// @yamada-ui/transitions/dist/chunk-7HOXUTFT.mjs
"use client"

// src/slide-fade.tsx
import { forwardRef } from "@yamada-ui/core";
import {
  AnimatePresence,
  transitionEnter,
  transitionExit,
  Motion
} from "@yamada-ui/motion";
import { useValue } from "@yamada-ui/use-value";
import { cx } from "@yamada-ui/utils";
import { jsx } from "react/jsx-runtime";
var variants = {
  initial: ({
    offsetX,
    offsetY,
    transition,
    transitionEnd,
    delay,
    duration,
    initial
  }) => ({
    opacity: 0,
    x: offsetX,
    y: offsetY,
    transition: transitionExit(transition == null ? void 0 : transition.exit)(delay, duration),
    transitionEnd: transitionEnd == null ? void 0 : transitionEnd.exit,
    ...initial
  }),
  enter: ({ transition, transitionEnd, delay, duration, enter } = {}) => ({
    opacity: 1,
    x: 0,
    y: 0,
    transition: transitionEnter(transition == null ? void 0 : transition.enter)(delay, duration),
    transitionEnd: transitionEnd == null ? void 0 : transitionEnd.enter,
    ...enter
  }),
  exit: ({
    offsetX,
    offsetY,
    reverse,
    transition,
    transitionEnd,
    delay,
    duration,
    exit
  } = {}) => ({
    opacity: 0,
    transition: transitionExit(transition == null ? void 0 : transition.exit)(delay, duration),
    ...reverse ? { x: offsetX, y: offsetY, transitionEnd: transitionEnd == null ? void 0 : transitionEnd.exit } : { transitionEnd: { x: offsetX, y: offsetY, ...transitionEnd == null ? void 0 : transitionEnd.exit } },
    ...exit
  })
};
var slideFadeProps = {
  initial: "exit",
  animate: "enter",
  exit: "exit",
  variants
};
var SlideFade = forwardRef(
  ({
    unmountOnExit,
    isOpen,
    offsetX: _offsetX = 0,
    offsetY: _offsetY = 8,
    reverse = true,
    transition,
    transitionEnd,
    delay,
    duration,
    className,
    ...rest
  }, ref) => {
    const animate = isOpen || unmountOnExit ? "enter" : "exit";
    const offsetX = useValue(_offsetX);
    const offsetY = useValue(_offsetY);
    const custom = {
      offsetX,
      offsetY,
      reverse,
      transition,
      transitionEnd,
      delay,
      duration
    };
    isOpen = unmountOnExit ? isOpen && unmountOnExit : true;
    const css = {
      w: "100%"
    };
    return /* @__PURE__ */ jsx(AnimatePresence, { custom, children: isOpen ? /* @__PURE__ */ jsx(
      Motion,
      {
        ref,
        className: cx("ui-slide-fade", className),
        custom,
        ...slideFadeProps,
        animate,
        __css: css,
        ...rest
      }
    ) : null });
  }
);

export {
  slideFadeProps,
  SlideFade
};
