// @yamada-ui/transitions/dist/slide-fade.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/slide-fade.tsx
var slide_fade_exports = {};
__export(slide_fade_exports, {
  SlideFade: () => SlideFade,
  slideFadeProps: () => slideFadeProps
});
module.exports = __toCommonJS(slide_fade_exports);
var import_core = require("@yamada-ui/core");
var import_motion = require("@yamada-ui/motion");
var import_use_value = require("@yamada-ui/use-value");
var import_utils = require("@yamada-ui/utils");
var import_jsx_runtime = require("react/jsx-runtime");
var variants = {
  initial: ({
    offsetX,
    offsetY,
    transition,
    transitionEnd,
    delay,
    duration,
    initial
  }) => ({
    opacity: 0,
    x: offsetX,
    y: offsetY,
    transition: (0, import_motion.transitionExit)(transition == null ? void 0 : transition.exit)(delay, duration),
    transitionEnd: transitionEnd == null ? void 0 : transitionEnd.exit,
    ...initial
  }),
  enter: ({ transition, transitionEnd, delay, duration, enter } = {}) => ({
    opacity: 1,
    x: 0,
    y: 0,
    transition: (0, import_motion.transitionEnter)(transition == null ? void 0 : transition.enter)(delay, duration),
    transitionEnd: transitionEnd == null ? void 0 : transitionEnd.enter,
    ...enter
  }),
  exit: ({
    offsetX,
    offsetY,
    reverse,
    transition,
    transitionEnd,
    delay,
    duration,
    exit
  } = {}) => ({
    opacity: 0,
    transition: (0, import_motion.transitionExit)(transition == null ? void 0 : transition.exit)(delay, duration),
    ...reverse ? { x: offsetX, y: offsetY, transitionEnd: transitionEnd == null ? void 0 : transitionEnd.exit } : { transitionEnd: { x: offsetX, y: offsetY, ...transitionEnd == null ? void 0 : transitionEnd.exit } },
    ...exit
  })
};
var slideFadeProps = {
  initial: "exit",
  animate: "enter",
  exit: "exit",
  variants
};
var SlideFade = (0, import_core.forwardRef)(
  ({
    unmountOnExit,
    isOpen,
    offsetX: _offsetX = 0,
    offsetY: _offsetY = 8,
    reverse = true,
    transition,
    transitionEnd,
    delay,
    duration,
    className,
    ...rest
  }, ref) => {
    const animate = isOpen || unmountOnExit ? "enter" : "exit";
    const offsetX = (0, import_use_value.useValue)(_offsetX);
    const offsetY = (0, import_use_value.useValue)(_offsetY);
    const custom = {
      offsetX,
      offsetY,
      reverse,
      transition,
      transitionEnd,
      delay,
      duration
    };
    isOpen = unmountOnExit ? isOpen && unmountOnExit : true;
    const css = {
      w: "100%"
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_motion.AnimatePresence, { custom, children: isOpen ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_motion.Motion,
      {
        ref,
        className: (0, import_utils.cx)("ui-slide-fade", className),
        custom,
        ...slideFadeProps,
        animate,
        __css: css,
        ...rest
      }
    ) : null });
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SlideFade,
  slideFadeProps
});
