// @yamada-ui/snacks/dist/chunk-RWY4IRNZ.mjs
"use client"

// src/use-snacks.tsx
import {
  Alert,
  AlertDescription,
  AlertIcon,
  AlertTitle
} from "@yamada-ui/alert";
import { CloseButton } from "@yamada-ui/close-button";
import { forwardRef, ui, useTheme } from "@yamada-ui/core";
import { cx, merge } from "@yamada-ui/utils";
import { useCallback, useMemo, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var createRender = (options) => {
  const { component } = options;
  const Render = (props) => {
    if (typeof component === "function") {
      return component({ ...props, ...options });
    } else {
      return /* @__PURE__ */ jsx(Snack, { ...props, ...options });
    }
  };
  return Render;
};
var counter = 0;
var useSnacks = (defaultOptions = {}) => {
  const [items, setItems] = useState([]);
  const { theme } = useTheme();
  const themeOptions = useMemo(
    () => {
      var _a, _b;
      return (_b = (_a = theme.__config.snacks) == null ? void 0 : _a.options) != null ? _b : {};
    },
    [theme]
  );
  const computedOptions = useMemo(
    () => merge(themeOptions, defaultOptions),
    [defaultOptions, themeOptions]
  );
  const { direction, startIndex, limit = 3 } = computedOptions;
  const getOptions = useCallback(
    (options) => merge(computedOptions, options),
    [computedOptions]
  );
  const snack = useMemo(() => {
    const methods = (options = {}) => {
      options = getOptions(options);
      const message = createRender(options);
      counter += 1;
      let {
        id = counter,
        duration,
        onCloseComplete,
        status,
        boxShadow,
        style
      } = options;
      const onClose = () => setItems((prev) => prev.filter((child) => child.id !== id));
      const snack2 = {
        id,
        status,
        duration,
        message,
        onClose,
        onCloseComplete,
        boxShadow,
        style
      };
      setItems((prev) => [
        ...prev.splice(-1 * ((limit != null ? limit : Infinity) - 1)),
        snack2
      ]);
      return id;
    };
    methods.update = (id, options) => {
      options = getOptions(options);
      setItems(
        (prev) => prev.map((props) => {
          if (props.id === id) {
            return { ...props, ...options, message: createRender(options) };
          } else {
            return props;
          }
        })
      );
    };
    methods.closeAll = () => {
      setItems([]);
    };
    methods.close = (id) => {
      setItems((prev) => prev.filter((props) => props.id !== id));
    };
    methods.isActive = (id) => !!items.find((props) => props.id === id);
    return methods;
  }, [items, limit, getOptions]);
  const snacks = useMemo(
    () => ({ direction, startIndex, items }),
    [direction, startIndex, items]
  );
  return { snack, snacks };
};
var defaultBoxShadow = [
  "0 -1px 3px 0 rgba(0, 0, 0, 0.1), 0 -1px 2px 0 rgba(0, 0, 0, 0.06), 0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)",
  "0px 0px 0px 1px rgba(0, 0, 0, 0.1), 0px -1px 2px rgba(0, 0, 0, 0.2), 0px 1px 2px rgba(0, 0, 0, 0.2)"
];
var Snack = forwardRef(
  ({
    variant = "basic",
    colorScheme,
    status,
    icon,
    title,
    description,
    isClosable = true,
    boxShadow = defaultBoxShadow,
    className,
    index,
    onClose
  }, ref) => {
    return /* @__PURE__ */ jsxs(
      Alert,
      {
        ref,
        className: cx("ui-snack", className),
        status,
        variant,
        colorScheme,
        alignItems: "start",
        boxShadow: index ? boxShadow : void 0,
        pe: isClosable ? 8 : void 0,
        children: [
          /* @__PURE__ */ jsx(
            AlertIcon,
            {
              variant: icon == null ? void 0 : icon.variant,
              className: "ui-snack__icon",
              ...(icon == null ? void 0 : icon.color) ? { color: icon.color } : {},
              children: icon == null ? void 0 : icon.children
            }
          ),
          /* @__PURE__ */ jsxs(ui.div, { flex: "1", children: [
            title ? /* @__PURE__ */ jsx(AlertTitle, { className: "ui-snack__title", lineClamp: 1, children: title }) : null,
            description ? /* @__PURE__ */ jsx(AlertDescription, { className: "ui-snack__desc", lineClamp: 3, children: description }) : null
          ] }),
          isClosable ? /* @__PURE__ */ jsx(
            CloseButton,
            {
              className: "ui-snack__close-button",
              size: "sm",
              onClick: onClose,
              position: "absolute",
              top: 2,
              right: 2
            }
          ) : null
        ]
      }
    );
  }
);

export {
  useSnacks
};
