// @yamada-ui/snacks/dist/snacks.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/snacks.tsx
var snacks_exports = {};
__export(snacks_exports, {
  Snacks: () => Snacks
});
module.exports = __toCommonJS(snacks_exports);
var import_core = require("@yamada-ui/core");
var import_motion = require("@yamada-ui/motion");
var import_use_timeout = require("@yamada-ui/use-timeout");
var import_use_token = require("@yamada-ui/use-token");
var import_use_value = require("@yamada-ui/use-value");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var defaultContainerVariants = {
  initial: { padding: 0 },
  animate: ({ gutter }) => ({
    padding: `${gutter[0]} 0 ${gutter[1]}`,
    transition: { duration: 0.4 }
  }),
  exit: { padding: 0 }
};
var defaultListVariants = {
  initial: { opacity: 1, height: 0 },
  animate: ({ height }) => ({
    opacity: 1,
    height,
    transition: { duration: 0.4 }
  }),
  exit: { opacity: 0, height: 0 }
};
var Snacks = (0, import_core.forwardRef)(
  ({
    className,
    snacks,
    containerVariants = defaultContainerVariants,
    listVariants = defaultListVariants,
    listProps,
    ...props
  }, ref) => {
    var _a, _b;
    const { items, ...computedSnacks } = snacks;
    const count = items.length;
    const refMap = (0, import_react.useRef)(/* @__PURE__ */ new Map());
    const [height, setHeight] = (0, import_react.useState)(0);
    const [isExist, setIsExist] = (0, import_react.useState)(!!count);
    const { theme } = (0, import_core.useTheme)();
    const {
      variants,
      direction = "top",
      startIndex = 0,
      gutter = [0, 0],
      gap = "fallback(4, 1rem)",
      negateMargin = true,
      ...rest
    } = (0, import_react.useMemo)(
      () => ({ ...computedSnacks, ...theme.__config.snacks, ...props }),
      [computedSnacks, theme, props]
    );
    const top = (_a = (0, import_use_token.useToken)("spaces", (0, import_use_value.useValue)(gutter[0]))) != null ? _a : 0;
    const bottom = (_b = (0, import_use_token.useToken)("spaces", (0, import_use_value.useValue)(gutter[1]))) != null ? _b : 0;
    const negatedTop = (0, import_utils.calc)(top).negate().toString();
    const negatedBottom = (0, import_utils.calc)(bottom).negate().toString();
    const isShow = !!count || isExist;
    const css = {
      w: "100%",
      var: [{ __prefix: "ui", name: "space", token: "spaces", value: gap }],
      margin: negateMargin ? `${negatedTop} 0 ${negatedBottom}` : void 0
    };
    (0, import_react.useEffect)(() => {
      let height2 = 0;
      if (!count)
        return;
      const refs = [...refMap.current.values()].slice(0, count);
      for (const ref2 of refs) {
        if (!ref2.current)
          continue;
        let { offsetHeight, offsetTop } = ref2.current;
        offsetHeight += offsetTop;
        if (offsetHeight > height2)
          height2 = offsetHeight;
      }
      setHeight(height2);
    }, [count, direction]);
    (0, import_utils.useUpdateEffect)(() => {
      if (!!count)
        setIsExist(true);
    }, [count]);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_motion.AnimatePresence, { initial: false, children: isShow ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_motion.Motion,
      {
        ref,
        className: (0, import_utils.cx)("ui-snacks", className),
        __css: css,
        variants: containerVariants,
        custom: { gutter: [top, bottom] },
        initial: "initial",
        animate: "animate",
        exit: "exit",
        ...rest,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_motion.Motion,
          {
            as: "ul",
            className: "ui-snacks__list",
            variants: listVariants,
            custom: { height },
            initial: "initial",
            animate: "animate",
            exit: "exit",
            __css: {
              position: "relative",
              w: "100%"
            },
            ...listProps,
            children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_motion.AnimatePresence,
              {
                onExitComplete: () => {
                  if (!count)
                    setIsExist(false);
                },
                children: items.map((props2, index) => {
                  const ref2 = (0, import_react.createRef)();
                  refMap.current.set(index, ref2);
                  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
                    SnackComponent,
                    {
                      ref: ref2,
                      index,
                      lastIndex: count - index - 1,
                      startIndex,
                      variants,
                      direction,
                      ...props2
                    },
                    props2.id
                  );
                })
              }
            )
          }
        )
      }
    ) : null });
  }
);
var defaultItemVariants = {
  initial: ({ index, direction }) => ({
    opacity: 0,
    ...index ? { y: (direction === "top" ? -1 : 1) * 16 } : {}
  }),
  animate: ({ index }) => ({
    opacity: 1,
    y: 0,
    transition: {
      delay: !index ? 0.4 : 0,
      duration: 0.4,
      ease: [0.4, 0, 0.2, 1]
    }
  }),
  exit: {
    opacity: 0,
    transition: {
      duration: 0.2,
      ease: [0.4, 0, 1, 1]
    }
  }
};
var SnackComponent = (0, import_react.memo)(
  (0, import_core.forwardRef)(
    ({
      index,
      lastIndex,
      startIndex,
      direction,
      variants = defaultItemVariants,
      duration = null,
      message,
      onClose: onCloseProp,
      onCloseComplete,
      style
    }, ref) => {
      const [delay, setDelay] = (0, import_react.useState)(duration);
      const isPresent = (0, import_motion.useIsPresent)();
      const onClose = isPresent ? onCloseProp : import_utils.noop;
      const onMouseEnter = () => setDelay(null);
      const onMouseLeave = () => setDelay(duration);
      const zIndex = startIndex + index;
      const space = `calc(var(--ui-space) * ${direction === "top" ? lastIndex : index})`;
      const css = {
        position: "absolute",
        top: space,
        left: 0,
        right: 0,
        zIndex,
        w: "100%",
        maxW: "100%",
        ...style
      };
      (0, import_utils.useUpdateEffect)(() => {
        if (!isPresent)
          onCloseComplete == null ? void 0 : onCloseComplete();
      }, [isPresent]);
      (0, import_utils.useUpdateEffect)(() => {
        setDelay(duration);
      }, [duration]);
      (0, import_use_timeout.useTimeout)(onClose, delay);
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_core.ui.li, { ref, className: "ui-snacks__item", __css: css, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_motion.Motion,
        {
          className: "ui-snacks__item-inner",
          layout: true,
          variants,
          custom: { index, lastIndex, direction },
          initial: "initial",
          animate: "animate",
          exit: "exit",
          onHoverStart: onMouseEnter,
          onHoverEnd: onMouseLeave,
          children: (0, import_utils.runIfFunc)(message, { index, onClose })
        }
      ) });
    }
  )
);
SnackComponent.displayName = "SnackComponent";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Snacks
});
