// @yamada-ui/utils/dist/calc.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/calc.ts
var calc_exports = {};
__export(calc_exports, {
  calc: () => calc
});
module.exports = __toCommonJS(calc_exports);
var toExpression = (operator, ...args) => args.join(` ${operator} `).replace(/calc/g, "");
var add = (...args) => `calc(${toExpression("+", ...args)})`;
var subtract = (...args) => `calc(${toExpression("-", ...args)})`;
var multiply = (...args) => `calc(${toExpression("*", ...args)})`;
var divide = (...args) => `calc(${toExpression("/", ...args)})`;
var negate = (value) => {
  if (value != null && !isNaN(parseFloat(value.toString())))
    return String(value).startsWith("-") ? String(value).slice(1) : `-${value}`;
  return multiply(value, -1);
};
var calc = Object.assign(
  (x) => ({
    add: (...args) => calc(add(x, ...args)),
    subtract: (...args) => calc(subtract(x, ...args)),
    multiply: (...args) => calc(multiply(x, ...args)),
    divide: (...args) => calc(divide(x, ...args)),
    negate: () => calc(negate(x)),
    toString: () => x.toString()
  }),
  {
    add,
    subtract,
    multiply,
    divide,
    negate
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  calc
});
