// @yamada-ui/utils/dist/chunk-P52P4NMT.mjs
// src/number.ts
var toNumber = (n) => {
  const num = parseFloat(n);
  return typeof num !== "number" || Number.isNaN(num) ? 0 : num;
};
var toPrecision = (n, precision) => {
  n = toNumber(n);
  const scale = 10 ** (precision != null ? precision : 10);
  n = Math.round(n * scale) / scale;
  return precision ? n.toFixed(precision) : n.toString();
};
var countDecimal = (n) => {
  if (!Number.isFinite(n))
    return 0;
  let e = 1;
  let p = 0;
  while (Math.round(n * e) / e !== n) {
    e *= 10;
    p += 1;
  }
  return p;
};
var roundNumberToStep = (n, from, step) => {
  const nextValue = Math.round((n - from) / step) * step + from;
  const precision = countDecimal(step);
  return toPrecision(nextValue, precision);
};
var valueToPercent = (n, min, max) => (n - min) * 100 / (max - min);
var percentToValue = (n, min, max) => (max - min) * n + min;
var clampNumber = (n, min, max) => Math.min(Math.max(n, min), max);

export {
  toNumber,
  toPrecision,
  countDecimal,
  roundNumberToStep,
  valueToPercent,
  percentToValue,
  clampNumber
};
