// @yamada-ui/utils/dist/dom.mjs
import {
  ariaAttr,
  createdDom,
  dataAttr,
  getActiveElement,
  getAllFocusable,
  getEventRelatedTarget,
  getOwnerDocument,
  getOwnerWindow,
  getPlatform,
  getPx,
  hasNegativeTabIndex,
  hasTabIndex,
  isActiveElement,
  isApple,
  isContains,
  isContentEditable,
  isDisabled,
  isElement,
  isFocusable,
  isHTMLElement,
  isHidden,
  isMac,
  isSafari,
  isTabbable,
  platform,
  vendor
} from "./chunk-YI55JNAD.mjs";
import "./chunk-RSDBT4DI.mjs";
import "./chunk-5OUUTZYQ.mjs";
import "./chunk-P52P4NMT.mjs";
import "./chunk-L76GM3GN.mjs";
import "./chunk-WV6P7ZCI.mjs";
import "./chunk-P5JCSYB3.mjs";
import "./chunk-L56KWSQO.mjs";
import "./chunk-A74F6ICI.mjs";
export {
  ariaAttr,
  createdDom,
  dataAttr,
  getActiveElement,
  getAllFocusable,
  getEventRelatedTarget,
  getOwnerDocument,
  getOwnerWindow,
  getPlatform,
  getPx,
  hasNegativeTabIndex,
  hasTabIndex,
  isActiveElement,
  isApple,
  isContains,
  isContentEditable,
  isDisabled,
  isElement,
  isFocusable,
  isHTMLElement,
  isHidden,
  isMac,
  isSafari,
  isTabbable,
  platform,
  vendor
};
