// @yamada-ui/utils/dist/index.mjs
import {
  alphaToHex,
  ariaAttr,
  assignAfter,
  assignRef,
  calcFormat,
  convertColor,
  createContext,
  createId,
  createdDom,
  cx,
  darkenColor,
  dataAttr,
  filterObject,
  filterUndefined,
  findChildren,
  flattenObject,
  funcAll,
  getActiveElement,
  getAllFocusable,
  getAlpha,
  getColor,
  getEventRelatedTarget,
  getMemoizedObject,
  getObject,
  getOwnerDocument,
  getOwnerWindow,
  getPlatform,
  getPx,
  getValidChildren,
  handlerAll,
  hasNegativeTabIndex,
  hasTabIndex,
  hslaTo,
  hsvTo,
  includesChildren,
  isAccessible,
  isActiveElement,
  isApple,
  isContains,
  isContentEditable,
  isDark,
  isDisabled,
  isElement,
  isFocusable,
  isGray,
  isHTMLElement,
  isHidden,
  isLight,
  isMac,
  isRefObject,
  isSafari,
  isTabbable,
  isTone,
  isValidElement,
  keysFormObject,
  lightenColor,
  memoizeObject,
  merge,
  mergeRefs,
  noop,
  objectFromEntries,
  omitChildren,
  omitObject,
  parseToHsla,
  parseToHsv,
  parseToRgba,
  pickChildren,
  pickObject,
  platform,
  randomColor,
  replaceObject,
  rgbaTo,
  runIfFunc,
  sameColor,
  shadeColor,
  splitObject,
  tintColor,
  tones,
  transparentizeColor,
  useAsync,
  useAsyncFunc,
  useAsyncRetry,
  useCallbackRef,
  useIsMounted,
  useMergeRefs,
  useSafeLayoutEffect,
  useUnmountEffect,
  useUpdateEffect,
  vendor
} from "./chunk-YI55JNAD.mjs";
import "./chunk-RSDBT4DI.mjs";
import {
  interopDefault
} from "./chunk-5OUUTZYQ.mjs";
import {
  clampNumber,
  countDecimal,
  percentToValue,
  roundNumberToStep,
  toNumber,
  toPrecision,
  valueToPercent
} from "./chunk-P52P4NMT.mjs";
import {
  antonym,
  escape,
  toCamelCase,
  toKebabCase
} from "./chunk-L76GM3GN.mjs";
import {
  filterEmpty
} from "./chunk-WV6P7ZCI.mjs";
import {
  cast,
  isArray,
  isBoolean,
  isEmpty,
  isFunction,
  isNotNumber,
  isNull,
  isNumber,
  isNumeric,
  isObject,
  isString,
  isUndefined,
  isUnit
} from "./chunk-P5JCSYB3.mjs";
import {
  calc
} from "./chunk-L56KWSQO.mjs";
import {
  addDomEvent,
  addPointerEvent,
  getEventPoint,
  getEventWindow,
  isMouseEvent,
  isMultiTouchEvent,
  isTouchEvent,
  pointFromMouse,
  pointFromTouch
} from "./chunk-A74F6ICI.mjs";
export {
  addDomEvent,
  addPointerEvent,
  alphaToHex,
  antonym,
  ariaAttr,
  assignAfter,
  assignRef,
  calc,
  calcFormat,
  cast,
  clampNumber,
  convertColor,
  countDecimal,
  createContext,
  createId,
  createdDom,
  cx,
  darkenColor,
  dataAttr,
  escape,
  filterEmpty,
  filterObject,
  filterUndefined,
  findChildren,
  flattenObject,
  funcAll,
  getActiveElement,
  getAllFocusable,
  getAlpha,
  getColor,
  getEventPoint,
  getEventRelatedTarget,
  getEventWindow,
  getMemoizedObject,
  getObject,
  getOwnerDocument,
  getOwnerWindow,
  getPlatform,
  getPx,
  getValidChildren,
  handlerAll,
  hasNegativeTabIndex,
  hasTabIndex,
  hslaTo,
  hsvTo,
  includesChildren,
  interopDefault,
  isAccessible,
  isActiveElement,
  isApple,
  isArray,
  isBoolean,
  isContains,
  isContentEditable,
  isDark,
  isDisabled,
  isElement,
  isEmpty,
  isFocusable,
  isFunction,
  isGray,
  isHTMLElement,
  isHidden,
  isLight,
  isMac,
  isMouseEvent,
  isMultiTouchEvent,
  isNotNumber,
  isNull,
  isNumber,
  isNumeric,
  isObject,
  isRefObject,
  isSafari,
  isString,
  isTabbable,
  isTone,
  isTouchEvent,
  isUndefined,
  isUnit,
  isValidElement,
  keysFormObject,
  lightenColor,
  memoizeObject,
  merge,
  mergeRefs,
  noop,
  objectFromEntries,
  omitChildren,
  omitObject,
  parseToHsla,
  parseToHsv,
  parseToRgba,
  percentToValue,
  pickChildren,
  pickObject,
  platform,
  pointFromMouse,
  pointFromTouch,
  randomColor,
  replaceObject,
  rgbaTo,
  roundNumberToStep,
  runIfFunc,
  sameColor,
  shadeColor,
  splitObject,
  tintColor,
  toCamelCase,
  toKebabCase,
  toNumber,
  toPrecision,
  tones,
  transparentizeColor,
  useAsync,
  useAsyncFunc,
  useAsyncRetry,
  useCallbackRef,
  useIsMounted,
  useMergeRefs,
  useSafeLayoutEffect,
  useUnmountEffect,
  useUpdateEffect,
  valueToPercent,
  vendor
};
