// @yamada-ui/checkbox/dist/checkbox-group.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/checkbox-group.tsx
var checkbox_group_exports = {};
__export(checkbox_group_exports, {
  CheckboxGroup: () => CheckboxGroup,
  useCheckboxGroup: () => useCheckboxGroup,
  useCheckboxGroupContext: () => useCheckboxGroupContext
});
module.exports = __toCommonJS(checkbox_group_exports);
var import_form_control2 = require("@yamada-ui/form-control");
var import_layouts = require("@yamada-ui/layouts");
var import_use_controllable_state = require("@yamada-ui/use-controllable-state");
var import_utils2 = require("@yamada-ui/utils");
var import_react2 = require("react");

// src/checkbox.tsx
var import_core = require("@yamada-ui/core");
var import_form_control = require("@yamada-ui/form-control");
var import_motion = require("@yamada-ui/motion");
var import_use_focus_visible = require("@yamada-ui/use-focus-visible");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var useCheckbox = ({
  id,
  ...props
}) => {
  id != null ? id : id = (0, import_react.useId)();
  const {
    id: _id,
    name,
    value,
    isChecked: isCheckedProp,
    defaultIsChecked,
    tabIndex,
    required,
    disabled,
    readOnly,
    isIndeterminate,
    onChange: onChangeProp,
    onFocus: onFocusProp,
    onBlur: onBlurProp,
    ...computedProps
  } = (0, import_form_control.useFormControlProps)({ id, ...props });
  const [{ "aria-readonly": _ariaReadonly, ...formControlProps }, rest] = (0, import_utils.splitObject)(computedProps, import_form_control.formControlProperties);
  const [isFocusVisible, setIsFocusVisible] = (0, import_react.useState)(false);
  const [isFocused, setFocused] = (0, import_react.useState)(false);
  const [isHovered, setHovered] = (0, import_react.useState)(false);
  const [isActive, setActive] = (0, import_react.useState)(false);
  const inputRef = (0, import_react.useRef)(null);
  const [isLabel, setIsLabel] = (0, import_react.useState)(true);
  const [isChecked, setIsChecked] = (0, import_react.useState)(!!defaultIsChecked);
  const isControlled = isCheckedProp !== void 0;
  const checked = isControlled ? isCheckedProp : isChecked;
  const onChange = (0, import_utils.useCallbackRef)(
    (ev) => {
      if (readOnly || disabled) {
        ev.preventDefault();
        return;
      }
      if (!isControlled)
        setIsChecked(!checked || isIndeterminate ? true : ev.target.checked);
      onChangeProp == null ? void 0 : onChangeProp(ev);
    },
    [readOnly, disabled, isControlled, checked, isIndeterminate]
  );
  const onFocus = (0, import_utils.useCallbackRef)(onFocusProp);
  const onBlur = (0, import_utils.useCallbackRef)(onBlurProp);
  const onKeyDown = (0, import_react.useCallback)(
    ({ key }) => {
      if (key === " ")
        setActive(true);
    },
    [setActive]
  );
  const onKeyUp = (0, import_react.useCallback)(
    ({ key }) => {
      if (key === " ")
        setActive(false);
    },
    [setActive]
  );
  (0, import_react.useEffect)(() => {
    return (0, import_use_focus_visible.trackFocusVisible)(setIsFocusVisible);
  }, []);
  (0, import_utils.useSafeLayoutEffect)(() => {
    if (inputRef.current)
      inputRef.current.indeterminate = Boolean(isIndeterminate);
  }, [isIndeterminate]);
  (0, import_utils.useUpdateEffect)(() => {
    if (disabled)
      setFocused(false);
  }, [disabled, setFocused]);
  (0, import_utils.useSafeLayoutEffect)(() => {
    var _a;
    if (!((_a = inputRef.current) == null ? void 0 : _a.form))
      return;
    inputRef.current.form.onreset = () => setIsChecked(!!defaultIsChecked);
  }, []);
  (0, import_utils.useSafeLayoutEffect)(() => {
    if (!inputRef.current)
      return;
    if (inputRef.current.checked !== checked)
      setIsChecked(inputRef.current.checked);
  }, [inputRef.current]);
  const getContainerProps = (0, import_react.useCallback)(
    (props2 = {}, ref = null) => ({
      ...formControlProps,
      ...props2,
      ref: (0, import_utils.mergeRefs)(ref, (el) => {
        if (el)
          setIsLabel(el.tagName === "LABEL");
      }),
      "data-checked": (0, import_utils.dataAttr)(checked),
      onClick: (0, import_utils.handlerAll)(props2.onClick, () => {
        var _a;
        if (isLabel)
          return;
        (_a = inputRef.current) == null ? void 0 : _a.click();
        requestAnimationFrame(() => {
          var _a2;
          return (_a2 = inputRef.current) == null ? void 0 : _a2.focus();
        });
      })
    }),
    [checked, isLabel, formControlProps]
  );
  const getIconProps = (0, import_react.useCallback)(
    (props2 = {}, ref = null) => ({
      ...formControlProps,
      ...props2,
      ref,
      "data-active": (0, import_utils.dataAttr)(isActive),
      "data-hover": (0, import_utils.dataAttr)(isHovered),
      "data-checked": (0, import_utils.dataAttr)(checked),
      "data-focus": (0, import_utils.dataAttr)(isFocused),
      "data-focus-visible": (0, import_utils.dataAttr)(isFocused && isFocusVisible),
      "data-indeterminate": (0, import_utils.dataAttr)(isIndeterminate),
      "aria-hidden": true,
      onMouseDown: (0, import_utils.handlerAll)(props2.onMouseDown, (ev) => {
        if (isFocused)
          ev.preventDefault();
        setActive(true);
      }),
      onMouseUp: (0, import_utils.handlerAll)(props2.onMouseUp, () => setActive(false)),
      onMouseEnter: (0, import_utils.handlerAll)(props2.onMouseEnter, () => setHovered(true)),
      onMouseLeave: (0, import_utils.handlerAll)(props2.onMouseLeave, () => setHovered(false))
    }),
    [
      isActive,
      checked,
      isFocused,
      isHovered,
      isFocusVisible,
      isIndeterminate,
      formControlProps
    ]
  );
  const getInputProps = (0, import_react.useCallback)(
    (props2 = {}, ref = null) => ({
      ...formControlProps,
      ...props2,
      ref: (0, import_utils.mergeRefs)(inputRef, ref),
      id,
      type: "checkbox",
      name,
      value,
      tabIndex,
      required,
      disabled,
      readOnly,
      checked,
      style: {
        border: "0px",
        clip: "rect(0px, 0px, 0px, 0px)",
        height: "1px",
        width: "1px",
        margin: "-1px",
        padding: "0px",
        overflow: "hidden",
        whiteSpace: "nowrap",
        position: "absolute"
      },
      onChange: (0, import_utils.handlerAll)(props2.onChange, onChange),
      onBlur: (0, import_utils.handlerAll)(props2.onBlur, onBlur, () => setFocused(false)),
      onFocus: (0, import_utils.handlerAll)(props2.onFocus, onFocus, () => setFocused(true)),
      onKeyDown: (0, import_utils.handlerAll)(props2.onKeyDown, onKeyDown),
      onKeyUp: (0, import_utils.handlerAll)(props2.onKeyUp, onKeyUp)
    }),
    [
      formControlProps,
      id,
      name,
      value,
      tabIndex,
      required,
      disabled,
      readOnly,
      checked,
      onChange,
      onBlur,
      onFocus,
      onKeyDown,
      onKeyUp
    ]
  );
  const getLabelProps = (0, import_react.useCallback)(
    (props2 = {}, ref = null) => ({
      ...formControlProps,
      ...props2,
      ref,
      "data-checked": (0, import_utils.dataAttr)(checked),
      onMouseDown: (0, import_utils.handlerAll)(props2.onMouseDown, (ev) => {
        ev.preventDefault();
        ev.stopPropagation();
      }),
      onTouchStart: (0, import_utils.handlerAll)(props2.onTouchStart, (ev) => {
        ev.preventDefault();
        ev.stopPropagation();
      })
    }),
    [checked, formControlProps]
  );
  return {
    props: rest,
    isFocusVisible,
    isFocused,
    isHovered,
    isActive,
    isChecked: checked,
    isIndeterminate,
    getContainerProps,
    getIconProps,
    getInputProps,
    getLabelProps
  };
};
var Checkbox = (0, import_react.forwardRef)(
  (props, ref) => {
    var _a, _b, _c, _d;
    const group = useCheckboxGroupContext();
    const { value: groupValue, ...groupProps } = { ...group };
    const control = (0, import_form_control.useFormControl)(props);
    const [styles, mergedProps] = (0, import_core.useMultiComponentStyle)("Checkbox", {
      ...groupProps,
      ...props
    });
    const {
      className,
      gap = "0.5rem",
      isRequired = (_a = groupProps.isRequired) != null ? _a : control.isRequired,
      isReadOnly = (_b = groupProps.isReadOnly) != null ? _b : control.isReadOnly,
      isDisabled = (_c = groupProps.isDisabled) != null ? _c : control.isDisabled,
      isInvalid = (_d = groupProps.isInvalid) != null ? _d : control.isInvalid,
      iconProps,
      inputProps,
      labelProps,
      children,
      ...computedProps
    } = (0, import_core.omitThemeProps)(mergedProps);
    const isCheckedProp = groupValue && computedProps.value ? groupValue.includes(computedProps.value) : computedProps.isChecked;
    const onChange = groupProps.onChange && computedProps.value ? (0, import_utils.funcAll)(groupProps.onChange, computedProps.onChange) : computedProps.onChange;
    const {
      isChecked,
      isIndeterminate,
      getContainerProps,
      getInputProps,
      getIconProps,
      getLabelProps,
      props: rest
    } = useCheckbox({
      ...computedProps,
      isRequired,
      isReadOnly,
      isDisabled,
      isInvalid,
      isChecked: isCheckedProp,
      onChange
    });
    const { children: customIcon, ...resolvedIconProps } = { ...iconProps };
    const cloneIcon = (0, import_react.cloneElement)(customIcon != null ? customIcon : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CheckboxIcon, {}), {
      __css: {
        opacity: isChecked || isIndeterminate ? 1 : 0,
        transform: isChecked || isIndeterminate ? "scale(1)" : "scale(0.95)",
        transitionProperty: "transform",
        transitionDuration: "normal"
      },
      isIndeterminate,
      isChecked,
      isRequired,
      isReadOnly,
      isDisabled,
      isInvalid
    });
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_core.ui.label,
      {
        className: (0, import_utils.cx)("ui-checkbox", className),
        ...getContainerProps(),
        __css: {
          cursor: "pointer",
          position: "relative",
          display: "inline-flex",
          alignItems: "center",
          verticalAlign: "top",
          gap,
          ...styles.container
        },
        ...rest,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_core.ui.input,
            {
              className: "ui-checkbox__input",
              ...getInputProps(inputProps, ref)
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_core.ui.span,
            {
              className: "ui-checkbox__icon",
              __css: {
                pointerEvents: isReadOnly ? "none" : void 0,
                position: "relative",
                display: "inline-block",
                userSelect: "none",
                ...styles.icon
              },
              ...getIconProps(resolvedIconProps),
              children: cloneIcon
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_core.ui.span,
            {
              className: "ui-checkbox__label",
              __css: { ...styles.label },
              ...getLabelProps(labelProps),
              children
            }
          )
        ]
      }
    );
  }
);
Checkbox.displayName = "Checkbox";
var CheckboxIcon = ({
  isIndeterminate,
  isChecked,
  isRequired: _isRequired,
  isReadOnly: _isReadOnly,
  isDisabled: _isDisabled,
  isInvalid: _isInvalid,
  ...rest
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_motion.AnimatePresence, { initial: false, children: isIndeterminate || isChecked ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_core.ui.div,
    {
      __css: {
        position: "absolute",
        top: "50%",
        left: "50%",
        transform: "translate(-50%, -50%)"
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_core.ui.div,
        {
          as: import_motion.motion.div,
          variants: {
            unchecked: { scale: 0.5 },
            checked: { scale: 1 }
          },
          initial: "unchecked",
          animate: "checked",
          exit: "unchecked",
          style: {
            display: "flex",
            alignItems: "center",
            justifyContent: "center"
          },
          children: isIndeterminate ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(IndeterminateIcon, { ...rest }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CheckIcon, { ...rest })
        }
      )
    }
  ) : null });
};
var CheckIcon = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_core.ui.svg,
    {
      as: import_motion.motion.svg,
      width: "1.2em",
      viewBox: "0 0 12 10",
      variants: {
        unchecked: {
          opacity: 0,
          strokeDashoffset: 16
        },
        checked: {
          opacity: 1,
          strokeDashoffset: 0,
          transition: { duration: 0.2 }
        }
      },
      style: {
        fill: "none",
        strokeWidth: 2,
        stroke: "currentColor",
        strokeDasharray: 16
      },
      ...props,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("polyline", { points: "1.5 6 4.5 9 10.5 1" })
    }
  );
};
var IndeterminateIcon = (props) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_core.ui.svg,
    {
      as: import_motion.motion.svg,
      width: "1.2em",
      viewBox: "0 0 24 24",
      variants: {
        unchecked: {
          scaleX: 0.65,
          opacity: 0
        },
        checked: {
          scaleX: 1,
          opacity: 1,
          transition: {
            scaleX: { duration: 0 },
            opacity: { duration: 0.02 }
          }
        }
      },
      style: { stroke: "currentColor", strokeWidth: 4 },
      ...props,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("line", { x1: "21", x2: "3", y1: "12", y2: "12" })
    }
  );
};

// src/checkbox-group.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
var isEvent = (value) => value && (0, import_utils2.isObject)(value) && (0, import_utils2.isObject)(value.target);
var useCheckboxGroup = ({
  value: valueProp,
  defaultValue = [],
  onChange: onChangeProp,
  isNative,
  ...props
}) => {
  const onChangeRef = (0, import_utils2.useCallbackRef)(onChangeProp);
  const [value, setValue] = (0, import_use_controllable_state.useControllableState)({
    value: valueProp,
    defaultValue,
    onChange: onChangeRef
  });
  const onChange = (0, import_react2.useCallback)(
    (evOrValue) => {
      const isChecked = isEvent(evOrValue) ? evOrValue.target.checked : !value.includes(evOrValue);
      const selectedValue = isEvent(evOrValue) ? evOrValue.target.value : evOrValue;
      const nextValue = isChecked ? [...value, selectedValue] : value.filter((v) => String(v) !== String(selectedValue));
      setValue(nextValue);
    },
    [value, setValue]
  );
  const getCheckboxProps = (0, import_react2.useCallback)(
    (props2, ref = null) => ({
      ...props2,
      ref,
      [isNative ? "checked" : "isChecked"]: value.some(
        (val) => String(props2 == null ? void 0 : props2.value) === String(val)
      ),
      onChange
    }),
    [onChange, isNative, value]
  );
  return { props, value, setValue, onChange, getCheckboxProps };
};
var [CheckboxGroupProvider, useCheckboxGroupContext] = (0, import_utils2.createContext)({
  strict: false,
  name: "CheckboxGroupContext"
});
var CheckboxGroup = (0, import_react2.forwardRef)(
  ({
    className,
    size,
    variant,
    colorScheme,
    children,
    items = [],
    direction = "column",
    gap,
    ...props
  }, ref) => {
    const { value, onChange, props: computedProps } = useCheckboxGroup(props);
    const { isRequired, isReadOnly, isDisabled, isInvalid, ...rest } = (0, import_form_control2.useFormControl)(computedProps);
    const validChildren = (0, import_utils2.getValidChildren)(children);
    let computedChildren = [];
    if (!validChildren.length && items.length) {
      computedChildren = items.map(({ label, value: value2, ...props2 }, i) => /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(Checkbox, { value: value2, ...props2, children: label }, i));
    }
    return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
      CheckboxGroupProvider,
      {
        value: {
          size,
          variant,
          colorScheme,
          isRequired,
          isReadOnly,
          isDisabled,
          isInvalid,
          value,
          onChange
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
          import_layouts.Flex,
          {
            ref,
            className: (0, import_utils2.cx)("ui-checkbox-group", className),
            role: "group",
            direction,
            gap: gap != null ? gap : direction === "row" ? "1rem" : void 0,
            ...rest,
            children: children != null ? children : computedChildren
          }
        )
      }
    );
  }
);
CheckboxGroup.displayName = "CheckboxGroup";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CheckboxGroup,
  useCheckboxGroup,
  useCheckboxGroupContext
});
