// @yamada-ui/layouts/dist/chunk-XKPB42SF.mjs
"use client"

// src/aspect-ratio.tsx
import { ui, forwardRef } from "@yamada-ui/core";
import { cx, replaceObject } from "@yamada-ui/utils";
import { Children } from "react";
import { jsx } from "react/jsx-runtime";
var AspectRatio = forwardRef(
  ({ className, ratio = 4 / 3, children, ...rest }, ref) => {
    const child = Children.only(children);
    const css = {
      position: "relative",
      _before: {
        content: `""`,
        display: "block",
        h: 0,
        pb: replaceObject(ratio, (r) => `${1 / r * 100}%`)
      },
      "& > *:not(style)": {
        overflow: "hidden",
        position: "absolute",
        top: "0",
        right: "0",
        bottom: "0",
        left: "0",
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        w: "100%",
        h: "100%"
      },
      "& > img, & > video": {
        objectFit: "cover"
      }
    };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        ref,
        className: cx("ui-aspect-ratio", className),
        __css: css,
        ...rest,
        children: child
      }
    );
  }
);

export {
  AspectRatio
};
