// @yamada-ui/layouts/dist/flex.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/flex.tsx
var flex_exports = {};
__export(flex_exports, {
  Flex: () => Flex,
  Wrap: () => Wrap
});
module.exports = __toCommonJS(flex_exports);
var import_core = require("@yamada-ui/core");
var import_jsx_runtime = require("react/jsx-runtime");
var Flex = (0, import_core.forwardRef)(
  ({
    direction: flexDirection,
    justify: justifyContent,
    align: alignItems,
    wrap: flexWrap,
    basis: flexBasis,
    grow: flexGrow,
    shrink: flexShrink,
    ...rest
  }, ref) => {
    const css = {
      display: "flex",
      flexDirection,
      justifyContent,
      alignItems,
      flexWrap,
      flexBasis,
      flexGrow,
      flexShrink
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_core.ui.div, { ref, __css: css, ...rest });
  }
);
var Wrap = (0, import_core.forwardRef)((props, ref) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Flex, { ref, wrap: "wrap", ...props }));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Flex,
  Wrap
});
