// @yamada-ui/infinite-scroll-area/dist/chunk-YKTVPMSD.mjs
"use client"

// src/infinite-scroll-area.tsx
import { ui, forwardRef } from "@yamada-ui/core";
import { useInfiniteScroll } from "@yamada-ui/use-infinite-scroll";
import { cx, mergeRefs } from "@yamada-ui/utils";
import { useMemo, useRef } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var InfiniteScrollArea = forwardRef(
  ({
    rootRef: rootRefProp,
    orientation = "vertical",
    rootMargin,
    threshold,
    startIndex,
    onLoad,
    resetRef,
    indexRef,
    className,
    children,
    triggerProps,
    loading,
    finish,
    isDisabled,
    isReverse,
    initialLoad,
    ...rest
  }, ref) => {
    const isVertical = orientation === "vertical";
    const rootRef = useRef(null);
    const { ref: triggerRef, isFinish } = useInfiniteScroll({
      orientation,
      rootRef: rootRefProp != null ? rootRefProp : rootRef,
      rootMargin,
      threshold,
      startIndex,
      onLoad,
      resetRef,
      indexRef,
      isDisabled,
      isReverse,
      initialLoad
    });
    const css = useMemo(
      () => ({
        w: "100%",
        display: "flex",
        flexDirection: isVertical ? "column" : "row",
        gap: "1rem"
      }),
      [isVertical]
    );
    const hasFinish = !!finish;
    const showTrigger = !isDisabled && (hasFinish || !isFinish);
    return /* @__PURE__ */ jsxs(
      ui.div,
      {
        ref: mergeRefs(rootRef, ref),
        tabIndex: 0,
        className: cx("ui-infinite-scroll-area", className),
        __css: css,
        ...rest,
        children: [
          isReverse && showTrigger ? /* @__PURE__ */ jsx(InfiniteScrollTrigger, { ref: triggerRef, ...triggerProps, children: isFinish ? finish : loading }) : null,
          children,
          !isReverse && showTrigger ? /* @__PURE__ */ jsx(InfiniteScrollTrigger, { ref: triggerRef, ...triggerProps, children: isFinish ? finish : loading }) : null
        ]
      }
    );
  }
);
var InfiniteScrollTrigger = forwardRef(
  ({ className, ...rest }, ref) => {
    const css = useMemo(
      () => ({
        w: "100%",
        flex: 1,
        display: "flex",
        justifyContent: "center",
        alignItems: "center"
      }),
      []
    );
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        ref,
        className: cx("ui-infinite-scroll-area__trigger", className),
        __css: css,
        ...rest
      }
    );
  }
);

export {
  InfiniteScrollArea
};
