// @yamada-ui/infinite-scroll-area/dist/infinite-scroll-area.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/infinite-scroll-area.tsx
var infinite_scroll_area_exports = {};
__export(infinite_scroll_area_exports, {
  InfiniteScrollArea: () => InfiniteScrollArea
});
module.exports = __toCommonJS(infinite_scroll_area_exports);
var import_core = require("@yamada-ui/core");
var import_use_infinite_scroll = require("@yamada-ui/use-infinite-scroll");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var InfiniteScrollArea = (0, import_core.forwardRef)(
  ({
    rootRef: rootRefProp,
    orientation = "vertical",
    rootMargin,
    threshold,
    startIndex,
    onLoad,
    resetRef,
    indexRef,
    className,
    children,
    triggerProps,
    loading,
    finish,
    isDisabled,
    isReverse,
    initialLoad,
    ...rest
  }, ref) => {
    const isVertical = orientation === "vertical";
    const rootRef = (0, import_react.useRef)(null);
    const { ref: triggerRef, isFinish } = (0, import_use_infinite_scroll.useInfiniteScroll)({
      orientation,
      rootRef: rootRefProp != null ? rootRefProp : rootRef,
      rootMargin,
      threshold,
      startIndex,
      onLoad,
      resetRef,
      indexRef,
      isDisabled,
      isReverse,
      initialLoad
    });
    const css = (0, import_react.useMemo)(
      () => ({
        w: "100%",
        display: "flex",
        flexDirection: isVertical ? "column" : "row",
        gap: "1rem"
      }),
      [isVertical]
    );
    const hasFinish = !!finish;
    const showTrigger = !isDisabled && (hasFinish || !isFinish);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      import_core.ui.div,
      {
        ref: (0, import_utils.mergeRefs)(rootRef, ref),
        tabIndex: 0,
        className: (0, import_utils.cx)("ui-infinite-scroll-area", className),
        __css: css,
        ...rest,
        children: [
          isReverse && showTrigger ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(InfiniteScrollTrigger, { ref: triggerRef, ...triggerProps, children: isFinish ? finish : loading }) : null,
          children,
          !isReverse && showTrigger ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(InfiniteScrollTrigger, { ref: triggerRef, ...triggerProps, children: isFinish ? finish : loading }) : null
        ]
      }
    );
  }
);
var InfiniteScrollTrigger = (0, import_core.forwardRef)(
  ({ className, ...rest }, ref) => {
    const css = (0, import_react.useMemo)(
      () => ({
        w: "100%",
        flex: 1,
        display: "flex",
        justifyContent: "center",
        alignItems: "center"
      }),
      []
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core.ui.div,
      {
        ref,
        className: (0, import_utils.cx)("ui-infinite-scroll-area__trigger", className),
        __css: css,
        ...rest
      }
    );
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  InfiniteScrollArea
});
