// @yamada-ui/providers/dist/ui-provider.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ui-provider.tsx
var ui_provider_exports = {};
__export(ui_provider_exports, {
  UIProvider: () => UIProvider
});
module.exports = __toCommonJS(ui_provider_exports);
var import_core = require("@yamada-ui/core");
var import_loading = require("@yamada-ui/loading");
var import_motion = require("@yamada-ui/motion");
var import_notice = require("@yamada-ui/notice");
var import_theme = require("@yamada-ui/theme");

// src/environment-provider.tsx
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var environment = {
  getDocument: () => document,
  getWindow: () => window
};
var EnvironmentContext = (0, import_react.createContext)(environment);
var EnvironmentProvider = ({
  children,
  environment: environment2,
  disabled
}) => {
  const ref = (0, import_react.useRef)(null);
  const context = (0, import_react.useMemo)(() => {
    if (environment2)
      return environment2;
    return {
      getDocument: () => {
        var _a, _b;
        return (_b = (_a = ref.current) == null ? void 0 : _a.ownerDocument) != null ? _b : document;
      },
      getWindow: () => {
        var _a, _b;
        return (_b = (_a = ref.current) == null ? void 0 : _a.ownerDocument.defaultView) != null ? _b : window;
      }
    };
  }, [environment2]);
  const enabled = !disabled || !environment2;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(EnvironmentContext.Provider, { value: context, children: [
    children,
    enabled ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { id: "__ui_dev", hidden: true, ref }) : null
  ] });
};
EnvironmentProvider.displayName = "EnvironmentProvider";

// src/ui-provider.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
var UIProvider = ({
  theme = import_theme.defaultTheme,
  config = import_theme.defaultConfig,
  disableResetStyle,
  disableGlobalStyle,
  colorModeManager,
  colorModeStorageKey,
  themeSchemeManager,
  themeSchemeStorageKey,
  environment: environment2,
  disableEnvironment,
  children
}) => {
  var _a;
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
    import_core.ThemeProvider,
    {
      theme,
      config,
      themeSchemeManager,
      storageKey: themeSchemeStorageKey,
      children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
        import_core.ColorModeProvider,
        {
          colorModeManager,
          storageKey: colorModeStorageKey,
          config,
          children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
            EnvironmentProvider,
            {
              environment: environment2,
              disabled: disableEnvironment,
              children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_motion.MotionConfig, { ...(_a = config.motion) == null ? void 0 : _a.config, children: /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(import_loading.LoadingProvider, { ...config.loading, children: [
                !disableResetStyle ? /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_core.ResetStyle, {}) : null,
                !disableGlobalStyle ? /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_core.GlobalStyle, {}) : null,
                children,
                /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_notice.NoticeProvider, { ...config.notice })
              ] }) })
            }
          )
        }
      )
    }
  );
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  UIProvider
});
