// @yamada-ui/link/dist/chunk-6VXZCMHN.mjs
"use client"

// src/link-box.tsx
import {
  ui,
  forwardRef,
  omitThemeProps,
  useComponentStyle
} from "@yamada-ui/core";
import { cx } from "@yamada-ui/utils";
import { jsx } from "react/jsx-runtime";
var LinkOverlay = forwardRef(
  ({ className, isExternal, target, rel, href, children, ...rest }, ref) => {
    const css = {
      position: "static",
      _before: {
        content: '""',
        cursor: "inherit",
        display: "block",
        position: "absolute",
        top: 0,
        left: 0,
        zIndex: 0,
        width: "100%",
        height: "100%"
      }
    };
    return /* @__PURE__ */ jsx(
      ui.a,
      {
        ref,
        target: isExternal ? "_blank" : target,
        rel: isExternal ? "noopener" : rel,
        className: cx("ui-link-box__overlay", className),
        href,
        __css: css,
        ...rest,
        children
      }
    );
  }
);
var LinkBox = forwardRef((props, ref) => {
  const [styles, mergedProps] = useComponentStyle("LinkBox", props);
  const { className, children, ...rest } = omitThemeProps(mergedProps);
  const css = {
    "a[href]:not(.ui-link-box__overlay), abbr[title]": {
      position: "relative",
      zIndex: "fallback(yamcha, 1)"
    },
    ...styles
  };
  return /* @__PURE__ */ jsx(
    ui.div,
    {
      ref,
      className: cx("ui-link-box", className),
      __css: css,
      position: "relative",
      ...rest,
      children
    }
  );
});

export {
  LinkOverlay,
  LinkBox
};
