// @yamada-ui/link/dist/link-box.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/link-box.tsx
var link_box_exports = {};
__export(link_box_exports, {
  LinkBox: () => LinkBox,
  LinkOverlay: () => LinkOverlay
});
module.exports = __toCommonJS(link_box_exports);
var import_core = require("@yamada-ui/core");
var import_utils = require("@yamada-ui/utils");
var import_jsx_runtime = require("react/jsx-runtime");
var LinkOverlay = (0, import_core.forwardRef)(
  ({ className, isExternal, target, rel, href, children, ...rest }, ref) => {
    const css = {
      position: "static",
      _before: {
        content: '""',
        cursor: "inherit",
        display: "block",
        position: "absolute",
        top: 0,
        left: 0,
        zIndex: 0,
        width: "100%",
        height: "100%"
      }
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core.ui.a,
      {
        ref,
        target: isExternal ? "_blank" : target,
        rel: isExternal ? "noopener" : rel,
        className: (0, import_utils.cx)("ui-link-box__overlay", className),
        href,
        __css: css,
        ...rest,
        children
      }
    );
  }
);
var LinkBox = (0, import_core.forwardRef)((props, ref) => {
  const [styles, mergedProps] = (0, import_core.useComponentStyle)("LinkBox", props);
  const { className, children, ...rest } = (0, import_core.omitThemeProps)(mergedProps);
  const css = {
    "a[href]:not(.ui-link-box__overlay), abbr[title]": {
      position: "relative",
      zIndex: "fallback(yamcha, 1)"
    },
    ...styles
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_core.ui.div,
    {
      ref,
      className: (0, import_utils.cx)("ui-link-box", className),
      __css: css,
      position: "relative",
      ...rest,
      children
    }
  );
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LinkBox,
  LinkOverlay
});
