// @yamada-ui/menu/dist/index.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Menu: () => Menu,
  MenuButton: () => MenuButton,
  MenuCommand: () => MenuCommand,
  MenuDivider: () => MenuDivider,
  MenuGroup: () => MenuGroup,
  MenuIcon: () => MenuIcon,
  MenuItem: () => MenuItem,
  MenuItemButton: () => MenuItemButton,
  MenuList: () => MenuList,
  MenuOptionGroup: () => MenuOptionGroup,
  MenuOptionItem: () => MenuOptionItem
});
module.exports = __toCommonJS(src_exports);

// src/menu.tsx
var import_core2 = require("@yamada-ui/core");
var import_popover = require("@yamada-ui/popover");
var import_use_descendant = require("@yamada-ui/use-descendant");
var import_use_disclosure = require("@yamada-ui/use-disclosure");
var import_utils2 = require("@yamada-ui/utils");
var import_react2 = require("react");

// src/menu-item.tsx
var import_core = require("@yamada-ui/core");
var import_use_clickable = require("@yamada-ui/use-clickable");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var [UpstreamMenuItemProvider, useUpstreamMenuItem] = (0, import_utils.createContext)({
  strict: false,
  name: "UpstreamMenuItemContext"
});
var isTargetMenuItem = (target) => {
  var _a;
  return (0, import_utils.isHTMLElement)(target) && !!((_a = target == null ? void 0 : target.getAttribute("role")) == null ? void 0 : _a.startsWith("menu-item"));
};
var MenuItem = (0, import_core.forwardRef)(
  ({
    as,
    className,
    type,
    isDisabled,
    isFocusable,
    closeOnSelect: customCloseOnSelect,
    icon,
    command,
    children,
    onClick: onClickProp,
    onFocus: onFocusProp,
    onMouseEnter: onMouseEnterProp,
    onMouseMove: onMouseMoveProp,
    onMouseLeave: onMouseLeaveProp,
    onKeyDown: onKeyDownProp,
    ...props
  }, ref) => {
    var _a;
    const {
      focusedIndex,
      setFocusedIndex,
      isOpen,
      onClose,
      onUpstreamClose,
      closeOnSelect: generalCloseOnSelect,
      menuRef,
      requestAnimationFrameId,
      isNested,
      styles
    } = useMenu();
    const { onUpstreamRestoreFocus } = (_a = useUpstreamMenuItem()) != null ? _a : {};
    const trulyDisabled = isDisabled && !isFocusable;
    const buttonRef = (0, import_react.useRef)(null);
    const hasDownstreamRef = (0, import_react.useRef)(false);
    const onKeyDownRef = (0, import_react.useRef)(
      () => void 0
    );
    const { index, register } = useMenuDescendant({ disabled: trulyDisabled });
    const [isDownstreamOpen, setDownstreamOpen] = (0, import_react.useState)(false);
    const isFocused = index === focusedIndex;
    const onMouseEnter = (0, import_react.useCallback)(
      (event) => {
        onMouseEnterProp == null ? void 0 : onMouseEnterProp(event);
        if (isDisabled)
          return;
        setFocusedIndex(index);
      },
      [setFocusedIndex, index, isDisabled, onMouseEnterProp]
    );
    const onMouseMove = (0, import_react.useCallback)(
      (event) => {
        onMouseMoveProp == null ? void 0 : onMouseMoveProp(event);
        if (buttonRef.current && !(0, import_utils.isActiveElement)(buttonRef.current))
          onMouseEnter(event);
      },
      [onMouseEnter, onMouseMoveProp]
    );
    const onMouseLeave = (0, import_react.useCallback)(
      (event) => {
        onMouseLeaveProp == null ? void 0 : onMouseLeaveProp(event);
        if (isDisabled)
          return;
        setFocusedIndex(-1);
      },
      [setFocusedIndex, isDisabled, onMouseLeaveProp]
    );
    const onClick = (0, import_react.useCallback)(
      (ev) => {
        onClickProp == null ? void 0 : onClickProp(ev);
        if (!isTargetMenuItem(ev.currentTarget))
          return;
        const hasDownstream = hasDownstreamRef.current;
        if (customCloseOnSelect != null ? customCloseOnSelect : !hasDownstream && generalCloseOnSelect) {
          onClose();
          onUpstreamClose == null ? void 0 : onUpstreamClose();
        }
      },
      [
        onClickProp,
        customCloseOnSelect,
        generalCloseOnSelect,
        onClose,
        onUpstreamClose
      ]
    );
    const onFocus = (0, import_react.useCallback)(
      (ev) => {
        onFocusProp == null ? void 0 : onFocusProp(ev);
        setFocusedIndex(index);
      },
      [onFocusProp, setFocusedIndex, index]
    );
    const onRestoreFocus = (0, import_react.useCallback)(() => {
      var _a2;
      (_a2 = buttonRef.current) == null ? void 0 : _a2.focus();
      setFocusedIndex(index);
    }, [setFocusedIndex, index]);
    const onKeyDown = (0, import_react.useCallback)(
      (ev) => {
        onKeyDownProp == null ? void 0 : onKeyDownProp(ev);
        const actions = {
          ArrowLeft: isNested ? (0, import_utils.funcAll)(onUpstreamRestoreFocus, onClose) : void 0
        };
        const action = actions[ev.key];
        if (!action)
          return;
        ev.preventDefault();
        ev.stopPropagation();
        action();
      },
      [onKeyDownProp, onUpstreamRestoreFocus, onClose, isNested]
    );
    const rest = (0, import_use_clickable.useClickable)({
      focusOnClick: false,
      ...props,
      onClick,
      onFocus,
      onMouseEnter,
      onMouseMove,
      onMouseLeave,
      onKeyDown: (0, import_utils.handlerAll)(onKeyDown, onKeyDownRef.current),
      ref: (0, import_utils.mergeRefs)(register, buttonRef, ref),
      isDisabled,
      isFocusable
    });
    (0, import_utils.useUpdateEffect)(() => {
      if (!isOpen)
        return;
      const id = requestAnimationFrameId.current;
      if (isFocused && !trulyDisabled && buttonRef.current) {
        if (id)
          cancelAnimationFrame(id);
        requestAnimationFrameId.current = requestAnimationFrame(() => {
          var _a2;
          (_a2 = buttonRef.current) == null ? void 0 : _a2.focus({ preventScroll: true });
          requestAnimationFrameId.current = null;
        });
      } else if (menuRef.current && !(0, import_utils.isActiveElement)(menuRef.current)) {
        menuRef.current.focus({ preventScroll: true });
      }
      return () => {
        if (id)
          cancelAnimationFrame(id);
      };
    }, [isFocused, trulyDisabled, menuRef, isOpen]);
    type = as || type ? type != null ? type : void 0 : "button";
    children = icon || command ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_core.ui.span, { style: { flex: 1 }, children }) : children;
    const css = {
      textDecoration: "none",
      color: "inherit",
      userSelect: "none",
      display: "flex",
      width: "100%",
      alignItems: "center",
      textAlign: "start",
      flex: "0 0 auto",
      outline: 0,
      gap: "0.75rem",
      ...styles.item
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      UpstreamMenuItemProvider,
      {
        value: {
          onKeyDownRef,
          onUpstreamRestoreFocus: onRestoreFocus,
          setDownstreamOpen,
          hasDownstreamRef
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
          import_core.ui.li,
          {
            ...rest,
            ...isDownstreamOpen ? { "data-active": "" } : {},
            as,
            type,
            role: "menu-item",
            tabIndex: isFocused ? 0 : -1,
            className: (0, import_utils.cx)("ui-menu__item", className),
            __css: css,
            children: [
              icon ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MenuIcon, { children: icon }) : null,
              children,
              command ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MenuCommand, { children: command }) : null
            ]
          }
        )
      }
    );
  }
);
var MenuOptionItem = (0, import_core.forwardRef)(
  ({ className, icon, isChecked, closeOnSelect = false, children, ...rest }, ref) => {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
      MenuItem,
      {
        ref,
        className: (0, import_utils.cx)("ui-menu__item--option", className),
        "aria-checked": (0, import_utils.ariaAttr)(isChecked),
        closeOnSelect,
        ...rest,
        children: [
          icon !== null ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(MenuIcon, { opacity: isChecked ? 1 : 0, children: icon || /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CheckIcon, {}) }) : null,
          children
        ]
      }
    );
  }
);
var MenuIcon = (0, import_core.forwardRef)(
  ({ className, ...rest }, ref) => {
    const { styles } = useMenu();
    const css = {
      flexShrink: 0,
      display: "inline-flex",
      justifyContent: "center",
      alignItems: "center",
      fontSize: "0.85em",
      ...styles.icon
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core.ui.span,
      {
        ref,
        className: (0, import_utils.cx)("ui-menu__item__icon", className),
        __css: css,
        ...rest
      }
    );
  }
);
var MenuCommand = (0, import_core.forwardRef)(
  ({ className, ...rest }, ref) => {
    const { styles } = useMenu();
    const css = { ...styles.command };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core.ui.span,
      {
        ref,
        className: (0, import_utils.cx)("ui-menu__item__command", className),
        __css: css,
        ...rest
      }
    );
  }
);
var CheckIcon = () => /* @__PURE__ */ (0, import_jsx_runtime.jsx)("svg", { viewBox: "0 0 14 14", width: "1em", height: "1em", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  "polygon",
  {
    fill: "currentColor",
    points: "5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
  }
) });

// src/menu.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
var {
  DescendantsContextProvider,
  useDescendantsContext: useMenuDescendantsContext,
  useDescendants,
  useDescendant: useMenuDescendant
} = (0, import_use_descendant.createDescendant)();
var [MenuProvider, useMenu] = (0, import_utils2.createContext)({
  name: "MenuContext",
  errorMessage: `useMenu returned is 'undefined'. Seems you forgot to wrap the components in "<Menu />"`
});
var [UpstreamMenuProvider, useUpstreamMenu] = (0, import_utils2.createContext)({
  strict: false,
  name: "UpstreamMenuContext"
});
var Menu = (props) => {
  var _a, _b;
  const [styles, mergedProps] = (0, import_core2.useMultiComponentStyle)("Menu", props);
  let {
    initialFocusRef,
    closeOnSelect = true,
    closeOnBlur,
    placement,
    duration = 0.2,
    offset,
    onOpen: onOpenProp,
    onClose: onCloseProp,
    ...rest
  } = (0, import_core2.omitThemeProps)(mergedProps);
  const { relatedRef, onUpstreamClose, onDownstreamCloseMapRef } = (_a = useUpstreamMenu()) != null ? _a : {};
  const { setDownstreamOpen, hasDownstreamRef } = (_b = useUpstreamMenuItem()) != null ? _b : {};
  const isNested = !!relatedRef;
  if (isNested) {
    placement != null ? placement : placement = "right-start";
    offset != null ? offset : offset = [-8, 8];
    closeOnBlur != null ? closeOnBlur : closeOnBlur = false;
  } else {
    placement != null ? placement : placement = "bottom-start";
  }
  const descendants = useDescendants();
  const [focusedIndex, setFocusedIndex] = (0, import_react2.useState)(-1);
  const menuRef = (0, import_react2.useRef)(null);
  const timeoutIds = (0, import_react2.useRef)(/* @__PURE__ */ new Set([]));
  const requestAnimationFrameId = (0, import_react2.useRef)(null);
  const onCloseMapRef = (0, import_react2.useRef)(/* @__PURE__ */ new Map());
  const onFocusMenu = (0, import_react2.useCallback)(() => {
    requestAnimationFrame(
      () => {
        var _a2;
        return (_a2 = menuRef.current) == null ? void 0 : _a2.focus({ preventScroll: false });
      }
    );
  }, []);
  const onFocusFirstItem = (0, import_react2.useCallback)(() => {
    const id2 = setTimeout(() => {
      if (initialFocusRef)
        return;
      const first = descendants.enabledFirstValue();
      if (first)
        setFocusedIndex(first.index);
    });
    timeoutIds.current.add(id2);
  }, [descendants, initialFocusRef]);
  const onFocusLastItem = (0, import_react2.useCallback)(() => {
    const id2 = setTimeout(() => {
      if (initialFocusRef)
        return;
      const last = descendants.enabledLastValue();
      if (last)
        setFocusedIndex(last.index);
    });
    timeoutIds.current.add(id2);
  }, [descendants, initialFocusRef]);
  const onOpenInternal = (0, import_react2.useCallback)(() => {
    onOpenProp == null ? void 0 : onOpenProp();
    if (!isNested)
      onFocusMenu();
  }, [onOpenProp, isNested, onFocusMenu]);
  const onCloseInternal = (0, import_react2.useCallback)(() => {
    onCloseProp == null ? void 0 : onCloseProp();
    for (const onClose2 of onCloseMapRef.current.values()) {
      onClose2();
    }
  }, [onCloseProp]);
  const id = (0, import_react2.useId)();
  const { isOpen, onOpen, onClose } = (0, import_use_disclosure.useDisclosure)({
    ...props,
    onOpen: onOpenInternal,
    onClose: onCloseInternal
  });
  (0, import_react2.useEffect)(() => {
    const onCloseMap = onDownstreamCloseMapRef == null ? void 0 : onDownstreamCloseMapRef.current;
    onCloseMap == null ? void 0 : onCloseMap.set(id, onClose);
    return () => {
      onCloseMap == null ? void 0 : onCloseMap.delete(id);
    };
  }, [id, onClose, onDownstreamCloseMapRef]);
  (0, import_react2.useEffect)(() => {
    if (setDownstreamOpen)
      setDownstreamOpen(isOpen);
  }, [setDownstreamOpen, isOpen]);
  (0, import_react2.useEffect)(() => {
    if (hasDownstreamRef)
      hasDownstreamRef.current = true;
    return () => {
      if (hasDownstreamRef)
        hasDownstreamRef.current = false;
    };
  });
  (0, import_utils2.useUpdateEffect)(() => {
    if (!isOpen)
      setFocusedIndex(-1);
  }, [isOpen]);
  (0, import_utils2.useUnmountEffect)(() => {
    timeoutIds.current.forEach((id2) => clearTimeout(id2));
    timeoutIds.current.clear();
  });
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(DescendantsContextProvider, { value: descendants, children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
    UpstreamMenuProvider,
    {
      value: {
        relatedRef: menuRef,
        onDownstreamCloseMapRef: onCloseMapRef,
        onUpstreamClose: (0, import_utils2.funcAll)(onUpstreamClose, onClose)
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
        MenuProvider,
        {
          value: {
            isOpen,
            onOpen,
            onClose,
            onUpstreamClose,
            onFocusFirstItem,
            onFocusLastItem,
            closeOnSelect,
            focusedIndex,
            setFocusedIndex,
            menuRef,
            requestAnimationFrameId,
            isNested,
            styles
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
            import_popover.Popover,
            {
              ...{
                trigger: isNested ? "hover" : "click",
                ...rest,
                isOpen,
                onOpen,
                onClose,
                placement,
                offset,
                duration,
                initialFocusRef,
                relatedRef,
                closeOnButton: false,
                closeOnBlur
              }
            }
          )
        }
      )
    }
  ) });
};

// src/menu-button.tsx
var import_core3 = require("@yamada-ui/core");
var import_icon = require("@yamada-ui/icon");
var import_popover2 = require("@yamada-ui/popover");
var import_utils3 = require("@yamada-ui/utils");
var import_react3 = require("react");
var import_jsx_runtime3 = require("react/jsx-runtime");
var MenuButton = (0, import_core3.forwardRef)(
  ({ className, children, as: As, ...rest }, ref) => {
    var _a;
    const { onKeyDownRef, onUpstreamRestoreFocus } = (_a = useUpstreamMenuItem()) != null ? _a : {};
    const { isOpen, onOpen, onClose, onFocusFirstItem, onFocusLastItem } = useMenu();
    const onKeyDown = (0, import_react3.useCallback)(
      (ev) => {
        const actions = {
          Enter: (0, import_utils3.funcAll)(onOpen, onFocusFirstItem),
          ArrowDown: (0, import_utils3.funcAll)(onOpen, onFocusFirstItem),
          ArrowUp: (0, import_utils3.funcAll)(onOpen, onFocusLastItem)
        };
        const action = actions[ev.key];
        if (!action)
          return;
        ev.preventDefault();
        action();
      },
      [onFocusFirstItem, onFocusLastItem, onOpen]
    );
    const onItemKeyDown = (0, import_react3.useCallback)(
      (ev) => {
        const actions = {
          ArrowRight: !isOpen ? (0, import_utils3.funcAll)(onOpen, onFocusFirstItem) : void 0,
          ArrowLeft: isOpen ? (0, import_utils3.funcAll)(onUpstreamRestoreFocus, onClose) : void 0
        };
        const action = actions[ev.key];
        if (!action)
          return;
        ev.preventDefault();
        action();
      },
      [isOpen, onOpen, onFocusFirstItem, onUpstreamRestoreFocus, onClose]
    );
    (0, import_utils3.assignRef)(onKeyDownRef, onItemKeyDown);
    const Component = As ? (0, import_core3.ui)(As) : Button;
    return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(import_popover2.PopoverTrigger, { children: /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      Component,
      {
        ref,
        className: (0, import_utils3.cx)("ui-menu", className),
        "aria-haspopup": "menu",
        "aria-expanded": isOpen,
        ...rest,
        "data-active": (0, import_utils3.dataAttr)(isOpen),
        onKeyDown: (0, import_utils3.handlerAll)(rest.onKeyDown, onKeyDown),
        children
      }
    ) });
  }
);
var Button = (0, import_core3.forwardRef)((rest, ref) => {
  const { styles } = useMenu();
  const css = {
    display: "inline-flex",
    appearance: "none",
    alignItems: "center",
    outline: 0,
    ...styles.button
  };
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(import_core3.ui.button, { ref, __css: css, ...rest });
});
var MenuItemButton = (0, import_core3.forwardRef)(
  ({ className, children, innerProps, icon, iconProps, ...rest }, ref) => {
    return /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(
      MenuButton,
      {
        ref,
        className: (0, import_utils3.cx)("ui-menu__item-button", className),
        flex: "1",
        ...rest,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(import_core3.ui.span, { as: "span", flex: "1", ...innerProps, children }),
          /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(MenuIcon, { ...iconProps, children: icon != null ? icon : /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(import_icon.ChevronIcon, { fontSize: "1.5em", transform: "rotate(-90deg)" }) })
        ]
      }
    );
  }
);

// src/menu-list.tsx
var import_core4 = require("@yamada-ui/core");
var import_popover3 = require("@yamada-ui/popover");
var import_utils4 = require("@yamada-ui/utils");
var import_react4 = require("react");
var import_jsx_runtime4 = require("react/jsx-runtime");
var MenuList = (0, import_core4.forwardRef)(
  ({ className, ...rest }, ref) => {
    const { menuRef, focusedIndex, setFocusedIndex, onClose, styles } = useMenu();
    const descendants = useMenuDescendantsContext();
    const onNext = (0, import_react4.useCallback)(() => {
      const next = descendants.enabledNextValue(focusedIndex);
      if (next)
        setFocusedIndex(next.index);
    }, [descendants, focusedIndex, setFocusedIndex]);
    const onPrev = (0, import_react4.useCallback)(() => {
      const prev = descendants.enabledPrevValue(focusedIndex);
      if (prev)
        setFocusedIndex(prev.index);
    }, [descendants, focusedIndex, setFocusedIndex]);
    const onFirst = (0, import_react4.useCallback)(() => {
      const first = descendants.enabledFirstValue();
      if (first)
        setFocusedIndex(first.index);
    }, [descendants, setFocusedIndex]);
    const onLast = (0, import_react4.useCallback)(() => {
      const last = descendants.enabledLastValue();
      if (last)
        setFocusedIndex(last.index);
    }, [descendants, setFocusedIndex]);
    const onKeyDown = (0, import_react4.useCallback)(
      (ev) => {
        const actions = {
          Tab: (ev2) => ev2.preventDefault(),
          Escape: onClose,
          ArrowDown: focusedIndex === -1 ? onFirst : onNext,
          ArrowUp: focusedIndex === -1 ? onLast : onPrev,
          Home: onFirst,
          End: onLast
        };
        const action = actions[ev.key];
        if (!action)
          return;
        ev.preventDefault();
        action(ev);
      },
      [focusedIndex, onClose, onFirst, onLast, onNext, onPrev]
    );
    const css = { ...styles.list };
    return /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
      import_popover3.PopoverContent,
      {
        as: "ul",
        ref: (0, import_utils4.mergeRefs)(menuRef, ref),
        className: (0, import_utils4.cx)("ui-menu__list", className),
        role: "menu",
        tabIndex: -1,
        __css: css,
        ...rest,
        onKeyDown: (0, import_utils4.handlerAll)(rest.onKeyDown, onKeyDown)
      }
    );
  }
);

// src/menu-group.tsx
var import_core5 = require("@yamada-ui/core");
var import_utils5 = require("@yamada-ui/utils");
var import_jsx_runtime5 = require("react/jsx-runtime");
var MenuGroup = (0, import_core5.forwardRef)(
  ({ className, label, children, ...rest }, ref) => {
    const { styles } = useMenu();
    const css = { ...styles.group };
    return /* @__PURE__ */ (0, import_jsx_runtime5.jsxs)(
      import_core5.ui.li,
      {
        ref,
        className: (0, import_utils5.cx)("ui-menu__item", "ui-menu__item--group", className),
        role: "group",
        __css: css,
        ...rest,
        children: [
          label ? /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
            import_core5.ui.span,
            {
              className: (0, import_utils5.cx)("ui-menu__item--group-label"),
              __css: styles.groupLabel,
              children: label
            }
          ) : null,
          /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(import_core5.ui.ul, { className: "ui-menu__item__group", children })
        ]
      }
    );
  }
);

// src/menu-option-group.tsx
var import_use_controllable_state = require("@yamada-ui/use-controllable-state");
var import_utils6 = require("@yamada-ui/utils");
var import_react5 = require("react");
var import_jsx_runtime6 = require("react/jsx-runtime");
var MenuOptionGroup = (0, import_react5.forwardRef)(
  ({
    className,
    value: valueProp,
    defaultValue,
    onChange: onChangeProp,
    type,
    children,
    ...rest
  }, ref) => {
    const isRadio = type === "radio";
    defaultValue != null ? defaultValue : defaultValue = isRadio ? "" : [];
    const [value, setValue] = (0, import_use_controllable_state.useControllableState)({
      value: valueProp,
      defaultValue,
      onChange: onChangeProp
    });
    const onChange = (0, import_react5.useCallback)(
      (selectedValue) => {
        if (isRadio && typeof value === "string")
          setValue(selectedValue);
        if (!isRadio && (0, import_utils6.isArray)(value)) {
          const nextValue = value.includes(selectedValue) ? value.filter((item) => item !== selectedValue) : value.concat(selectedValue);
          setValue(nextValue);
        }
      },
      [isRadio, value, setValue]
    );
    const validChildren = (0, import_utils6.getValidChildren)(children);
    const cloneChildren = validChildren.map((child) => {
      if (child.type !== MenuOptionItem)
        return child;
      const onClick = (ev) => {
        var _a, _b;
        onChange(child.props.value);
        (_b = (_a = child.props).onClick) == null ? void 0 : _b.call(_a, ev);
      };
      const isChecked = !isRadio && (0, import_utils6.isArray)(value) ? value.includes(child.props.value) : child.props.value === value;
      return (0, import_react5.cloneElement)(child, { type, onClick, isChecked });
    });
    return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
      MenuGroup,
      {
        ref,
        className: (0, import_utils6.cx)("ui-menu__item--group--option", className),
        ...rest,
        children: cloneChildren
      }
    );
  }
);
MenuOptionGroup.displayName = "MenuOptionGroup";

// src/menu-divider.tsx
var import_core6 = require("@yamada-ui/core");
var import_utils7 = require("@yamada-ui/utils");
var import_jsx_runtime7 = require("react/jsx-runtime");
var MenuDivider = (0, import_core6.forwardRef)(({ className, ...rest }, ref) => {
  const { styles } = useMenu();
  const css = { ...styles.divider };
  return /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
    import_core6.ui.hr,
    {
      ref,
      className: (0, import_utils7.cx)("ui-menu__divider", className),
      __css: css,
      ...rest
    }
  );
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Menu,
  MenuButton,
  MenuCommand,
  MenuDivider,
  MenuGroup,
  MenuIcon,
  MenuItem,
  MenuItemButton,
  MenuList,
  MenuOptionGroup,
  MenuOptionItem
});
