// @yamada-ui/menu/dist/menu-group.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/menu-group.tsx
var menu_group_exports = {};
__export(menu_group_exports, {
  MenuGroup: () => MenuGroup
});
module.exports = __toCommonJS(menu_group_exports);
var import_core2 = require("@yamada-ui/core");
var import_utils2 = require("@yamada-ui/utils");

// src/menu.tsx
var import_core = require("@yamada-ui/core");
var import_popover = require("@yamada-ui/popover");
var import_use_descendant = require("@yamada-ui/use-descendant");
var import_use_disclosure = require("@yamada-ui/use-disclosure");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var {
  DescendantsContextProvider,
  useDescendantsContext: useMenuDescendantsContext,
  useDescendants,
  useDescendant: useMenuDescendant
} = (0, import_use_descendant.createDescendant)();
var [MenuProvider, useMenu] = (0, import_utils.createContext)({
  name: "MenuContext",
  errorMessage: `useMenu returned is 'undefined'. Seems you forgot to wrap the components in "<Menu />"`
});
var [UpstreamMenuProvider, useUpstreamMenu] = (0, import_utils.createContext)({
  strict: false,
  name: "UpstreamMenuContext"
});

// src/menu-group.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
var MenuGroup = (0, import_core2.forwardRef)(
  ({ className, label, children, ...rest }, ref) => {
    const { styles } = useMenu();
    const css = { ...styles.group };
    return /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(
      import_core2.ui.li,
      {
        ref,
        className: (0, import_utils2.cx)("ui-menu__item", "ui-menu__item--group", className),
        role: "group",
        __css: css,
        ...rest,
        children: [
          label ? /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
            import_core2.ui.span,
            {
              className: (0, import_utils2.cx)("ui-menu__item--group-label"),
              __css: styles.groupLabel,
              children: label
            }
          ) : null,
          /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_core2.ui.ul, { className: "ui-menu__item__group", children })
        ]
      }
    );
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MenuGroup
});
