// @yamada-ui/notice/dist/notice.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/notice.tsx
var notice_exports = {};
__export(notice_exports, {
  noticeStore: () => noticeStore,
  useNotice: () => useNotice
});
module.exports = __toCommonJS(notice_exports);
var import_alert = require("@yamada-ui/alert");
var import_close_button = require("@yamada-ui/close-button");
var import_core = require("@yamada-ui/core");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var findId = (options, id) => options.find((notice) => notice.id === id);
var findNotice = (state, id) => {
  const placement = getNoticePlacement(state, id);
  const index = placement ? state[placement].findIndex((notice) => notice.id === id) : -1;
  return { placement, index };
};
var getNoticePlacement = (state, id) => {
  for (const [placement, values] of Object.entries(state)) {
    if (findId(values, id))
      return placement;
  }
};
var counter = 0;
var createNotice = (message, {
  id,
  placement = "top",
  duration,
  onCloseComplete,
  status,
  style
}) => {
  counter += 1;
  id != null ? id : id = counter;
  return {
    id,
    placement,
    status,
    duration,
    message,
    onDelete: () => noticeStore.remove(String(id), placement),
    isDelete: false,
    onCloseComplete,
    style
  };
};
var createRender = (options) => {
  const { component } = options;
  const Render = (props) => {
    if (typeof component === "function") {
      return component({ ...props, ...options });
    } else {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Notice, { ...props, ...options });
    }
  };
  return Render;
};
var createNoticeFunc = (defaultOptions, theme) => {
  var _a, _b;
  const themeOptions = (_b = (_a = theme.__config.notice) == null ? void 0 : _a.options) != null ? _b : {};
  const computedOptions = (options) => (0, import_utils.merge)(themeOptions, (0, import_utils.merge)(defaultOptions, options));
  const notice = (options = {}) => {
    options = computedOptions(options);
    const message = createRender(options);
    return noticeStore.create(message, options);
  };
  notice.update = (id, options) => {
    options = computedOptions(options);
    noticeStore.update(id, options);
  };
  notice.closeAll = noticeStore.closeAll;
  notice.close = noticeStore.close;
  notice.isActive = noticeStore.isActive;
  return notice;
};
var useNotice = (defaultOptions) => {
  const { theme } = (0, import_core.useTheme)();
  return (0, import_react.useMemo)(
    () => createNoticeFunc(defaultOptions != null ? defaultOptions : {}, theme),
    [defaultOptions, theme]
  );
};
var initialState = {
  top: [],
  "top-left": [],
  "top-right": [],
  bottom: [],
  "bottom-left": [],
  "bottom-right": []
};
var createNoticeStore = (initialState2) => {
  let state = initialState2;
  const storeChangeCache = /* @__PURE__ */ new Set();
  const setState = (setStateFunc) => {
    state = setStateFunc(state);
    storeChangeCache.forEach((onStoreChange) => onStoreChange());
  };
  return {
    getSnapshot: () => state,
    subscribe: (onStoreChange) => {
      storeChangeCache.add(onStoreChange);
      return () => {
        setState(() => initialState2);
        storeChangeCache.delete(onStoreChange);
      };
    },
    remove: (id, placement) => {
      setState((prevState) => ({
        ...prevState,
        [placement]: prevState[placement].filter((notice) => notice.id != id)
      }));
    },
    create: (message, options) => {
      const limit = options.limit;
      const notice = createNotice(message, options);
      const { placement, id } = notice;
      setState((prev) => {
        var _a;
        let prevNotices = (_a = prev[placement]) != null ? _a : [];
        if (limit !== void 0 && limit > 0 && prevNotices.length > limit - 1) {
          const n = prevNotices.length - (limit - 1);
          const notices2 = placement.includes("top") ? prevNotices.slice(n * -1) : prevNotices.slice(0, n);
          const ids = notices2.map(({ id: id2 }) => id2);
          prevNotices = prevNotices.map(
            (notice2) => ids.includes(notice2.id) ? { ...notice2, isDelete: true } : notice2
          );
        }
        const notices = placement.includes("top") ? [notice, ...prevNotices] : [...prevNotices, notice];
        return { ...prev, [placement]: notices };
      });
      return id;
    },
    update: (id, options) => {
      setState((prev) => {
        const next = { ...prev };
        const { placement, index } = findNotice(next, id);
        if (placement && index !== -1) {
          next[placement][index] = {
            ...next[placement][index],
            ...options,
            message: createRender(options)
          };
        }
        return next;
      });
    },
    closeAll: ({ placement } = {}) => {
      setState((prev) => {
        let placements = [
          "bottom",
          "bottom-right",
          "bottom-left",
          "top",
          "top-left",
          "top-right"
        ];
        if (placement)
          placements = placement;
        return placements.reduce(
          (acc, placement2) => {
            acc[placement2] = prev[placement2].map((notice) => ({
              ...notice,
              isDelete: true
            }));
            return acc;
          },
          { ...prev }
        );
      });
    },
    close: (id) => {
      setState((prev) => {
        const placement = getNoticePlacement(prev, id);
        if (!placement)
          return prev;
        return {
          ...prev,
          [placement]: prev[placement].map(
            (notice) => notice.id == id ? { ...notice, isDelete: true } : notice
          )
        };
      });
    },
    isActive: (id) => Boolean(findNotice(noticeStore.getSnapshot(), id).placement)
  };
};
var noticeStore = createNoticeStore(initialState);
var Notice = ({
  variant = "basic",
  colorScheme,
  status,
  icon,
  title,
  description,
  isClosable,
  className,
  onClose
}) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_alert.Alert,
    {
      status,
      variant,
      colorScheme,
      alignItems: "start",
      boxShadow: "fallback(lg, 0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05))",
      className: (0, import_utils.cx)("ui-notice", className),
      pe: isClosable ? 8 : void 0,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_alert.AlertIcon,
          {
            variant: icon == null ? void 0 : icon.variant,
            className: "ui-notice__icon",
            ...(icon == null ? void 0 : icon.color) ? { color: icon.color } : {},
            children: icon == null ? void 0 : icon.children
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_core.ui.div, { flex: "1", children: [
          title ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_alert.AlertTitle, { className: "ui-notice__title", lineClamp: 1, children: title }) : null,
          description ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_alert.AlertDescription, { className: "ui-notice__desc", lineClamp: 3, children: description }) : null
        ] }),
        isClosable ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_close_button.CloseButton,
          {
            className: "ui-notice__close-button",
            size: "sm",
            onClick: onClose,
            position: "absolute",
            top: 2,
            right: 2
          }
        ) : null
      ]
    }
  );
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  noticeStore,
  useNotice
});
