// @yamada-ui/use-clickable/dist/index.mjs
"use client"

// src/index.ts
import { useEventListeners } from "@yamada-ui/use-event-listener";
import { dataAttr, mergeRefs } from "@yamada-ui/utils";
import { useCallback, useState } from "react";
var isValidElement = (ev) => {
  const { tagName, isContentEditable } = ev.target;
  return tagName !== "INPUT" && tagName !== "TEXTAREA" && isContentEditable !== true;
};
var useClickable = ({
  ref,
  isDisabled,
  isFocusable,
  clickOnEnter = true,
  clickOnSpace = true,
  focusOnClick = true,
  onMouseDown,
  onMouseUp,
  onClick,
  onKeyDown,
  onKeyUp,
  tabIndex: _tabIndex,
  onMouseOver,
  onMouseLeave,
  ...props
} = {}) => {
  const [isButton, setIsButton] = useState(true);
  const [isPressed, setIsPressed] = useState(false);
  const listeners = useEventListeners();
  const tabIndex = isButton ? _tabIndex : _tabIndex || 0;
  const trulyDisabled = isDisabled && !isFocusable;
  const refCb = (node) => {
    if (!node)
      return;
    if (node.tagName !== "BUTTON")
      setIsButton(false);
  };
  const handleClick = useCallback(
    (ev) => {
      if (isDisabled) {
        ev.stopPropagation();
        ev.preventDefault();
        return;
      }
      if (focusOnClick)
        ev.currentTarget.focus();
      onClick == null ? void 0 : onClick(ev);
    },
    [isDisabled, focusOnClick, onClick]
  );
  const onDocumentKeyUp = useCallback(
    (ev) => {
      if (isPressed && isValidElement(ev)) {
        ev.preventDefault();
        ev.stopPropagation();
        setIsPressed(false);
        listeners.remove(document, "keyup", onDocumentKeyUp, false);
      }
    },
    [isPressed, listeners]
  );
  const handleKeyDown = useCallback(
    (ev) => {
      onKeyDown == null ? void 0 : onKeyDown(ev);
      if (isDisabled || ev.defaultPrevented || ev.metaKey)
        return;
      if (!isValidElement(ev.nativeEvent) || isButton)
        return;
      if (clickOnSpace && ev.key === " ") {
        ev.preventDefault();
        setIsPressed(true);
      }
      if (clickOnEnter && ev.key === "Enter") {
        ev.preventDefault();
        ev.currentTarget.click();
      }
      listeners.add(document, "keyup", onDocumentKeyUp, false);
    },
    [
      isDisabled,
      isButton,
      onKeyDown,
      clickOnEnter,
      clickOnSpace,
      listeners,
      onDocumentKeyUp
    ]
  );
  const handleKeyUp = useCallback(
    (ev) => {
      onKeyUp == null ? void 0 : onKeyUp(ev);
      if (isDisabled || ev.defaultPrevented || ev.metaKey)
        return;
      if (!isValidElement(ev.nativeEvent) || isButton)
        return;
      if (clickOnSpace && ev.key === " ") {
        ev.preventDefault();
        setIsPressed(false);
        ev.currentTarget.click();
      }
    },
    [clickOnSpace, isButton, isDisabled, onKeyUp]
  );
  const onDocumentMouseUp = useCallback(
    (ev) => {
      if (ev.button !== 0)
        return;
      setIsPressed(false);
      listeners.remove(document, "mouseup", onDocumentMouseUp, false);
    },
    [listeners]
  );
  const handleMouseDown = useCallback(
    (ev) => {
      if (ev.button !== 0)
        return;
      if (isDisabled) {
        ev.stopPropagation();
        ev.preventDefault();
        return;
      }
      if (!isButton)
        setIsPressed(true);
      ev.currentTarget.focus({ preventScroll: true });
      listeners.add(document, "mouseup", onDocumentMouseUp, false);
      onMouseDown == null ? void 0 : onMouseDown(ev);
    },
    [isDisabled, isButton, onMouseDown, listeners, onDocumentMouseUp]
  );
  const handleMouseUp = useCallback(
    (ev) => {
      if (ev.button !== 0)
        return;
      if (!isButton)
        setIsPressed(false);
      onMouseUp == null ? void 0 : onMouseUp(ev);
    },
    [onMouseUp, isButton]
  );
  const handleMouseOver = useCallback(
    (ev) => {
      if (isDisabled) {
        ev.preventDefault();
        return;
      }
      onMouseOver == null ? void 0 : onMouseOver(ev);
    },
    [isDisabled, onMouseOver]
  );
  const handleMouseLeave = useCallback(
    (ev) => {
      if (isPressed) {
        ev.preventDefault();
        setIsPressed(false);
      }
      onMouseLeave == null ? void 0 : onMouseLeave(ev);
    },
    [isPressed, onMouseLeave]
  );
  if (isButton) {
    return {
      ...props,
      ref: mergeRefs(ref, refCb),
      type: "button",
      "aria-disabled": trulyDisabled ? void 0 : isDisabled,
      disabled: trulyDisabled,
      onClick: handleClick,
      onMouseDown,
      onMouseUp,
      onKeyUp,
      onKeyDown,
      onMouseOver,
      onMouseLeave
    };
  } else {
    return {
      ...props,
      ref: mergeRefs(ref, refCb),
      role: "button",
      "data-active": dataAttr(isPressed),
      "aria-disabled": isDisabled ? "true" : void 0,
      tabIndex: trulyDisabled ? void 0 : tabIndex,
      onClick: handleClick,
      onMouseDown: handleMouseDown,
      onMouseUp: handleMouseUp,
      onKeyUp: handleKeyUp,
      onKeyDown: handleKeyDown,
      onMouseOver: handleMouseOver,
      onMouseLeave: handleMouseLeave
    };
  }
};
export {
  useClickable
};
