// @yamada-ui/tag/dist/chunk-VEIAKGIF.mjs
"use client"

// src/tag.tsx
import {
  ui,
  forwardRef,
  useMultiComponentStyle,
  omitThemeProps
} from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { useClickable } from "@yamada-ui/use-clickable";
import { cx } from "@yamada-ui/utils";
import { useRef } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var Tag = forwardRef((props, ref) => {
  const [styles, mergedProps] = useMultiComponentStyle("Tag", props);
  const {
    className,
    leftIcon,
    rightIcon,
    onClose,
    isDisabled,
    children,
    ...rest
  } = omitThemeProps(mergedProps);
  const css = {
    maxW: "100%",
    display: "inline-flex",
    alignItems: "center",
    gap: "fallback(1, 0.25rem)",
    verticalAlign: "top",
    ...styles.container
  };
  return /* @__PURE__ */ jsxs(
    ui.span,
    {
      ref,
      className: cx("ui-tag", className),
      __css: css,
      ...rest,
      children: [
        leftIcon,
        /* @__PURE__ */ jsx(TagLabel, { children }),
        rightIcon,
        onClose ? /* @__PURE__ */ jsx(CloseButton, { isDisabled, onClick: onClose, children: /* @__PURE__ */ jsx(CloseIcon, {}) }) : null
      ]
    }
  );
});
var TagLabel = ({ children, ...props }) => {
  const [styles] = useMultiComponentStyle("Tag", props);
  const css = styles.label;
  return /* @__PURE__ */ jsx(ui.span, { lineClamp: 1, __css: css, children });
};
var CloseIcon = () => {
  return /* @__PURE__ */ jsx(Icon, { verticalAlign: "inherit", viewBox: "0 0 512 512", fontSize: "1.125rem", children: /* @__PURE__ */ jsx(
    "path",
    {
      fill: "currentColor",
      d: "M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
    }
  ) });
};
var CloseButton = ({ children, ...props }) => {
  const ref = useRef(null);
  const [styles] = useMultiComponentStyle("Tag", props);
  const css = {
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    outline: "0",
    cursor: "pointer",
    ...styles.closeButton
  };
  const rest = useClickable({ ref, ...props });
  return /* @__PURE__ */ jsx(ui.span, { "aria-label": "Close tag", __css: css, ...rest, children });
};

export {
  Tag
};
