// @yamada-ui/tag/dist/index.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Tag: () => Tag
});
module.exports = __toCommonJS(src_exports);

// src/tag.tsx
var import_core = require("@yamada-ui/core");
var import_icon = require("@yamada-ui/icon");
var import_use_clickable = require("@yamada-ui/use-clickable");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var Tag = (0, import_core.forwardRef)((props, ref) => {
  const [styles, mergedProps] = (0, import_core.useMultiComponentStyle)("Tag", props);
  const {
    className,
    leftIcon,
    rightIcon,
    onClose,
    isDisabled,
    children,
    ...rest
  } = (0, import_core.omitThemeProps)(mergedProps);
  const css = {
    maxW: "100%",
    display: "inline-flex",
    alignItems: "center",
    gap: "fallback(1, 0.25rem)",
    verticalAlign: "top",
    ...styles.container
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_core.ui.span,
    {
      ref,
      className: (0, import_utils.cx)("ui-tag", className),
      __css: css,
      ...rest,
      children: [
        leftIcon,
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TagLabel, { children }),
        rightIcon,
        onClose ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CloseButton, { isDisabled, onClick: onClose, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CloseIcon, {}) }) : null
      ]
    }
  );
});
var TagLabel = ({ children, ...props }) => {
  const [styles] = (0, import_core.useMultiComponentStyle)("Tag", props);
  const css = styles.label;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_core.ui.span, { lineClamp: 1, __css: css, children });
};
var CloseIcon = () => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icon.Icon, { verticalAlign: "inherit", viewBox: "0 0 512 512", fontSize: "1.125rem", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "path",
    {
      fill: "currentColor",
      d: "M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z"
    }
  ) });
};
var CloseButton = ({ children, ...props }) => {
  const ref = (0, import_react.useRef)(null);
  const [styles] = (0, import_core.useMultiComponentStyle)("Tag", props);
  const css = {
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    outline: "0",
    cursor: "pointer",
    ...styles.closeButton
  };
  const rest = (0, import_use_clickable.useClickable)({ ref, ...props });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_core.ui.span, { "aria-label": "Close tag", __css: css, ...rest, children });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Tag
});
