// @yamada-ui/reorder/dist/reorder-trigger.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/reorder-trigger.tsx
var reorder_trigger_exports = {};
__export(reorder_trigger_exports, {
  ReorderTrigger: () => ReorderTrigger
});
module.exports = __toCommonJS(reorder_trigger_exports);
var import_core3 = require("@yamada-ui/core");
var import_icon = require("@yamada-ui/icon");
var import_utils3 = require("@yamada-ui/utils");

// src/reorder.tsx
var import_core = require("@yamada-ui/core");
var import_motion = require("@yamada-ui/motion");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var [ReorderProvider, useReorderContext] = (0, import_utils.createContext)({
  name: "ReorderContext",
  errorMessage: `useReorderContext returned is 'undefined'. Seems you forgot to wrap the components in "<Reorder />"`
});
var omitDuplicated = (values) => Array.from(new Set(values));
var pickDuplicated = (values) => values.filter(
  (value, index, self) => self.indexOf(value) === index && index !== self.lastIndexOf(value)
);
var Reorder = (0, import_react.forwardRef)(
  (props, ref) => {
    const [styles, mergedProps] = (0, import_core.useMultiComponentStyle)("Reorder", props);
    const {
      className,
      orientation = "vertical",
      gap = "fallback(4, 1rem)",
      onChange,
      onCompleteChange,
      children,
      ...rest
    } = (0, import_core.omitThemeProps)(mergedProps);
    const axis = orientation === "vertical" ? "y" : "x";
    const validChildren = (0, import_utils.getValidChildren)(children);
    const defaultValues = (0, import_react.useMemo)(() => {
      const values2 = validChildren.map(({ props: props2 }) => props2.label);
      const duplicatedValues = pickDuplicated(values2);
      if (duplicatedValues.length)
        console.warn(
          `Reorder: 'label' of 'ReorderItem' must not be duplicated. duplicate 'label' is '${duplicatedValues.join(
            `', '`
          )}' `
        );
      return omitDuplicated(values2);
    }, [validChildren]);
    const prevDefaultValues = (0, import_react.useRef)(defaultValues);
    const [values, setValues] = (0, import_react.useState)(defaultValues);
    const prevValues = (0, import_react.useRef)(defaultValues);
    const onReorder = (0, import_react.useCallback)(
      (newValues) => {
        setValues(newValues);
        onChange == null ? void 0 : onChange(newValues);
      },
      [onChange]
    );
    const onCompleteReorder = (0, import_react.useCallback)(() => {
      const isEqual = JSON.stringify(prevValues.current) === JSON.stringify(values);
      if (isEqual)
        return;
      prevValues.current = values;
      onCompleteChange == null ? void 0 : onCompleteChange(values);
    }, [onCompleteChange, values]);
    (0, import_utils.useUpdateEffect)(() => {
      const isEqual = JSON.stringify(defaultValues) === JSON.stringify(prevDefaultValues.current);
      if (isEqual)
        return;
      prevValues.current = defaultValues;
      prevDefaultValues.current = defaultValues;
      setValues(defaultValues);
    }, [defaultValues]);
    const cloneChildren = (0, import_react.useMemo)(
      () => values.map(
        (value) => validChildren.find(({ props: props2 }) => props2.label === value)
      ),
      [values, validChildren]
    );
    const css = {
      display: "flex",
      flexDirection: orientation === "vertical" ? "column" : "row",
      gap,
      ...styles.container
    };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ReorderProvider, { value: { orientation, styles }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core.ui.ul,
      {
        ref,
        as: import_motion.MotionReorder.Group,
        className: (0, import_utils.cx)("ui-reorder", className),
        axis,
        values,
        onReorder,
        __css: css,
        ...rest,
        onMouseUp: (0, import_utils.handlerAll)(rest.onMouseUp, onCompleteReorder),
        onTouchEnd: (0, import_utils.handlerAll)(rest.onTouchEnd, onCompleteReorder),
        children: cloneChildren
      }
    ) });
  }
);
Reorder.displayName = "Reorder";

// src/reorder-item.tsx
var import_core2 = require("@yamada-ui/core");
var import_motion2 = require("@yamada-ui/motion");
var import_utils2 = require("@yamada-ui/utils");
var import_react2 = require("react");
var import_jsx_runtime2 = require("react/jsx-runtime");
var [ReorderItemProvider, useReorderItemContext] = (0, import_utils2.createContext)({
  name: "ReorderContext",
  errorMessage: `useReorderItemContext returned is 'undefined'. Seems you forgot to wrap the components in "<ReorderItem />"`
});
var ReorderItem = (0, import_react2.forwardRef)(
  ({ className, label, ...rest }, ref) => {
    var _a;
    const { orientation, styles } = useReorderContext();
    const dragControls = (0, import_motion2.useDragControls)();
    const [hasTrigger, setHasTrigger] = (0, import_react2.useState)(false);
    const [isDrag, setIsDrag] = (0, import_react2.useState)(false);
    const x = (0, import_motion2.useMotionValue)(0);
    const y = (0, import_motion2.useMotionValue)(0);
    const register = (0, import_react2.useCallback)(
      (node) => setHasTrigger(!!node),
      []
    );
    (0, import_react2.useEffect)(() => {
      const unsubscribeX = x.on("change", (x2) => {
        if (orientation === "horizontal")
          setIsDrag(x2 !== 0);
      });
      const unsubscribeY = y.on("change", (y2) => {
        if (orientation === "vertical")
          setIsDrag(y2 !== 0);
      });
      return () => {
        unsubscribeX();
        unsubscribeY();
      };
    }, [orientation, x, y]);
    const css = {
      ...!hasTrigger ? { cursor: "grab" } : { userSelect: "none" },
      ...styles.item,
      ...!hasTrigger ? {
        _selected: {
          ...(_a = styles.item) == null ? void 0 : _a._selected,
          cursor: "grabbing"
        }
      } : {}
    };
    return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(ReorderItemProvider, { value: { register, isDrag, dragControls }, children: /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
      import_core2.ui.li,
      {
        ref,
        as: import_motion2.MotionReorder.Item,
        className: (0, import_utils2.cx)("ui-reorder__item", className),
        value: label,
        __css: css,
        ...rest,
        dragListener: !hasTrigger,
        dragControls,
        "data-selected": (0, import_utils2.dataAttr)(isDrag),
        style: { ...rest.style, x, y }
      }
    ) });
  }
);
ReorderItem.displayName = "ReorderItem";

// src/reorder-trigger.tsx
var import_jsx_runtime3 = require("react/jsx-runtime");
var ReorderTrigger = (0, import_core3.forwardRef)(
  ({ className, children = /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(ReorderTriggerIcon, {}), ...rest }, ref) => {
    const { styles } = useReorderContext();
    const { register, isDrag, dragControls } = useReorderItemContext();
    const css = {
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      cursor: "grab",
      _selected: {
        cursor: "grabbing"
      },
      ...styles.trigger
    };
    return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      import_core3.ui.div,
      {
        ref: (0, import_utils3.mergeRefs)(register, ref),
        className: (0, import_utils3.cx)("ui-reorder__trigger", className),
        __css: css,
        ...rest,
        "data-selected": (0, import_utils3.dataAttr)(isDrag),
        onPointerDown: (0, import_utils3.handlerAll)(
          rest.onPointerDown,
          (ev) => dragControls.start(ev)
        ),
        children
      }
    );
  }
);
var ReorderTriggerIcon = (rest) => {
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(import_icon.Icon, { viewBox: "0 0 39 39", ...rest, children: [
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 5 0 C 7.761 0 10 2.239 10 5 C 10 7.761 7.761 10 5 10 C 2.239 10 0 7.761 0 5 C 0 2.239 2.239 0 5 0 Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 19 0 C 21.761 0 24 2.239 24 5 C 24 7.761 21.761 10 19 10 C 16.239 10 14 7.761 14 5 C 14 2.239 16.239 0 19 0 Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 33 0 C 35.761 0 38 2.239 38 5 C 38 7.761 35.761 10 33 10 C 30.239 10 28 7.761 28 5 C 28 2.239 30.239 0 33 0 Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 33 14 C 35.761 14 38 16.239 38 19 C 38 21.761 35.761 24 33 24 C 30.239 24 28 21.761 28 19 C 28 16.239 30.239 14 33 14 Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 19 14 C 21.761 14 24 16.239 24 19 C 24 21.761 21.761 24 19 24 C 16.239 24 14 21.761 14 19 C 14 16.239 16.239 14 19 14 Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 5 14 C 7.761 14 10 16.239 10 19 C 10 21.761 7.761 24 5 24 C 2.239 24 0 21.761 0 19 C 0 16.239 2.239 14 5 14 Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 5 28 C 7.761 28 10 30.239 10 33 C 10 35.761 7.761 38 5 38 C 2.239 38 0 35.761 0 33 C 0 30.239 2.239 28 5 28 Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 19 28 C 21.761 28 24 30.239 24 33 C 24 35.761 21.761 38 19 38 C 16.239 38 14 35.761 14 33 C 14 30.239 16.239 28 19 28 Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "path",
      {
        d: "M 33 28 C 35.761 28 38 30.239 38 33 C 38 35.761 35.761 38 33 38 C 30.239 38 28 35.761 28 33 C 28 30.239 30.239 28 33 28 Z",
        fill: "currentColor"
      }
    )
  ] });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ReorderTrigger
});
