// @yamada-ui/motion/dist/index.mjs
"use client"
import {
  Motion
} from "./chunk-3XJJIQVE.mjs";
import "./chunk-R3D52RAW.mjs";
import {
  MOTION_TRANSITION_DEFAULTS,
  MOTION_TRANSITION_EASINGS,
  MOTION_TRANSITION_VARIANTS,
  transitionEnter,
  transitionExit
} from "./chunk-2RLQTJG7.mjs";

// src/index.ts
import {
  m,
  motion,
  AnimatePresence,
  LayoutGroup,
  LazyMotion,
  MotionConfig,
  Reorder,
  useMotionTemplate,
  useScroll,
  useSpring,
  useTime,
  useTransform,
  transform,
  useVelocity,
  useViewportScroll,
  useAnimation,
  useAnimationControls,
  useAnimationFrame,
  useDragControls,
  useInView,
  useIsPresent,
  useReducedMotion,
  useMotionValue,
  useMotionValueEvent,
  useCycle
} from "framer-motion";
export {
  AnimatePresence,
  LayoutGroup,
  LazyMotion,
  MOTION_TRANSITION_DEFAULTS,
  MOTION_TRANSITION_EASINGS,
  MOTION_TRANSITION_VARIANTS,
  Motion,
  MotionConfig,
  Reorder as MotionReorder,
  m,
  motion,
  transform,
  transitionEnter,
  transitionExit,
  useCycle,
  useDragControls,
  useInView,
  useIsPresent,
  useAnimation as useMotionAnimation,
  useAnimationControls as useMotionAnimationControls,
  useAnimationFrame as useMotionAnimationFrame,
  useMotionTemplate,
  useMotionValue,
  useMotionValueEvent,
  useReducedMotion,
  useScroll,
  useSpring,
  useTime,
  useTransform,
  useVelocity,
  useViewportScroll
};
