// @yamada-ui/stepper/dist/chunk-YY2V4TC4.mjs
"use client"
import {
  useStepContext
} from "./chunk-4MXRJFQH.mjs";
import {
  useStepperContext
} from "./chunk-S2LCEE4Q.mjs";

// src/step-status.tsx
import { ui, forwardRef } from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { cx, runIfFunc } from "@yamada-ui/utils";
import { Fragment, jsx } from "react/jsx-runtime";
var StepStatus = forwardRef(
  ({
    className,
    complete = /* @__PURE__ */ jsx(StepIcon, {}),
    incomplete = /* @__PURE__ */ jsx(StepNumber, {}),
    active = /* @__PURE__ */ jsx(StepNumber, {}),
    ...rest
  }, ref) => {
    const { styles } = useStepperContext();
    const { status, ...props } = useStepContext();
    const css = { ...styles.status };
    let component = null;
    switch (status) {
      case "complete":
        component = runIfFunc(complete, props);
        break;
      case "incomplete":
        component = runIfFunc(incomplete, props);
        break;
      case "active":
        component = runIfFunc(active, props);
        break;
    }
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        ref,
        className: cx("ui-step__status", className),
        "data-status": status,
        __css: css,
        ...rest,
        children: component ? /* @__PURE__ */ jsx(Fragment, { children: component }) : null
      }
    );
  }
);
var StepNumber = forwardRef(
  ({ className, children, ...rest }, ref) => {
    const { styles } = useStepperContext();
    const { status, index } = useStepContext();
    const css = { ...styles.number };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        ref,
        className: cx("ui-step__number", className),
        "data-status": status,
        __css: css,
        ...rest,
        children: children || index + 1
      }
    );
  }
);
var StepIcon = forwardRef(
  ({ className, ...rest }, ref) => {
    const { styles } = useStepperContext();
    const { status } = useStepContext();
    const as = status === "complete" ? CheckIcon : void 0;
    const css = { ...styles.icon };
    return /* @__PURE__ */ jsx(
      Icon,
      {
        ref,
        as,
        className: cx("ui-step__icon", className),
        "data-status": status,
        __css: css,
        ...rest
      }
    );
  }
);
var CheckIcon = (props) => {
  return /* @__PURE__ */ jsx(
    "svg",
    {
      stroke: "currentColor",
      fill: "currentColor",
      strokeWidth: "0",
      viewBox: "0 0 20 20",
      "aria-hidden": "true",
      height: "1em",
      width: "1em",
      ...props,
      children: /* @__PURE__ */ jsx(
        "path",
        {
          fillRule: "evenodd",
          d: "M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z",
          clipRule: "evenodd"
        }
      )
    }
  );
};

export {
  StepStatus,
  StepNumber,
  StepIcon
};
