// @yamada-ui/stepper/dist/chunk-S2LCEE4Q.mjs
"use client"

// src/use-stepper.ts
import { createDescendant } from "@yamada-ui/use-descendant";
import { createContext, mergeRefs } from "@yamada-ui/utils";
import { useCallback } from "react";
var {
  DescendantsContextProvider: StepperDescendantsContextProvider,
  useDescendants: useStepperDescendants,
  useDescendant: useStepperDescendant
} = createDescendant();
var [StepperProvider, useStepperContext] = createContext({
  name: "StepperContext",
  errorMessage: `useStepperContext returned is 'undefined'. Seems you forgot to wrap the components in "<Stepper />"`
});
var useStepper = ({
  index,
  orientation = "horizontal",
  showLastSeparator = false,
  ...rest
}) => {
  const descendants = useStepperDescendants();
  const getStepStatus = useCallback(
    (step) => {
      if (step < index)
        return "complete";
      if (step > index)
        return "incomplete";
      return "active";
    },
    [index]
  );
  const getContainerProps = useCallback(
    (props = {}, ref = null) => ({
      ...rest,
      ...props,
      ref,
      "data-orientation": orientation
    }),
    [orientation, rest]
  );
  return {
    descendants,
    index,
    orientation,
    showLastSeparator,
    getStepStatus,
    getContainerProps
  };
};
var useStep = () => {
  var _a;
  const { orientation, getStepStatus } = useStepperContext();
  const { index, register, descendants } = useStepperDescendant();
  const isFirst = index === 0;
  const isLast = index === ((_a = descendants.lastValue()) == null ? void 0 : _a.index);
  const status = getStepStatus(index);
  const getStepProps = useCallback(
    (props = {}, ref = null) => ({
      ...props,
      ref: mergeRefs(ref, register),
      "data-status": status,
      "data-orientation": orientation
    }),
    [orientation, register, status]
  );
  return { index, status, isFirst, isLast, getStepProps };
};

export {
  StepperDescendantsContextProvider,
  useStepperDescendants,
  useStepperDescendant,
  StepperProvider,
  useStepperContext,
  useStepper,
  useStep
};
