// @yamada-ui/stepper/dist/chunk-4MXRJFQH.mjs
"use client"
import {
  useStep,
  useStepperContext
} from "./chunk-S2LCEE4Q.mjs";

// src/step.tsx
import { ui, forwardRef } from "@yamada-ui/core";
import { cx, createContext, dataAttr } from "@yamada-ui/utils";
import { jsx } from "react/jsx-runtime";
var [StepProvider, useStepContext] = createContext({
  name: "StepperContext",
  errorMessage: `useStepContext returned is 'undefined'. Seems you forgot to wrap the components in "<Step />"`
});
var Step = forwardRef(
  ({ className, ...rest }, ref) => {
    const { orientation, showLastSeparator, styles } = useStepperContext();
    const { index, status, isFirst, isLast, getStepProps } = useStep();
    const css = { ...styles.step };
    return /* @__PURE__ */ jsx(StepProvider, { value: { index, status, isFirst, isLast }, children: /* @__PURE__ */ jsx(
      ui.div,
      {
        className: cx("ui-step", className),
        __css: css,
        "data-orientation": orientation,
        "data-stretch": dataAttr(showLastSeparator),
        ...getStepProps(rest, ref)
      }
    ) });
  }
);

export {
  StepProvider,
  useStepContext,
  Step
};
