// @yamada-ui/stepper/dist/step-title.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/step-title.tsx
var step_title_exports = {};
__export(step_title_exports, {
  StepTitle: () => StepTitle
});
module.exports = __toCommonJS(step_title_exports);
var import_core2 = require("@yamada-ui/core");
var import_utils3 = require("@yamada-ui/utils");

// src/step.tsx
var import_core = require("@yamada-ui/core");
var import_utils2 = require("@yamada-ui/utils");

// src/use-stepper.ts
var import_use_descendant = require("@yamada-ui/use-descendant");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var {
  DescendantsContextProvider: StepperDescendantsContextProvider,
  useDescendants: useStepperDescendants,
  useDescendant: useStepperDescendant
} = (0, import_use_descendant.createDescendant)();
var [StepperProvider, useStepperContext] = (0, import_utils.createContext)({
  name: "StepperContext",
  errorMessage: `useStepperContext returned is 'undefined'. Seems you forgot to wrap the components in "<Stepper />"`
});
var useStep = () => {
  var _a;
  const { orientation, getStepStatus } = useStepperContext();
  const { index, register, descendants } = useStepperDescendant();
  const isFirst = index === 0;
  const isLast = index === ((_a = descendants.lastValue()) == null ? void 0 : _a.index);
  const status = getStepStatus(index);
  const getStepProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      ...props,
      ref: (0, import_utils.mergeRefs)(ref, register),
      "data-status": status,
      "data-orientation": orientation
    }),
    [orientation, register, status]
  );
  return { index, status, isFirst, isLast, getStepProps };
};

// src/step.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var [StepProvider, useStepContext] = (0, import_utils2.createContext)({
  name: "StepperContext",
  errorMessage: `useStepContext returned is 'undefined'. Seems you forgot to wrap the components in "<Step />"`
});
var Step = (0, import_core.forwardRef)(
  ({ className, ...rest }, ref) => {
    const { orientation, showLastSeparator, styles } = useStepperContext();
    const { index, status, isFirst, isLast, getStepProps } = useStep();
    const css = { ...styles.step };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(StepProvider, { value: { index, status, isFirst, isLast }, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core.ui.div,
      {
        className: (0, import_utils2.cx)("ui-step", className),
        __css: css,
        "data-orientation": orientation,
        "data-stretch": (0, import_utils2.dataAttr)(showLastSeparator),
        ...getStepProps(rest, ref)
      }
    ) });
  }
);

// src/step-title.tsx
var import_jsx_runtime2 = require("react/jsx-runtime");
var StepTitle = (0, import_core2.forwardRef)(
  ({ className, ...rest }, ref) => {
    const { styles } = useStepperContext();
    const { status } = useStepContext();
    const css = { ...styles.title };
    return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
      import_core2.ui.h3,
      {
        ref,
        className: (0, import_utils3.cx)("ui-step__title", className),
        "data-status": status,
        __css: css,
        ...rest
      }
    );
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StepTitle
});
