// @yamada-ui/theme/dist/components/alert.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/alert.ts
var alert_exports = {};
__export(alert_exports, {
  Alert: () => Alert
});
module.exports = __toCommonJS(alert_exports);
var import_utils = require("@yamada-ui/utils");
var Alert = {
  baseStyle: {
    container: {
      px: 4,
      py: 3,
      rounded: "md"
    },
    icon: {
      flexShrink: 0,
      marginEnd: 3,
      boxSize: 5
    },
    loading: {
      flexShrink: 0,
      marginEnd: 3,
      fontSize: "xl"
    },
    title: {
      marginEnd: 2,
      fontWeight: "bold",
      lineHeight: 5
    },
    description: {
      lineHeight: 5
    }
  },
  variants: {
    basic: ({ theme: t, colorMode: m, colorScheme: c = "primary" }) => ({
      container: { bg: ["white", "black"], borderWidth: "1px" },
      icon: { color: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), `${c}.400`] }
    }),
    subtle: ({ theme: t, colorMode: m, colorScheme: c = "primary" }) => ({
      container: { bg: [`${c}.100`, (0, import_utils.shadeColor)(`${c}.300`, 58)(t, m)] },
      icon: { color: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), `${c}.400`] }
    }),
    solid: ({ theme: t, colorMode: m, colorScheme: c = "primary" }) => ({
      container: {
        bg: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), (0, import_utils.shadeColor)(`${c}.600`, 16)(t, m)],
        color: "white"
      }
    }),
    "island-accent": ({
      theme: t,
      colorMode: m,
      colorScheme: c = "primary"
    }) => ({
      container: {
        bg: ["white", "black"],
        borderWidth: "1px",
        pl: 7,
        _before: {
          content: '""',
          position: "absolute",
          top: "50%",
          left: 3,
          transform: "translateY(-50%)",
          w: 1,
          h: "calc(100% - 1.5rem)",
          bg: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), `${c}.400`],
          rounded: "full"
        }
      },
      icon: { color: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), `${c}.400`] }
    }),
    "left-accent": ({
      theme: t,
      colorMode: m,
      colorScheme: c = "primary"
    }) => ({
      container: {
        bg: [`${c}.100`, (0, import_utils.shadeColor)(`${c}.300`, 58)(t, m)],
        pl: 3,
        borderLeft: "0.25rem solid",
        borderLeftColor: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), `${c}.400`],
        rounded: 4
      },
      icon: { color: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), `${c}.400`] }
    }),
    "top-accent": ({ theme: t, colorMode: m, colorScheme: c = "primary" }) => ({
      container: {
        bg: [`${c}.100`, (0, import_utils.shadeColor)(`${c}.300`, 58)(t, m)],
        pt: 3,
        borderTop: "0.25rem solid",
        borderTopColor: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), `${c}.400`],
        rounded: 4
      },
      icon: { color: [(0, import_utils.tintColor)(`${c}.600`, 16)(t, m), `${c}.400`] }
    })
  },
  defaultProps: {
    variant: "basic",
    colorScheme: "primary"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Alert
});
