// @yamada-ui/theme/dist/components/alpha-slider.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/alpha-slider.ts
var alpha_slider_exports = {};
__export(alpha_slider_exports, {
  AlphaSlider: () => AlphaSlider
});
module.exports = __toCommonJS(alpha_slider_exports);
var import_core = require("@yamada-ui/core");

// src/components/hue-slider.ts
var HueSlider = {
  baseStyle: {
    container: {
      w: "100%",
      rounded: "full",
      cursor: "pointer",
      _readOnly: { cursor: "auto" },
      _disabled: {
        opacity: 0.6,
        cursor: "not-allowed"
      }
    },
    overlay: {
      rounded: "full"
    },
    track: {},
    thumb: {
      rounded: "full",
      outline: 0,
      borderWidth: "3px",
      borderColor: "whiteAlpha.950",
      boxShadow: ["md", "dark-md"],
      transitionProperty: "transform",
      transitionDuration: "normal",
      _active: {
        transform: `scale(1.15)`
      },
      _focusVisible: {
        boxShadow: ["outline", "outline"]
      }
    }
  },
  sizes: {
    sm: {
      container: { h: "3" },
      thumb: { boxSize: "3" }
    },
    md: {
      container: { h: "4" },
      thumb: { boxSize: "4" }
    },
    lg: {
      container: { h: "5" },
      thumb: { boxSize: "5" }
    }
  },
  defaultProps: {
    size: "md"
  }
};

// src/components/alpha-slider.ts
var AlphaSlider = (0, import_core.mergeMultiStyle)(HueSlider)();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AlphaSlider
});
