// @yamada-ui/theme/dist/components/carousel.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/carousel.ts
var carousel_exports = {};
__export(carousel_exports, {
  Carousel: () => Carousel
});
module.exports = __toCommonJS(carousel_exports);
var Carousel = {
  baseStyle: {
    container: {
      w: "100%"
    },
    inner: {},
    slide: {},
    control: {},
    prev: ({ orientation: o }) => ({
      ...o === "vertical" ? { left: "50%", top: "4", transform: "translateX(-50%)" } : { top: "50%", left: "4", transform: "translateY(-50%)" }
    }),
    next: ({ orientation: o }) => ({
      ...o === "vertical" ? { left: "50%", bottom: "4", transform: "translateX(-50%)" } : { top: "50%", right: "4", transform: "translateY(-50%)" }
    }),
    indicators: ({ orientation: o }) => ({
      gap: "md",
      ...o === "vertical" ? {
        py: "4",
        h: "100%",
        top: "50%",
        right: "4",
        transform: "translateY(-50%)"
      } : {
        px: "4",
        w: "100%",
        h: "2",
        left: "50%",
        bottom: "4",
        transform: "translateX(-50%)"
      }
    }),
    indicator: ({ orientation: o }) => ({
      rounded: "full",
      bg: ["whiteAlpha.400", "blackAlpha.400"],
      _hover: {
        bg: ["whiteAlpha.500", "blackAlpha.500"]
      },
      _active: {
        bg: [`whiteAlpha.600`, `blackAlpha.600`]
      },
      _selected: {
        bg: [`whiteAlpha.700`, `blackAlpha.700`]
      },
      transitionProperty: "common",
      transitionDuration: "slower",
      ...o === "vertical" ? { h: "8", right: "4" } : { w: "8", bottom: "4" }
    })
  },
  sizes: {
    sm: {
      inner: {
        h: "sm"
      },
      indicators: ({ orientation: o }) => ({
        ...o === "vertical" ? { w: "1.5" } : { h: "1.5" }
      })
    },
    md: {
      inner: {
        h: "md"
      },
      indicators: ({ orientation: o }) => ({
        ...o === "vertical" ? { w: "1.5" } : { h: "1.5" }
      })
    },
    lg: {
      inner: {
        h: "lg"
      },
      indicators: ({ orientation: o }) => ({
        ...o === "vertical" ? { w: "2" } : { h: "2" }
      })
    },
    xl: {
      inner: {
        h: "xl"
      },
      indicators: ({ orientation: o }) => ({
        ...o === "vertical" ? { w: "2" } : { h: "2" }
      })
    }
  },
  defaultProps: {
    size: "md"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Carousel
});
