// @yamada-ui/theme/dist/components/color-swatch.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/color-swatch.ts
var color_swatch_exports = {};
__export(color_swatch_exports, {
  ColorSwatch: () => ColorSwatch
});
module.exports = __toCommonJS(color_swatch_exports);
var ColorSwatch = {
  baseStyle: {
    container: {},
    overlay: { boxSize: "full" }
  },
  variants: {
    basic: {
      container: { rounded: "md" },
      overlay: { rounded: "md" }
    },
    rounded: {
      container: { rounded: "full" },
      overlay: { rounded: "full" }
    }
  },
  sizes: {
    sm: {
      container: { boxSize: "6" }
    },
    md: {
      container: { boxSize: "8" }
    },
    lg: {
      container: { boxSize: "10" }
    }
  },
  defaultProps: {
    variant: "basic",
    size: "md"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ColorSwatch
});
