// @yamada-ui/theme/dist/components/form-control.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/form-control.ts
var form_control_exports = {};
__export(form_control_exports, {
  FormControl: () => FormControl
});
module.exports = __toCommonJS(form_control_exports);
var FormControl = {
  baseStyle: {
    container: {
      w: "100%",
      position: "relative"
    },
    label: {
      fontSize: "md",
      me: "3",
      mb: "2",
      fontWeight: "medium",
      transitionProperty: "common",
      transitionDuration: "normal",
      opacity: 1,
      _disabled: {
        opacity: 0.4
      }
    },
    requiredIndicator: {
      ms: "1",
      color: ["danger.500", "danger.400"]
    },
    helperMessage: {
      mt: "2",
      color: ["blackAlpha.700", "whiteAlpha.600"],
      fontSize: "sm"
    },
    errorMessage: {
      mt: "2",
      color: ["danger.500", "danger.400"],
      fontSize: "sm"
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FormControl
});
