// @yamada-ui/theme/dist/components/heading.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/heading.ts
var heading_exports = {};
__export(heading_exports, {
  Heading: () => Heading
});
module.exports = __toCommonJS(heading_exports);
var Heading = {
  baseStyle: {
    fontFamily: "heading",
    fontWeight: "bold"
  },
  sizes: {
    "4xl": {
      fontSize: { base: "7xl", sm: "6xl" }
    },
    "3xl": {
      fontSize: { base: "6xl", sm: "5xl" }
    },
    "2xl": {
      fontSize: { base: "5xl", sm: "4xl" }
    },
    xl: {
      fontSize: { base: "4xl", sm: "3xl" }
    },
    lg: {
      fontSize: { base: "3xl", sm: "2xl" }
    },
    md: { fontSize: "xl" },
    sm: { fontSize: "md" },
    xs: { fontSize: "sm" }
  },
  defaultProps: { size: "xl" }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Heading
});
