// @yamada-ui/theme/dist/components/indicator.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/indicator.ts
var indicator_exports = {};
__export(indicator_exports, {
  Indicator: () => Indicator
});
module.exports = __toCommonJS(indicator_exports);
var import_utils = require("@yamada-ui/utils");
var Indicator = {
  baseStyle: ({ withBorder }) => ({
    zIndex: "kurillin",
    rounded: "full",
    w: "fit-content",
    h: "fit-content",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    outline: 0,
    whiteSpace: "nowrap",
    ...withBorder ? { borderWidth: "0.2em", borderColor: ["white", "black"] } : {}
  }),
  sizes: {
    sm: { minH: 2, minW: 2, fontSize: "xs", px: 1, lineHeight: 1.6 },
    md: { minH: 3, minW: 3, fontSize: "sm", px: 1.5, lineHeight: 1.5 },
    lg: { minH: 4, minW: 4, fontSize: "md", px: 2, lineHeight: 1.4 }
  },
  variants: {
    solid: ({ colorScheme: c = "primary" }) => ({
      bg: [`${c}.500`, `${c}.600`],
      color: `white`,
      var: [
        {
          __prefix: "ui",
          name: "ping",
          token: "colors",
          value: [`${c}.300`, `${c}.400`]
        }
      ]
    }),
    subtle: ({ theme: t, colorMode: m, colorScheme: c = "primary" }) => ({
      bg: [
        (0, import_utils.isGray)(c) ? `${c}.50` : `${c}.100`,
        (0, import_utils.shadeColor)(`${c}.300`, 58)(t, m)
      ],
      color: [`${c}.800`, (0, import_utils.isGray)(c) ? `${c}.50` : `${c}.200`],
      var: [
        {
          __prefix: "ui",
          name: "ping",
          token: "colors",
          value: ["blackAlpha.400", "whiteAlpha.500"]
        }
      ]
    })
  },
  defaultProps: {
    size: "md",
    variant: "solid",
    colorScheme: "primary"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Indicator
});
