// @yamada-ui/theme/dist/components/mark.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/mark.ts
var mark_exports = {};
__export(mark_exports, {
  Mark: () => Mark
});
module.exports = __toCommonJS(mark_exports);
var import_core = require("@yamada-ui/core");
var import_utils = require("@yamada-ui/utils");
var Mark = {
  baseStyle: {
    px: "1",
    py: "1"
  },
  variants: {
    solid: ({ theme: t, colorMode: m, colorScheme: c = "gray" }) => ({
      bg: [(0, import_utils.tintColor)(`${c}.600`, 24)(t, m), (0, import_utils.shadeColor)(`${c}.600`, 16)(t, m)],
      color: `white`
    }),
    subtle: ({ theme: t, colorMode: m, colorScheme: c = "gray" }) => ({
      bg: [
        (0, import_utils.isGray)(c) ? `${c}.50` : `${c}.100`,
        (0, import_utils.shadeColor)(`${c}.300`, 58)(t, m)
      ],
      color: [`${c}.800`, (0, import_utils.isGray)(c) ? `${c}.50` : `${c}.200`]
    }),
    outline: ({ theme: t, colorMode: m, colorScheme: c = "gray" }) => {
      const color = (0, import_core.mode)(
        (0, import_utils.getColor)(`${c}.500`)(t, m),
        (0, import_utils.getColor)((0, import_utils.isGray)(c) ? `${c}.100` : `${c}.400`)(t, m)
      )(m);
      return {
        color,
        boxShadow: `inset 0 0 0px 1px ${color}`
      };
    },
    "text-accent": ({ colorScheme: c = "gray" }) => ({
      color: [`${c}.500`, (0, import_utils.isGray)(c) ? `${c}.100` : `${c}.400`],
      p: 0
    }),
    unstyled: {
      p: 0
    }
  },
  defaultProps: {
    variant: "subtle",
    colorScheme: "secondary"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Mark
});
