// @yamada-ui/theme/dist/components/progress.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/progress.ts
var progress_exports = {};
__export(progress_exports, {
  Progress: () => Progress
});
module.exports = __toCommonJS(progress_exports);
var import_core = require("@yamada-ui/core");
var import_utils = require("@yamada-ui/utils");
var Progress = {
  baseStyle: {
    track: {
      bg: "border"
    },
    filledTrack: ({
      theme: t,
      colorMode: m,
      colorScheme: c = "primary",
      hasStripe,
      isAnimation,
      filledTrackColor: f
    }) => {
      hasStripe = !isAnimation && hasStripe;
      const color = (0, import_core.mode)("rgba(255, 255, 255, 0.15)", "rgba(0,0,0,0.1)")(m);
      const bgColor = f ? (0, import_utils.isArray)(f) ? (0, import_core.mode)((0, import_utils.getColor)(f[0], f[0])(t, m), (0, import_utils.getColor)(f[1], f[1])(t, m))(m) : (0, import_utils.getColor)(f, f)(t, m) : (0, import_utils.getColor)(`${c}.500`)(t, m);
      const bgImage = hasStripe ? `linear-gradient(
            45deg,
            ${color} 25%,
            transparent 25%,
            transparent 50%,
            ${color} 50%,
            ${color} 75%,
            transparent 75%,
            transparent
          )` : isAnimation ? `linear-gradient(
            to right,
            transparent 0%,
            ${bgColor} 50%,
            transparent 100%
          )` : void 0;
      return {
        transitionProperty: "common",
        transitionDuration: "slow",
        bgImage,
        ...hasStripe ? { bgSize: "1rem 1rem" } : {},
        ...!isAnimation ? { bgColor } : {}
      };
    }
  },
  sizes: {
    xs: {
      track: {
        h: "1"
      }
    },
    sm: {
      track: {
        h: "2"
      }
    },
    md: {
      track: {
        h: "3"
      }
    },
    lg: {
      track: {
        h: "4"
      }
    }
  },
  defaultProps: {
    size: "md",
    colorScheme: "primary"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Progress
});
