// @yamada-ui/theme/dist/components/radio.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/radio.ts
var radio_exports = {};
__export(radio_exports, {
  Radio: () => Radio
});
module.exports = __toCommonJS(radio_exports);
var Radio = {
  baseStyle: {
    container: {
      _readOnly: { cursor: "auto" },
      _disabled: { cursor: "not-allowed" }
    },
    icon: ({ colorScheme: c = "primary" }) => ({
      transitionProperty: "box-shadow",
      transitionDuration: "normal",
      border: "2px solid",
      borderColor: "inherit",
      rounded: "full",
      color: "white",
      display: "inline-flex",
      justifyContent: "center",
      alignItems: "center",
      _checked: {
        _before: {
          content: `""`,
          display: "inline-block",
          rounded: "full",
          bg: [`${c}.500`, `${c}.600`]
        },
        _hover: {
          _before: {
            bg: [`${c}.600`, `${c}.700`]
          }
        },
        _disabled: {
          _before: {
            bg: ["blackAlpha.400", "whiteAlpha.300"]
          },
          _dark: {
            _before: {
              bg: "whiteAlpha.300"
            }
          }
        }
      },
      _disabled: {
        bg: ["blackAlpha.200", "whiteAlpha.100"]
      },
      _focusVisible: {
        boxShadow: "outline"
      },
      _invalid: {
        borderColor: ["danger.500", "danger.400"],
        _focusVisible: {
          borderColor: ["inherit", "inherit"]
        }
      }
    }),
    label: {
      userSelect: "none",
      _disabled: { opacity: 0.4 }
    }
  },
  sizes: {
    sm: {
      icon: {
        boxSize: "3.5",
        _before: {
          boxSize: "1.5"
        }
      },
      label: { fontSize: "sm" }
    },
    md: {
      icon: {
        boxSize: "4",
        _before: {
          boxSize: "2"
        }
      },
      label: { fontSize: "md" }
    },
    lg: {
      icon: {
        boxSize: "5",
        _before: {
          boxSize: "3"
        }
      },
      label: { fontSize: "lg" }
    }
  },
  defaultProps: {
    size: "md",
    colorScheme: "primary"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Radio
});
