// @yamada-ui/theme/dist/components/rating.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/rating.ts
var rating_exports = {};
__export(rating_exports, {
  Rating: () => Rating
});
module.exports = __toCommonJS(rating_exports);
var Rating = {
  baseStyle: ({ colorScheme: c = "yellow" }) => ({
    container: {
      w: "max-content"
    },
    group: {
      position: "relative"
    },
    item: {
      cursor: "pointer",
      rounded: "sm",
      color: ["blackAlpha.300", "whiteAlpha.300"],
      outline: "none",
      _notLast: {
        position: "absolute",
        top: 0,
        left: 0
      },
      _readOnly: {
        cursor: "default"
      },
      _disabled: {
        opacity: 0.4,
        cursor: "not-allowed"
      },
      _focusVisible: {
        boxShadow: "outline"
      },
      _filled: {
        color: [`${c}.500`, `${c}.600`]
      }
    },
    icon: {
      fill: "currentColor"
    }
  }),
  sizes: {
    xs: {
      icon: {
        fontSize: "md"
      }
    },
    sm: {
      icon: {
        fontSize: "lg"
      }
    },
    md: {
      icon: {
        fontSize: "xl"
      }
    },
    lg: {
      icon: {
        fontSize: "2xl"
      }
    },
    xl: {
      icon: {
        fontSize: "3xl"
      }
    }
  },
  defaultProps: {
    size: "md",
    colorScheme: "yellow"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Rating
});
