// @yamada-ui/theme/dist/components/reorder.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/reorder.ts
var reorder_exports = {};
__export(reorder_exports, {
  Reorder: () => Reorder
});
module.exports = __toCommonJS(reorder_exports);
var Reorder = {
  baseStyle: {
    container: {
      w: "100%"
    },
    item: {
      w: "100%",
      h: "100%",
      _selected: {
        boxShadow: ["md", "dark-md"]
      }
    },
    trigger: {
      color: ["blackAlpha.300", "whiteAlpha.300"]
    }
  },
  variants: {
    elevated: {
      item: {
        bg: ["white", "black"],
        boxShadow: ["base", "dark-md"]
      }
    },
    outline: {
      item: {
        bg: ["white", "black"],
        borderWidth: "1px"
      }
    },
    unstyled: {
      item: { rounded: 0, p: 0, _selected: { boxShadow: "unset" } }
    }
  },
  sizes: {
    sm: {
      item: { rounded: "base", p: "sm" }
    },
    md: {
      item: { rounded: "md", p: "md" }
    },
    normal: {
      item: { rounded: "lg", p: "normal" }
    },
    lg: {
      item: { rounded: "xl", p: "lg" }
    }
  },
  defaultProps: {
    variant: "outline",
    size: "md"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Reorder
});
