// @yamada-ui/theme/dist/components/resizable.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/resizable.ts
var resizable_exports = {};
__export(resizable_exports, {
  Resizable: () => Resizable
});
module.exports = __toCommonJS(resizable_exports);
var import_utils = require("@yamada-ui/utils");
var Resizable = {
  baseStyle: ({ direction: d }) => ({
    container: {},
    item: {},
    trigger: {},
    icon: {
      color: ["blackAlpha.600", "whiteAlpha.700"],
      rotate: d === "vertical" ? "90deg" : "0deg"
    }
  }),
  variants: {
    border: ({ direction: d }) => ({
      trigger: {
        bg: "border",
        ...d === "vertical" ? { h: "px" } : { w: "px" },
        _focus: {
          outline: "none"
        },
        _focusVisible: {
          outline: "none",
          bg: "focus"
        },
        _after: {
          content: "''",
          position: "absolute",
          ...d === "vertical" ? { left: "0", right: "0", transform: "translateY(-50%)" } : { top: "0", bottom: "0", transform: "translateX(-50%)" },
          ...d === "vertical" ? { h: "2" } : { w: "2" }
        }
      },
      icon: {
        p: "1",
        bg: "border",
        rounded: "full"
      }
    }),
    spacer: ({
      direction: d,
      theme: t,
      colorMode: m,
      colorScheme: c = "gray"
    }) => ({
      trigger: {
        ...d === "vertical" ? { p: "1" } : { p: "1" },
        _after: {
          transitionProperty: "common",
          transitionDuration: "slower",
          bg: "transparent",
          content: "''",
          display: "block",
          rounded: "full",
          ...d === "vertical" ? { w: "full", h: "2" } : { w: "2", h: "full" }
        },
        _dark: {
          _after: {
            bg: "transparent"
          }
        },
        _focus: {
          outline: "none"
        },
        _active: {
          _after: {
            bg: (0, import_utils.isGray)(c) ? "border" : `${c}.50`
          },
          _dark: {
            _after: {
              bg: (0, import_utils.isGray)(c) ? "border" : (0, import_utils.transparentizeColor)(`${c}.600`, 0.12)(t, m)
            }
          }
        },
        _focusVisible: {
          outline: "none",
          _after: {
            bg: (0, import_utils.isGray)(c) ? "border" : `${c}.100`
          },
          _dark: {
            _after: {
              bg: (0, import_utils.isGray)(c) ? "border" : (0, import_utils.transparentizeColor)(`${c}.600`, 0.24)(t, m)
            }
          }
        }
      },
      icon: {
        transitionProperty: "common",
        transitionDuration: "slower",
        _active: {
          color: ["transparent", "transparent"]
        }
      }
    }),
    unstyled: ({ direction: d }) => ({
      trigger: {
        _after: {
          content: "''",
          position: "absolute",
          ...d === "vertical" ? { left: "0", right: "0", transform: "translateY(-50%)" } : { top: "0", bottom: "0", transform: "translateX(-50%)" },
          ...d === "vertical" ? { h: "2" } : { w: "2" }
        }
      }
    })
  },
  defaultProps: {
    variant: "border",
    colorScheme: "gray"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Resizable
});
