// @yamada-ui/theme/dist/components/tag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/components/tag.ts
var tag_exports = {};
__export(tag_exports, {
  Tag: () => Tag
});
module.exports = __toCommonJS(tag_exports);
var import_core = require("@yamada-ui/core");
var import_utils = require("@yamada-ui/utils");
var Tag = {
  baseStyle: {
    container: {
      outline: 0,
      fontWeight: "medium",
      rounded: "md",
      _focusVisible: {
        boxShadow: "outline"
      }
    },
    label: {
      overflow: "visible"
    },
    closeButton: {
      transitionProperty: "common",
      transitionDuration: "normal",
      rounded: "full",
      opacity: 0.5,
      _disabled: {
        opacity: 0.4
      },
      _hover: {
        opacity: 0.8,
        _disabled: {
          opacity: 0.4,
          cursor: "not-allowed"
        }
      },
      _active: {
        opacity: 1
      },
      _focusVisible: {
        boxShadow: "outline",
        bg: "rgba(0, 0, 0, 0.14)"
      }
    }
  },
  variants: {
    solid: {
      container: ({ theme: t, colorMode: m, colorScheme: c = "primary" }) => ({
        bg: [(0, import_utils.tintColor)(`${c}.600`, 24)(t, m), (0, import_utils.shadeColor)(`${c}.600`, 16)(t, m)],
        color: `white`
      })
    },
    subtle: {
      container: ({ theme: t, colorMode: m, colorScheme: c = "primary" }) => ({
        bg: [
          (0, import_utils.isGray)(c) ? `${c}.50` : `${c}.100`,
          (0, import_utils.shadeColor)(`${c}.300`, 58)(t, m)
        ],
        color: [`${c}.800`, (0, import_utils.isGray)(c) ? `${c}.50` : `${c}.200`]
      })
    },
    outline: {
      container: ({ theme: t, colorMode: m, colorScheme: c = "primary" }) => {
        const color = (0, import_core.mode)(
          (0, import_utils.getColor)(`${c}.500`)(t, m),
          (0, import_utils.getColor)((0, import_utils.isGray)(c) ? `${c}.100` : `${c}.400`)(t, m)
        )(m);
        return {
          color,
          boxShadow: `inset 0 0 0px 1px ${color}`
        };
      }
    }
  },
  sizes: {
    sm: ({ theme: t }) => ({
      container: {
        minH: "6",
        minW: "6",
        fontSize: "xs",
        px: "2",
        lineHeight: (0, import_utils.getMemoizedObject)(t, "sizes.6")
      }
    }),
    md: ({ theme: t }) => ({
      container: {
        minH: "7",
        minW: "7",
        fontSize: "sm",
        px: "2",
        lineHeight: (0, import_utils.getMemoizedObject)(t, "sizes.7")
      }
    }),
    lg: ({ theme: t }) => ({
      container: {
        minH: "8",
        minW: "8",
        fontSize: "md",
        px: "3",
        lineHeight: (0, import_utils.getMemoizedObject)(t, "sizes.8")
      }
    })
  },
  defaultProps: {
    size: "md",
    variant: "subtle",
    colorScheme: "primary"
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Tag
});
