// @yamada-ui/textarea/dist/chunk-H3EGLDRH.mjs
"use client"

// src/use-autosize.ts
import { pickObject } from "@yamada-ui/utils";
import { useRef } from "react";
var useAutosize = (ref, maxRows, minRows) => {
  const valueRef = useRef();
  const resizeTextarea = () => {
    const el = ref.current;
    if (!el)
      return;
    let { value, placeholder } = el;
    if (value === valueRef.current)
      return;
    else
      valueRef.current = value;
    value != null ? value : value = placeholder != null ? placeholder : "x";
    const nodeSizeData = getSizingData(el);
    if (!nodeSizeData)
      return;
    const rows = calcRows(el, nodeSizeData, value, maxRows, minRows);
    el.rows = rows;
  };
  return resizeTextarea;
};
var SIZING_STYLE = [
  "borderBottomWidth",
  "borderLeftWidth",
  "borderRightWidth",
  "borderTopWidth",
  "boxSizing",
  "fontFamily",
  "fontSize",
  "fontStyle",
  "fontWeight",
  "letterSpacing",
  "lineHeight",
  "paddingBottom",
  "paddingLeft",
  "paddingRight",
  "paddingTop",
  // non-standard
  "tabSize",
  "textIndent",
  // non-standard
  "textRendering",
  "textTransform",
  "width",
  "wordBreak"
];
var getSizingData = (el) => {
  const style = window == null ? void 0 : window.getComputedStyle(el);
  if (style == null)
    return null;
  const sizingStyle = pickObject(
    style,
    SIZING_STYLE
  );
  const { boxSizing } = sizingStyle;
  if (boxSizing === "")
    return null;
  const paddingSize = parseFloat(sizingStyle.paddingBottom) + parseFloat(sizingStyle.paddingTop);
  const borderSize = parseFloat(sizingStyle.borderBottomWidth) + parseFloat(sizingStyle.borderTopWidth);
  const singleRowHeight = parseFloat(sizingStyle.lineHeight);
  return {
    sizingStyle,
    paddingSize,
    borderSize,
    singleRowHeight
  };
};
var calcRows = (el, sizingData, value, maxRows, minRows) => {
  const clone = el.cloneNode();
  Object.assign(clone.style, sizingData.sizingStyle);
  forceHiddenStyles(clone);
  clone.value = value;
  document.body.appendChild(clone);
  let rows;
  if (clone.scrollHeight) {
    const rowHeight = sizingData.singleRowHeight;
    rows = Math.min(
      maxRows,
      Math.max(minRows, Math.floor(clone.scrollHeight / rowHeight))
    );
  } else {
    const lineBreaks = (value.match(/\n/g) || []).length;
    rows = Math.min(maxRows, Math.max(minRows, lineBreaks + 1));
  }
  document.body.removeChild(clone);
  return rows;
};
var HIDDEN_TEXTAREA_STYLE = {
  "min-height": "0",
  "max-height": "none",
  height: "0",
  visibility: "hidden",
  overflow: "hidden",
  position: "absolute",
  "z-index": "-1000",
  top: "0",
  right: "0"
};
var forceHiddenStyles = (el) => {
  Object.keys(HIDDEN_TEXTAREA_STYLE).forEach((key) => {
    el.style.setProperty(
      key,
      HIDDEN_TEXTAREA_STYLE[key],
      "important"
    );
  });
};
var use_autosize_default = useAutosize;

export {
  use_autosize_default
};
