// @yamada-ui/avatar/dist/avatar-group.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/avatar-group.tsx
var avatar_group_exports = {};
__export(avatar_group_exports, {
  AvatarGroup: () => AvatarGroup
});
module.exports = __toCommonJS(avatar_group_exports);
var import_core = require("@yamada-ui/core");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var import_jsx_runtime = require("react/jsx-runtime");
var AvatarGroup = (0, import_core.forwardRef)((props, ref) => {
  const [styles] = (0, import_core.useMultiComponentStyle)("Avatar", props);
  const {
    className,
    max,
    borderColor,
    gap = "fallback(-2, -0.5rem)",
    borderRadius = "fallback(full, 9999px)",
    rounded = "fallback(full, 9999px)",
    children,
    ...rest
  } = (0, import_core.omitThemeProps)(props);
  let validChildren = (0, import_utils.getValidChildren)(children);
  const excess = !(0, import_utils.isUndefined)(max) ? validChildren.length - max : 0;
  validChildren = !(0, import_utils.isUndefined)(max) ? validChildren.slice(0, max) : validChildren;
  validChildren = validChildren.reverse();
  const cloneChildren = validChildren.map(
    (child, i) => {
      var _a;
      return (0, import_react.cloneElement)(
        child,
        (0, import_utils.filterUndefined)({
          marginEnd: !i ? 0 : gap,
          size: props.size,
          borderColor: (_a = child.props.borderColor) != null ? _a : borderColor,
          borderWidth: "2px"
        })
      );
    }
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_core.ui.div,
    {
      ref,
      className: (0, import_utils.cx)("ui-avatar-group", className),
      __css: {
        display: "flex",
        flexDirection: "row-reverse",
        justifyContent: "flex-end",
        alignItems: "center",
        ...styles.group
      },
      ...rest,
      children: [
        excess > 0 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          import_core.ui.span,
          {
            className: "ui-avatar__excess",
            borderRadius,
            rounded,
            ms: gap,
            __css: {
              position: "relative",
              display: "inline-flex",
              justifyContent: "center",
              alignItems: "center",
              flexShrink: 0,
              textAlign: "center",
              textTransform: "uppercase",
              fontWeight: "medium",
              borderWidth: "2px",
              ...styles.excess
            },
            children: `+${excess}`
          }
        ) : null,
        cloneChildren
      ]
    }
  );
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AvatarGroup
});
