// @yamada-ui/avatar/dist/chunk-PVC4MSTZ.mjs
"use client"

// src/avatar-group.tsx
import {
  ui,
  forwardRef,
  useMultiComponentStyle,
  omitThemeProps
} from "@yamada-ui/core";
import {
  cx,
  getValidChildren,
  isUndefined,
  filterUndefined
} from "@yamada-ui/utils";
import { cloneElement } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var AvatarGroup = forwardRef((props, ref) => {
  const [styles] = useMultiComponentStyle("Avatar", props);
  const {
    className,
    max,
    borderColor,
    gap = "fallback(-2, -0.5rem)",
    borderRadius = "fallback(full, 9999px)",
    rounded = "fallback(full, 9999px)",
    children,
    ...rest
  } = omitThemeProps(props);
  let validChildren = getValidChildren(children);
  const excess = !isUndefined(max) ? validChildren.length - max : 0;
  validChildren = !isUndefined(max) ? validChildren.slice(0, max) : validChildren;
  validChildren = validChildren.reverse();
  const cloneChildren = validChildren.map(
    (child, i) => {
      var _a;
      return cloneElement(
        child,
        filterUndefined({
          marginEnd: !i ? 0 : gap,
          size: props.size,
          borderColor: (_a = child.props.borderColor) != null ? _a : borderColor,
          borderWidth: "2px"
        })
      );
    }
  );
  return /* @__PURE__ */ jsxs(
    ui.div,
    {
      ref,
      className: cx("ui-avatar-group", className),
      __css: {
        display: "flex",
        flexDirection: "row-reverse",
        justifyContent: "flex-end",
        alignItems: "center",
        ...styles.group
      },
      ...rest,
      children: [
        excess > 0 ? /* @__PURE__ */ jsx(
          ui.span,
          {
            className: "ui-avatar__excess",
            borderRadius,
            rounded,
            ms: gap,
            __css: {
              position: "relative",
              display: "inline-flex",
              justifyContent: "center",
              alignItems: "center",
              flexShrink: 0,
              textAlign: "center",
              textTransform: "uppercase",
              fontWeight: "medium",
              borderWidth: "2px",
              ...styles.excess
            },
            children: `+${excess}`
          }
        ) : null,
        cloneChildren
      ]
    }
  );
});

export {
  AvatarGroup
};
