// @yamada-ui/avatar/dist/chunk-Y3MK4PTJ.mjs
"use client"
import {
  useAvatarContext
} from "./chunk-F7MHUXUL.mjs";

// src/avatar-badge.tsx
import { ui, forwardRef } from "@yamada-ui/core";
import { useAnimation } from "@yamada-ui/use-animation";
import { cx } from "@yamada-ui/utils";
import { jsx, jsxs } from "react/jsx-runtime";
var placementStyles = {
  "top-start": {
    top: "0",
    insetStart: "0",
    transform: "translate(-25%, -25%)"
  },
  "top-end": {
    top: "0",
    insetEnd: "0",
    transform: "translate(25%, -25%)"
  },
  "bottom-start": {
    bottom: "0",
    insetStart: "0",
    transform: "translate(-25%, 25%)"
  },
  "bottom-end": {
    bottom: "0",
    insetEnd: "0",
    transform: "translate(25%, 25%)"
  }
};
var AvatarBadge = forwardRef(
  ({
    className,
    boxSize = "1em",
    placement = "bottom-end",
    ping,
    pingColor = ["blackAlpha.400", "whiteAlpha.500"],
    pingDuration = "1.4s",
    pingCount = "infinite",
    pingScale = 1.8,
    children,
    ...rest
  }, ref) => {
    const styles = useAvatarContext();
    const animation = useAnimation({
      keyframes: {
        "75%, 100%": {
          transform: `scale(${pingScale})`,
          opacity: 0
        }
      },
      fillMode: "forwards",
      duration: pingDuration,
      timingFunction: "cubic-bezier(0, 0, 0.2, 1)",
      iterationCount: pingCount
    });
    const css = {
      position: "absolute",
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      ...placementStyles[placement],
      ...styles.badge
    };
    return /* @__PURE__ */ jsxs(
      ui.div,
      {
        ref,
        className: cx("ui-avatar__badge", className),
        boxSize,
        __css: css,
        ...rest,
        children: [
          children,
          ping ? /* @__PURE__ */ jsx(
            ui.div,
            {
              className: "ui-avatar__badge__ping",
              __css: {
                position: "absolute",
                boxSize: "100%",
                rounded: "fallback(full, 9999px)",
                opacity: 0.75,
                zIndex: -1,
                bg: pingColor
              },
              animation
            }
          ) : null
        ]
      }
    );
  }
);

export {
  AvatarBadge
};
