// @yamada-ui/theme/dist/styles/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/styles/index.ts
var styles_exports = {};
__export(styles_exports, {
  globalStyle: () => globalStyle,
  layerStyles: () => layerStyles,
  resetStyle: () => resetStyle,
  textStyles: () => textStyles
});
module.exports = __toCommonJS(styles_exports);

// src/styles/global-style.ts
var globalStyle = {
  body: {
    fontFamily: "body",
    bg: ["white", "black"],
    color: ["black", "white"],
    transitionProperty: "background-color",
    transitionDuration: "normal",
    lineHeight: "base",
    overflowX: "hidden"
  },
  "*::placeholder, *[data-placeholder]": {
    color: "blackAlpha.600"
  },
  _dark: {
    "*::placeholder, *[data-placeholder]": {
      color: "whiteAlpha.400"
    }
  },
  "*, *::before, *::after": {
    borderWidth: "0",
    borderStyle: "solid",
    borderColor: "border",
    wordWrap: "break-word"
  }
};

// src/styles/reset-style.ts
var resetStyle = {
  "*": {
    boxSizing: "border-box"
  },
  "::before, ::after": {
    boxSizing: "inherit"
  },
  html: {
    lineHeight: 1.15,
    WebkitTextSizeAdjust: "100%",
    WebkitTapHighlightColor: "transparent"
  },
  body: {
    margin: 0
  },
  main: {
    display: "block"
  },
  "p, table, blockquote, address, pre, iframe, form, figure, dl": {
    margin: 0
  },
  "h1, h2, h3, h4, h5, h6": {
    fontSize: "inherit",
    lineHeight: "inherit",
    fontWeight: "inherit",
    margin: 0
  },
  "ul, ol": {
    margin: 0,
    padding: 0,
    listStyle: "none"
  },
  dt: {
    fontWeight: "bold"
  },
  dd: {
    marginLeft: 0
  },
  hr: {
    boxSizing: "content-box",
    height: 0,
    overflow: "visible",
    border: 0,
    borderTop: "1px solid",
    margin: 0,
    clear: "both",
    color: "inherit"
  },
  pre: {
    fontFamily: "monospace, monospace",
    fontSize: "inherit"
  },
  address: {
    fontStyle: "inherit"
  },
  a: {
    backgroundColor: "transparent",
    textDecoration: "none",
    color: "inherit"
  },
  "abbr[title]": {
    borderBottom: "none",
    textDecoration: "underline dotted"
  },
  "b, strong": {
    fontWeight: "bolder"
  },
  "code, kbd, samp": {
    fontFamily: "monospace, monospace",
    fontSize: "inherit"
  },
  small: {
    fontSize: "80%"
  },
  "sub, sup": {
    fontSize: "75%",
    lineHeight: 0,
    position: "relative",
    verticalAlign: "baseline"
  },
  sub: {
    bottom: "-0.25em"
  },
  sup: {
    top: "-0.5em"
  },
  img: {
    borderStyle: "none",
    verticalAlign: "bottom"
  },
  "embed, object, iframe": {
    border: 0,
    verticalAlign: "bottom"
  },
  "button, input, optgroup, select, textarea": {
    WebkitAppearance: "none",
    appearance: "none",
    verticalAlign: "middle",
    color: "inherit",
    font: "inherit",
    border: 0,
    background: "transparent",
    padding: 0,
    margin: 0,
    outline: 0,
    borderRadius: 0,
    textAlign: "inherit"
  },
  "[type='checkbox']": {
    WebkitAppearance: "checkbox",
    appearance: "checkbox"
  },
  "[type='radio']": {
    WebkitAppearance: "radio",
    appearance: "radio"
  },
  "button, input": {
    overflow: "visible"
  },
  "button, select": {
    textTransform: "none"
  },
  "button, [type='button'], [type='reset'], [type='submit']": {
    cursor: "pointer",
    WebkitAppearance: "none",
    appearance: "none"
  },
  "button[disabled], [type='button'][disabled], [type='reset'][disabled], [type='submit'][disabled]": {
    cursor: "default"
  },
  "button::-moz-focus-inner, [type='button']::-moz-focus-inner, [type='reset']::-moz-focus-inner, [type='submit']::-moz-focus-inner": {
    borderStyle: "none",
    padding: 0
  },
  "button:-moz-focusring, [type='button']:-moz-focusring, [type='reset']:-moz-focusring, [type='submit']:-moz-focusring": {
    outline: "1px dotted ButtonText"
  },
  "select::-ms-expand": {
    display: "none"
  },
  option: {
    padding: 0
  },
  fieldset: {
    margin: 0,
    padding: 0,
    border: 0,
    minWidth: 0
  },
  legend: {
    color: "inherit",
    display: "table",
    maxWidth: "100%",
    padding: 0,
    whiteSpace: "normal"
  },
  progress: {
    verticalAlign: "baseline"
  },
  textarea: {
    overflow: "auto"
  },
  "[type='number']::-webkit-inner-spin-button, [type='number']::-webkit-outer-spin-button": {
    display: "none"
  },
  "[type='search']": {
    outlineOffset: "-2px"
  },
  "[type='search']::-webkit-search-decoration": {
    WebkitAppearance: "none"
  },
  "[type='time']::-webkit-calendar-picker-indicator": {
    display: "none"
  },
  "::-webkit-file-upload-button": {
    WebkitAppearance: "button",
    font: "inherit"
  },
  "::-webkit-search-cancel-button": {
    WebkitAppearance: "none"
  },
  "label[for]": {
    cursor: "pointer"
  },
  details: {
    display: "block"
  },
  summary: {
    display: "list-item"
  },
  "[contenteditable]": {
    outline: "none"
  },
  table: {
    borderCollapse: "collapse",
    borderSpacing: 0
  },
  caption: {
    textAlign: "left"
  },
  "td, th": {
    verticalAlign: "top",
    padding: 0
  },
  th: {
    textAlign: "left",
    fontWeight: "bold"
  },
  template: {
    display: "none"
  },
  "[hidden]": {
    display: "none !important"
  }
};

// src/styles/layer-styles.ts
var layerStyles = {};

// src/styles/text-styles.ts
var textStyles = {};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  globalStyle,
  layerStyles,
  resetStyle,
  textStyles
});
