// @yamada-ui/typography/dist/index.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Heading: () => Heading,
  Text: () => Text
});
module.exports = __toCommonJS(src_exports);

// src/heading.tsx
var import_core = require("@yamada-ui/core");
var import_utils = require("@yamada-ui/utils");
var import_jsx_runtime = require("react/jsx-runtime");
var Heading = (0, import_core.forwardRef)((props, ref) => {
  const [css, mergedProps] = (0, import_core.useComponentStyle)("Heading", props);
  const { className, ...rest } = (0, import_core.omitThemeProps)(mergedProps);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_core.ui.h1,
    {
      ref,
      className: (0, import_utils.cx)("ui-heading", className),
      __css: css,
      ...rest
    }
  );
});

// src/text.tsx
var import_core2 = require("@yamada-ui/core");
var import_utils2 = require("@yamada-ui/utils");
var import_jsx_runtime2 = require("react/jsx-runtime");
var Text = (0, import_core2.forwardRef)((props, ref) => {
  const [css, mergedProps] = (0, import_core2.useComponentStyle)("Text", props);
  const {
    className,
    align: textAlign,
    decoration: textDecoration,
    casing: textTransform,
    ...rest
  } = (0, import_core2.omitThemeProps)(mergedProps);
  const textProps = (0, import_utils2.filterUndefined)({
    textAlign,
    textDecoration,
    textTransform
  });
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
    import_core2.ui.p,
    {
      ref,
      className: (0, import_utils2.cx)("ui-text", className),
      __css: css,
      ...textProps,
      ...rest
    }
  );
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Heading,
  Text
});
