// @yamada-ui/select/dist/select.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/select.tsx
var select_exports = {};
__export(select_exports, {
  Select: () => Select
});
module.exports = __toCommonJS(select_exports);
var import_core6 = require("@yamada-ui/core");
var import_popover2 = require("@yamada-ui/popover");
var import_portal = require("@yamada-ui/portal");
var import_utils6 = require("@yamada-ui/utils");

// src/select-icon.tsx
var import_core2 = require("@yamada-ui/core");
var import_icon = require("@yamada-ui/icon");
var import_use_clickable = require("@yamada-ui/use-clickable");
var import_utils2 = require("@yamada-ui/utils");
var import_react2 = require("react");

// src/use-select.ts
var import_core = require("@yamada-ui/core");
var import_form_control = require("@yamada-ui/form-control");
var import_use_controllable_state = require("@yamada-ui/use-controllable-state");
var import_use_descendant = require("@yamada-ui/use-descendant");
var import_use_disclosure = require("@yamada-ui/use-disclosure");
var import_use_outside_click = require("@yamada-ui/use-outside-click");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var isTargetOption = (target) => {
  var _a;
  return (0, import_utils.isHTMLElement)(target) && !!((_a = target == null ? void 0 : target.getAttribute("role")) == null ? void 0 : _a.startsWith("select-item"));
};
var {
  DescendantsContextProvider: SelectDescendantsContextProvider,
  useDescendantsContext: useSelectDescendantsContext,
  useDescendants: useSelectDescendants,
  useDescendant: useSelectDescendant
} = (0, import_use_descendant.createDescendant)();
var [SelectProvider, useSelectContext] = (0, import_utils.createContext)({
  strict: false,
  name: "SelectContext"
});
var useSelect = ({
  placeholder,
  closeOnBlur = true,
  closeOnEsc = true,
  closeOnSelect = true,
  placeholderInOptions = true,
  omitSelectedValues = false,
  maxSelectValues,
  isEmpty,
  placement = "bottom-start",
  duration = 0.2,
  isOpen: isOpenProp,
  defaultIsOpen,
  onOpen: onOpenProp,
  onClose: onCloseProp,
  optionProps,
  ...rest
}) => {
  rest = (0, import_form_control.useFormControlProps)(rest);
  const { "aria-readonly": _ariaReadonly, ...formControlProps } = (0, import_utils.pickObject)(
    rest,
    import_form_control.formControlProperties
  );
  const [containerProps, fieldProps] = (0, import_utils.splitObject)(
    (0, import_utils.omitObject)(rest, ["value", "defaultValue", "onChange", "aria-readonly"]),
    import_core.layoutStyleProperties
  );
  const descendants = useSelectDescendants();
  const [focusedIndex, setFocusedIndex] = (0, import_react.useState)(-1);
  const [isAllSelected, setIsAllSelected] = (0, import_react.useState)(false);
  const containerRef = (0, import_react.useRef)(null);
  const fieldRef = (0, import_react.useRef)(null);
  const listRef = (0, import_react.useRef)(null);
  const timeoutIds = (0, import_react.useRef)(/* @__PURE__ */ new Set([]));
  const [value, setValue] = (0, import_use_controllable_state.useControllableState)({
    value: rest.value,
    defaultValue: rest.defaultValue,
    onChange: rest.onChange
  });
  const [label, setLabel] = (0, import_react.useState)(void 0);
  const isFocused = focusedIndex > -1;
  const isMulti = (0, import_utils.isArray)(value);
  const isEmptyValue = (!isMulti ? !value : !value.length) && !(placeholder && placeholderInOptions);
  const selectedValues = descendants.values(
    ({ node }) => {
      var _a;
      return isMulti && value.includes((_a = node.dataset.value) != null ? _a : "");
    }
  );
  const selectedIndexes = selectedValues.map(({ index }) => index);
  const enabledValues = descendants.enabledValues(
    ({ index }) => !selectedIndexes.includes(index)
  );
  const onFocusFirst = (0, import_react.useCallback)(() => {
    const id = setTimeout(() => {
      const first = descendants.enabledFirstValue();
      if (!first)
        return;
      if (!isMulti || !omitSelectedValues) {
        setFocusedIndex(first.index);
      } else {
        if (selectedIndexes.includes(first.index)) {
          const enabledFirst = enabledValues[0];
          setFocusedIndex(enabledFirst.index);
        } else {
          setFocusedIndex(first.index);
        }
      }
    });
    timeoutIds.current.add(id);
  }, [descendants, enabledValues, isMulti, omitSelectedValues, selectedIndexes]);
  const onFocusLast = (0, import_react.useCallback)(() => {
    const id = setTimeout(() => {
      const last = descendants.enabledLastValue();
      if (!last)
        return;
      if (!isMulti || !omitSelectedValues) {
        setFocusedIndex(last.index);
      } else {
        if (selectedIndexes.includes(last.index)) {
          const enabledLast = enabledValues.reverse()[0];
          setFocusedIndex(enabledLast.index);
        } else {
          setFocusedIndex(last.index);
        }
      }
    });
    timeoutIds.current.add(id);
  }, [descendants, enabledValues, isMulti, omitSelectedValues, selectedIndexes]);
  const onFocusSelected = (0, import_react.useCallback)(() => {
    const id = setTimeout(() => {
      const values = descendants.enabledValues();
      const selected = values.find(
        ({ node }) => {
          var _a;
          return !isMulti ? node.dataset.value === value : value.includes((_a = node.dataset.value) != null ? _a : "");
        }
      );
      if (selected)
        setFocusedIndex(selected.index);
    });
    timeoutIds.current.add(id);
  }, [descendants, isMulti, value]);
  const onFocusNext = (0, import_react.useCallback)(() => {
    const id = setTimeout(() => {
      var _a;
      const next = descendants.enabledNextValue(focusedIndex);
      if (!next)
        return;
      if (!isMulti || !omitSelectedValues) {
        setFocusedIndex(next.index);
      } else {
        if (selectedIndexes.includes(next.index)) {
          const enabledNext = (_a = enabledValues.find(({ index }) => next.index < index)) != null ? _a : enabledValues[0];
          setFocusedIndex(enabledNext.index);
        } else {
          setFocusedIndex(next.index);
        }
      }
    });
    timeoutIds.current.add(id);
  }, [
    descendants,
    enabledValues,
    focusedIndex,
    isMulti,
    omitSelectedValues,
    selectedIndexes,
    setFocusedIndex
  ]);
  const onFocusPrev = (0, import_react.useCallback)(() => {
    const id = setTimeout(() => {
      var _a;
      const prev = descendants.enabledPrevValue(focusedIndex);
      if (!prev)
        return;
      if (!isMulti || !omitSelectedValues) {
        setFocusedIndex(prev.index);
      } else {
        if (selectedIndexes.includes(prev.index)) {
          const enabledPrev = (_a = enabledValues.reverse().find(({ index }) => index < prev.index)) != null ? _a : enabledValues[0];
          setFocusedIndex(enabledPrev.index);
        } else {
          setFocusedIndex(prev.index);
        }
      }
    });
    timeoutIds.current.add(id);
  }, [
    descendants,
    enabledValues,
    focusedIndex,
    isMulti,
    omitSelectedValues,
    selectedIndexes,
    setFocusedIndex
  ]);
  const onFocusFirstOrSelected = isEmptyValue || omitSelectedValues ? onFocusFirst : onFocusSelected;
  const onFocusLastOrSelected = isEmptyValue || omitSelectedValues ? onFocusLast : onFocusSelected;
  const onChangeLabel = (0, import_react.useCallback)(
    (newValue, runOmit = true) => {
      const values = descendants.values();
      const selectedValues2 = values.filter(({ node }) => node.dataset.value === newValue).map(({ node, index }) => {
        var _a;
        if (!(!!placeholder && placeholderInOptions) || index !== 0) {
          const el = Array.from(node.children).find(
            (child) => child.getAttribute("data-label") !== null
          );
          return (_a = el == null ? void 0 : el.innerHTML) != null ? _a : "";
        } else {
          return void 0;
        }
      });
      setLabel((prev) => {
        if (!isMulti) {
          return selectedValues2[0];
        } else {
          selectedValues2.forEach((selectedValue) => {
            const isSelected = (0, import_utils.isArray)(prev) && prev.includes(selectedValue != null ? selectedValue : "");
            if (!isSelected) {
              prev = [...(0, import_utils.isArray)(prev) ? prev : [], selectedValue];
            } else if (runOmit) {
              prev = (0, import_utils.isArray)(prev) ? prev.filter((value2) => value2 !== selectedValue) : void 0;
            }
          });
          return prev;
        }
      });
    },
    [descendants, isMulti, placeholder, placeholderInOptions]
  );
  const onChange = (0, import_react.useCallback)(
    (newValue) => {
      setValue((prev) => {
        if (!(0, import_utils.isArray)(prev)) {
          return newValue;
        } else {
          const isSelected = prev.includes(newValue);
          if (!isSelected) {
            return [...prev, newValue];
          } else {
            return prev.filter((value2) => value2 !== newValue);
          }
        }
      });
      onChangeLabel(newValue);
    },
    [onChangeLabel, setValue]
  );
  const onClear = (0, import_react.useCallback)(
    (ev) => {
      ev.stopPropagation();
      setValue([]);
      setLabel(void 0);
    },
    [setLabel, setValue]
  );
  const {
    isOpen,
    onOpen: onInternalOpen,
    onClose
  } = (0, import_use_disclosure.useDisclosure)({
    isOpen: isOpenProp,
    defaultIsOpen,
    onOpen: onOpenProp,
    onClose: onCloseProp
  });
  const onOpen = (0, import_react.useCallback)(() => {
    if (formControlProps.disabled || formControlProps.readOnly)
      return;
    if (isEmpty || isAllSelected)
      return;
    onInternalOpen();
  }, [formControlProps, isEmpty, isAllSelected, onInternalOpen]);
  const onSelect = (0, import_react.useCallback)(() => {
    var _a, _b;
    let enabledValue = descendants.value(focusedIndex);
    if ("disabled" in ((_a = enabledValue == null ? void 0 : enabledValue.node.dataset) != null ? _a : {}))
      enabledValue = void 0;
    if (!enabledValue)
      return;
    const value2 = (_b = enabledValue.node.dataset.value) != null ? _b : "";
    onChange(value2);
    if (closeOnSelect)
      onClose();
    if (omitSelectedValues)
      onFocusNext();
  }, [
    closeOnSelect,
    descendants,
    focusedIndex,
    omitSelectedValues,
    onChange,
    onClose,
    onFocusNext
  ]);
  const onClick = (0, import_react.useCallback)(() => {
    if (isOpen)
      return;
    onOpen();
    onFocusFirstOrSelected();
  }, [isOpen, onFocusFirstOrSelected, onOpen]);
  const onFocus = (0, import_react.useCallback)(() => {
    if (isOpen)
      return;
    onOpen();
    onFocusFirstOrSelected();
  }, [isOpen, onFocusFirstOrSelected, onOpen]);
  const onBlur = (0, import_react.useCallback)(
    (ev) => {
      const relatedTarget = (0, import_utils.getEventRelatedTarget)(ev);
      if ((0, import_utils.isContains)(containerRef.current, relatedTarget))
        return;
      if (!closeOnBlur)
        return;
      if (isOpen)
        onClose();
    },
    [closeOnBlur, isOpen, onClose]
  );
  const onKeyDown = (0, import_react.useCallback)(
    (ev) => {
      if (ev.key === " ")
        ev.key = ev.code;
      if (formControlProps.disabled || formControlProps.readOnly)
        return;
      const actions = {
        ArrowDown: isFocused ? () => onFocusNext() : !isOpen ? (0, import_utils.funcAll)(onOpen, onFocusFirstOrSelected) : void 0,
        ArrowUp: isFocused ? () => onFocusPrev() : !isOpen ? (0, import_utils.funcAll)(onOpen, onFocusLastOrSelected) : void 0,
        Space: isFocused ? onSelect : !isOpen ? (0, import_utils.funcAll)(onOpen, onFocusFirstOrSelected) : void 0,
        Enter: isFocused ? onSelect : !isOpen ? (0, import_utils.funcAll)(onOpen, onFocusFirstOrSelected) : void 0,
        Home: isOpen ? onFocusFirst : void 0,
        End: isOpen ? onFocusLast : void 0,
        Escape: closeOnEsc ? onClose : void 0
      };
      const action = actions[ev.key];
      if (!action)
        return;
      ev.preventDefault();
      ev.stopPropagation();
      action();
    },
    [
      formControlProps.disabled,
      formControlProps.readOnly,
      isFocused,
      isOpen,
      onOpen,
      onFocusFirstOrSelected,
      onFocusLastOrSelected,
      onSelect,
      onFocusFirst,
      onFocusLast,
      closeOnEsc,
      onClose,
      onFocusNext,
      onFocusPrev
    ]
  );
  (0, import_use_outside_click.useOutsideClick)({
    ref: containerRef,
    handler: onClose,
    enabled: isOpen && closeOnBlur
  });
  (0, import_react.useEffect)(() => {
    if (!isMulti)
      return;
    if (!omitSelectedValues && (0, import_utils.isUndefined)(maxSelectValues))
      return;
    const isAll = value.length > 0 && value.length === descendants.count();
    const isMax = value.length === maxSelectValues;
    if (isAll || isMax) {
      onClose();
      setIsAllSelected(true);
    } else {
      setIsAllSelected(false);
    }
  }, [
    omitSelectedValues,
    value,
    descendants,
    isMulti,
    onClose,
    maxSelectValues
  ]);
  (0, import_utils.useUpdateEffect)(() => {
    if (!isOpen)
      setFocusedIndex(-1);
  }, [isOpen]);
  (0, import_utils.useUnmountEffect)(() => {
    timeoutIds.current.forEach((id) => clearTimeout(id));
    timeoutIds.current.clear();
  });
  const getPopoverProps = (0, import_react.useCallback)(
    (props) => ({
      matchWidth: true,
      ...rest,
      ...props,
      isOpen,
      onOpen,
      onClose,
      placement,
      duration,
      trigger: "never",
      closeOnButton: false,
      closeOnBlur
    }),
    [duration, closeOnBlur, onClose, onOpen, placement, rest, isOpen]
  );
  const getContainerProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      ref: (0, import_utils.mergeRefs)(containerRef, ref),
      ...containerProps,
      ...props,
      ...formControlProps,
      onClick: (0, import_utils.handlerAll)(props.onClick, rest.onClick, onClick),
      onBlur: (0, import_utils.handlerAll)(props.onBlur, rest.onBlur, onBlur)
    }),
    [containerProps, formControlProps, onBlur, onClick, rest]
  );
  const getFieldProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      ref: (0, import_utils.mergeRefs)(fieldRef, ref),
      tabIndex: 0,
      ...fieldProps,
      ...props,
      "data-active": (0, import_utils.dataAttr)(isOpen),
      "data-placeholder": (0, import_utils.dataAttr)(
        !isMulti ? label === void 0 : !(label == null ? void 0 : label.length)
      ),
      "aria-expanded": (0, import_utils.dataAttr)(isOpen),
      onFocus: (0, import_utils.handlerAll)(props.onFocus, rest.onFocus, onFocus),
      onKeyDown: (0, import_utils.handlerAll)(props.onKeyDown, rest.onKeyDown, onKeyDown)
    }),
    [fieldProps, isOpen, isMulti, label, rest, onFocus, onKeyDown]
  );
  return {
    descendants,
    value,
    label,
    focusedIndex,
    placeholder,
    placeholderInOptions,
    omitSelectedValues,
    closeOnSelect,
    isOpen,
    containerRef,
    fieldRef,
    listRef,
    optionProps,
    formControlProps,
    onChangeLabel,
    onChange,
    onClear,
    onOpen,
    onClose,
    onFocusFirst,
    onFocusLast,
    onFocusSelected,
    onFocusNext,
    onFocusPrev,
    setFocusedIndex,
    getPopoverProps,
    getContainerProps,
    getFieldProps
  };
};
var useSelectList = () => {
  const { listRef, focusedIndex } = useSelectContext();
  const descendants = useSelectDescendantsContext();
  const beforeFocusedIndex = (0, import_react.useRef)(-1);
  const selectedValue = descendants.value(focusedIndex);
  (0, import_react.useEffect)(() => {
    if (!listRef.current || !selectedValue)
      return;
    if (beforeFocusedIndex.current === selectedValue.index)
      return;
    const parent = listRef.current;
    const child = selectedValue.node;
    const parentHeight = parent.clientHeight;
    const viewTop = parent.scrollTop;
    const viewBottom = viewTop + parentHeight;
    const childHeight = child.clientHeight;
    const childTop = child.offsetTop;
    const childBottom = childTop + childHeight;
    const isInView = viewTop <= childTop && childBottom <= viewBottom;
    const isScrollBottom = beforeFocusedIndex.current < selectedValue.index;
    if (!isInView) {
      if (childBottom <= parentHeight) {
        listRef.current.scrollTo({ top: 0 });
      } else {
        if (!isScrollBottom) {
          listRef.current.scrollTo({ top: childTop + 1 });
        } else {
          listRef.current.scrollTo({ top: childBottom - parentHeight });
        }
      }
    }
    beforeFocusedIndex.current = selectedValue.index;
  }, [listRef, selectedValue]);
  const getListProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      as: "ul",
      ref: (0, import_utils.mergeRefs)(listRef, ref),
      role: "select",
      tabIndex: -1,
      ...props
    }),
    [listRef]
  );
  return {
    getListProps
  };
};
var useSelectOptionGroup = ({
  label,
  ...rest
}) => {
  const { value, omitSelectedValues } = useSelectContext();
  const isMulti = (0, import_utils.isArray)(value);
  const descendants = useSelectDescendantsContext();
  const values = descendants.values();
  const selectedValues = isMulti && omitSelectedValues ? descendants.values(
    ({ node }) => {
      var _a;
      return value.includes((_a = node.dataset.value) != null ? _a : "");
    }
  ) : [];
  const selectedIndexes = selectedValues.map(({ index }) => index);
  const childValues = values.filter(
    ({ node, index }) => {
      var _a;
      return ((_a = node.parentElement) == null ? void 0 : _a.dataset.label) === label && !selectedIndexes.includes(index);
    }
  );
  const isEmpty = !childValues.length;
  const computedRest = (0, import_utils.splitObject)(rest, import_core.layoutStyleProperties);
  const getContainerProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => {
      const style = {
        border: "0px",
        clip: "rect(0px, 0px, 0px, 0px)",
        height: "1px",
        width: "1px",
        margin: "-1px",
        padding: "0px",
        overflow: "hidden",
        whiteSpace: "nowrap",
        position: "absolute"
      };
      return {
        ref,
        ...props,
        ...computedRest[0],
        style: isEmpty ? style : void 0
      };
    },
    [computedRest, isEmpty]
  );
  const getGroupProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      ref,
      ...props,
      ...computedRest[1],
      "data-label": label
    }),
    [computedRest, label]
  );
  return {
    label,
    getContainerProps,
    getGroupProps
  };
};
var useSelectOption = (ref, props) => {
  var _a, _b;
  const {
    fieldRef,
    value,
    placeholder,
    placeholderInOptions,
    omitSelectedValues,
    closeOnSelect: generalCloseOnSelect,
    focusedIndex,
    optionProps,
    onChange,
    onChangeLabel,
    onFocusNext,
    onClose,
    setFocusedIndex
  } = useSelectContext();
  let {
    icon: customIcon,
    isDisabled,
    isFocusable,
    closeOnSelect: customCloseOnSelect,
    children,
    ...computedProps
  } = { ...optionProps, ...props };
  const trulyDisabled = !!isDisabled && !isFocusable;
  const itemRef = (0, import_react.useRef)(null);
  const { index, register, descendants } = useSelectDescendant({
    disabled: trulyDisabled
  });
  const values = descendants.values();
  const frontValues = values.slice(0, index);
  const isMulti = (0, import_utils.isArray)(value);
  const isDuplicated = !isMulti ? frontValues.some(
    ({ node }) => {
      var _a2;
      return node.dataset.value === ((_a2 = computedProps.value) != null ? _a2 : "");
    }
  ) : false;
  const isSelected = !isDuplicated && (!isMulti ? ((_a = computedProps.value) != null ? _a : "") === value : value.includes((_b = computedProps.value) != null ? _b : ""));
  const isFocused = index === focusedIndex;
  if (!!placeholder && index > 0 && placeholderInOptions && !computedProps.value) {
    console.warn(
      `${!isMulti ? "Select" : "MultiSelect"}: If placeholders are present, All options must be set value. If want to set an empty value, either don't set the placeholder or set 'placeholderInOptions' to false.`
    );
  }
  const onClick = (0, import_react.useCallback)(
    (ev) => {
      var _a2;
      ev.preventDefault();
      ev.stopPropagation();
      if (isDisabled) {
        if (fieldRef.current)
          fieldRef.current.focus();
        return;
      }
      if (!isTargetOption(ev.currentTarget)) {
        if (fieldRef.current)
          fieldRef.current.focus();
        return;
      }
      setFocusedIndex(index);
      onChange((_a2 = computedProps.value) != null ? _a2 : "");
      if (fieldRef.current)
        fieldRef.current.focus();
      if (customCloseOnSelect != null ? customCloseOnSelect : generalCloseOnSelect)
        onClose();
      if (omitSelectedValues)
        onFocusNext();
    },
    [
      isDisabled,
      setFocusedIndex,
      index,
      onChange,
      computedProps.value,
      fieldRef,
      customCloseOnSelect,
      generalCloseOnSelect,
      onClose,
      omitSelectedValues,
      onFocusNext
    ]
  );
  (0, import_react.useEffect)(() => {
    var _a2;
    if (isSelected)
      onChangeLabel((_a2 = computedProps.value) != null ? _a2 : "", false);
  }, [computedProps, isSelected, onChangeLabel]);
  const getOptionProps = (0, import_react.useCallback)(
    (props2 = {}) => {
      var _a2;
      const style = {
        border: "0px",
        clip: "rect(0px, 0px, 0px, 0px)",
        height: "1px",
        width: "1px",
        margin: "-1px",
        padding: "0px",
        overflow: "hidden",
        whiteSpace: "nowrap",
        position: "absolute"
      };
      return {
        ref: (0, import_utils.mergeRefs)(itemRef, ref, register),
        ...(0, import_utils.omitObject)(computedProps, ["value"]),
        ...props2,
        role: "select-item",
        tabIndex: -1,
        style: omitSelectedValues && isSelected ? style : void 0,
        "data-value": (_a2 = computedProps.value) != null ? _a2 : "",
        "data-focus": (0, import_utils.dataAttr)(isFocused),
        "data-disabled": (0, import_utils.dataAttr)(isDisabled),
        "aria-checked": (0, import_utils.ariaAttr)(isSelected),
        "aria-disabled": (0, import_utils.ariaAttr)(isDisabled),
        onClick: (0, import_utils.handlerAll)(computedProps.onClick, props2.onClick, onClick)
      };
    },
    [
      computedProps,
      isDisabled,
      isFocused,
      isSelected,
      omitSelectedValues,
      onClick,
      ref,
      register
    ]
  );
  return {
    isSelected,
    isFocused,
    customIcon,
    children,
    getOptionProps
  };
};

// src/select-icon.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var SelectIcon = (0, import_core2.forwardRef)(
  ({ className, children, __css, ...rest }, ref) => {
    const { styles } = useSelectContext();
    const css = {
      position: "absolute",
      top: "50%",
      transform: "translateY(-50%)",
      display: "inline-flex",
      alignItems: "center",
      justifyContent: "center",
      pointerEvents: "none",
      cursor: "pointer",
      ...styles.icon,
      ...__css
    };
    const validChildren = (0, import_utils2.getValidChildren)(children);
    const cloneChildren = validChildren.map(
      (child) => (0, import_react2.cloneElement)(child, {
        focusable: false,
        "aria-hidden": true,
        style: {
          maxWidth: "1em",
          maxHeight: "1em",
          color: "currentColor"
        }
      })
    );
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core2.ui.div,
      {
        ref,
        className: (0, import_utils2.cx)("ui-select__icon", className),
        __css: css,
        ...rest,
        children: (0, import_utils2.isValidElement)(children) ? cloneChildren : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icon.ChevronIcon, {})
      }
    );
  }
);

// src/select-list.tsx
var import_core3 = require("@yamada-ui/core");
var import_popover = require("@yamada-ui/popover");
var import_utils3 = require("@yamada-ui/utils");
var import_jsx_runtime2 = require("react/jsx-runtime");
var SelectList = (0, import_core3.forwardRef)(
  ({ className, w, width, minW, minWidth, ...rest }, ref) => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const { styles } = useSelectContext();
    const { getListProps } = useSelectList();
    w = (_d = w != null ? w : width) != null ? _d : (_c = (_a = styles.list) == null ? void 0 : _a.w) != null ? _c : (_b = styles.list) == null ? void 0 : _b.width;
    minW = (_h = minW != null ? minW : minWidth) != null ? _h : (_g = (_e = styles.list) == null ? void 0 : _e.minW) != null ? _g : (_f = styles.list) == null ? void 0 : _f.minWidth;
    const css = { ...styles.list };
    return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
      import_popover.PopoverContent,
      {
        className: (0, import_utils3.cx)("ui-select__list", className),
        w,
        minW,
        __css: css,
        ...getListProps(rest, ref)
      }
    );
  }
);

// src/option-group.tsx
var import_core4 = require("@yamada-ui/core");
var import_utils4 = require("@yamada-ui/utils");
var import_jsx_runtime3 = require("react/jsx-runtime");
var OptionGroup = (0, import_core4.forwardRef)(
  ({ className, color, h, height, minH, minHeight, children, ...rest }, ref) => {
    const { styles } = useSelectContext();
    const { label, getContainerProps, getGroupProps } = useSelectOptionGroup(rest);
    h != null ? h : h = height;
    minH != null ? minH : minH = minHeight;
    return /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(
      import_core4.ui.li,
      {
        className: (0, import_utils4.cx)("ui-select__item", "ui-select__item--group", className),
        __css: { w: "100%", h: "fit-content", color },
        ...getContainerProps(),
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
            import_core4.ui.span,
            {
              className: "ui-select__item__group-label",
              __css: styles.groupLabel,
              lineClamp: 1,
              children: label
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
            import_core4.ui.ul,
            {
              ...getGroupProps({}, ref),
              className: "ui-select__item__group",
              __css: { h, minH, ...styles.group },
              children
            }
          )
        ]
      }
    );
  }
);

// src/option.tsx
var import_core5 = require("@yamada-ui/core");
var import_utils5 = require("@yamada-ui/utils");
var import_jsx_runtime4 = require("react/jsx-runtime");
var Option = (0, import_core5.forwardRef)(
  ({ className, icon, ...rest }, ref) => {
    const { styles } = useSelectContext();
    const { isSelected, customIcon, children, getOptionProps } = useSelectOption(ref, rest);
    icon != null ? icon : icon = customIcon;
    const css = {
      textDecoration: "none",
      color: "inherit",
      userSelect: "none",
      display: "flex",
      width: "100%",
      alignItems: "center",
      textAlign: "start",
      flex: "0 0 auto",
      outline: 0,
      gap: "0.75rem",
      ...styles.item
    };
    return /* @__PURE__ */ (0, import_jsx_runtime4.jsxs)(
      import_core5.ui.li,
      {
        className: (0, import_utils5.cx)("ui-select__item", className),
        __css: css,
        ...getOptionProps(),
        children: [
          icon !== null ? /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(OptionIcon, { opacity: isSelected ? 1 : 0, children: icon || /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(CheckIcon, {}) }) : null,
          /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(import_core5.ui.span, { style: { flex: 1 }, "data-label": true, children })
        ]
      }
    );
  }
);
var OptionIcon = (0, import_core5.forwardRef)(
  ({ className, ...rest }, ref) => {
    const { styles } = useSelectContext();
    const css = {
      flexShrink: 0,
      display: "inline-flex",
      justifyContent: "center",
      alignItems: "center",
      fontSize: "0.85em",
      ...styles.itemIcon
    };
    return /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
      import_core5.ui.span,
      {
        ref,
        className: (0, import_utils5.cx)("ui-select__item__icon", className),
        __css: css,
        ...rest
      }
    );
  }
);
var CheckIcon = () => /* @__PURE__ */ (0, import_jsx_runtime4.jsx)("svg", { viewBox: "0 0 14 14", width: "1em", height: "1em", children: /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
  "polygon",
  {
    fill: "currentColor",
    points: "5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
  }
) });

// src/select.tsx
var import_jsx_runtime5 = require("react/jsx-runtime");
var Select = (0, import_core6.forwardRef)((props, ref) => {
  const [styles, mergedProps] = (0, import_core6.useMultiComponentStyle)("Select", props);
  let {
    className,
    placeholder,
    defaultValue = "",
    placeholderInOptions = true,
    items = [],
    color,
    h,
    height,
    minH,
    minHeight,
    containerProps,
    listProps,
    iconProps,
    portalProps = { isDisabled: true },
    children,
    ...computedProps
  } = (0, import_core6.omitThemeProps)(mergedProps);
  const validChildren = (0, import_utils6.getValidChildren)(children);
  let computedChildren = [];
  if (!validChildren.length && items.length) {
    computedChildren = items.map((item, i) => {
      if ("value" in item) {
        const { label, value, ...props2 } = item;
        return /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(Option, { value, ...props2, children: label }, i);
      } else if ("items" in item) {
        const { label, items: items2 = [], ...props2 } = item;
        return /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
          OptionGroup,
          {
            label: label != null ? label : "",
            ...props2,
            children: items2.map(({ label: label2, value, ...props3 }, i2) => /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(Option, { value, ...props3, children: label2 }, i2))
          },
          i
        );
      }
    }).filter(Boolean);
  }
  const isEmpty = !validChildren.length && !computedChildren.length && !(!!placeholder && placeholderInOptions);
  const {
    descendants,
    formControlProps,
    getPopoverProps,
    getContainerProps,
    getFieldProps,
    ...rest
  } = useSelect({
    ...computedProps,
    placeholder,
    placeholderInOptions,
    defaultValue,
    isEmpty
  });
  h != null ? h : h = height;
  minH != null ? minH : minH = minHeight;
  const css = {
    w: "100%",
    h: "fit-content",
    color,
    ...styles.container
  };
  return /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(SelectDescendantsContextProvider, { value: descendants, children: /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
    SelectProvider,
    {
      value: { ...rest, placeholder, placeholderInOptions, styles },
      children: /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(import_popover2.Popover, { ...getPopoverProps(), children: /* @__PURE__ */ (0, import_jsx_runtime5.jsxs)(
        import_core6.ui.div,
        {
          className: (0, import_utils6.cx)("ui-select", className),
          __css: css,
          ...getContainerProps(containerProps),
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime5.jsxs)(
              import_core6.ui.div,
              {
                className: "ui-select__inner",
                __css: { position: "relative", ...styles.inner },
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(import_popover2.PopoverTrigger, { children: /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(SelectField, { h, minH, ...getFieldProps({}, ref) }) }),
                  /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(SelectIcon, { ...iconProps, ...formControlProps })
                ]
              }
            ),
            !isEmpty ? /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(import_portal.Portal, { ...portalProps, children: /* @__PURE__ */ (0, import_jsx_runtime5.jsxs)(SelectList, { ...listProps, children: [
              !!placeholder && placeholderInOptions ? /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(Option, { children: placeholder }) : null,
              children != null ? children : computedChildren
            ] }) }) : null
          ]
        }
      ) })
    }
  ) });
});
var SelectField = (0, import_core6.forwardRef)(
  ({ className, isTruncated = true, lineClamp, h, minH, ...rest }, ref) => {
    var _a;
    const { label, placeholder, styles } = useSelectContext();
    const css = {
      pe: "2rem",
      h,
      minH,
      display: "flex",
      alignItems: "center",
      ...styles.field
    };
    return /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
      import_core6.ui.div,
      {
        ref,
        className: (0, import_utils6.cx)("ui-select__field", className),
        __css: css,
        ...rest,
        children: /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
          import_core6.ui.span,
          {
            isTruncated,
            lineClamp,
            dangerouslySetInnerHTML: {
              __html: (_a = label != null ? label : placeholder) != null ? _a : ""
            }
          }
        )
      }
    );
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Select
});
