// @yamada-ui/slider/dist/chunk-R5DZPYSR.mjs
"use client"

// src/slider.tsx
import {
  ui,
  forwardRef,
  useMultiComponentStyle,
  omitThemeProps
} from "@yamada-ui/core";
import {
  formControlProperties,
  useFormControlProps
} from "@yamada-ui/form-control";
import { useControllableState } from "@yamada-ui/use-controllable-state";
import { useLatestRef } from "@yamada-ui/use-latest-ref";
import { usePanEvent } from "@yamada-ui/use-pan-event";
import { useSize } from "@yamada-ui/use-size";
import {
  createContext,
  cx,
  pickObject,
  useCallbackRef,
  valueToPercent,
  clampNumber,
  roundNumberToStep,
  useUpdateEffect,
  mergeRefs,
  dataAttr,
  handlerAll,
  percentToValue,
  getValidChildren,
  findChildren,
  isEmpty,
  omitChildren,
  includesChildren
} from "@yamada-ui/utils";
import { useCallback, useRef, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var useSlider = ({
  focusThumbOnChange = true,
  ...props
}) => {
  if (!focusThumbOnChange)
    props.isReadOnly = true;
  let {
    id,
    name,
    min = 0,
    max = 100,
    step = 1,
    defaultValue,
    orientation = "horizontal",
    isReversed,
    required,
    disabled,
    readOnly,
    value: valueProp,
    onChange,
    onChangeStart: onChangeStartProp,
    onChangeEnd: onChangeEndProp,
    onFocus,
    onBlur,
    "aria-readonly": ariaReadonly,
    ...rest
  } = useFormControlProps(props);
  const formControlProps = pickObject(rest, formControlProperties);
  if (max < min)
    throw new Error("Do not assign a number less than 'min' to 'max'");
  const onChangeStart = useCallbackRef(onChangeStartProp);
  const onChangeEnd = useCallbackRef(onChangeEndProp);
  const [computedValue, setValue] = useControllableState({
    value: valueProp,
    defaultValue: defaultValue != null ? defaultValue : min + (max - min) / 2,
    onChange
  });
  const [isDragging, setDragging] = useState(false);
  const [isFocused, setFocused] = useState(false);
  const isInteractive = !(disabled || readOnly);
  const tenStep = (max - min) / 10;
  const oneStep = step || (max - min) / 100;
  const value = clampNumber(computedValue, min, max);
  const reversedValue = max - value + min;
  const thumbValue = isReversed ? reversedValue : value;
  const thumbPercent = valueToPercent(thumbValue, min, max);
  const isVertical = orientation === "vertical";
  const latestRef = useLatestRef({
    min,
    max,
    step,
    value,
    isInteractive,
    focusThumbOnChange
  });
  const eventSourceRef = useRef(null);
  const containerRef = useRef(null);
  const trackRef = useRef(null);
  const thumbRef = useRef(null);
  const thumbSize = useSize(thumbRef);
  usePanEvent(containerRef, {
    onSessionStart: (ev) => {
      const { isInteractive: isInteractive2, value: value2 } = latestRef.current;
      if (!isInteractive2)
        return;
      setDragging(true);
      focusThumb();
      setValueFromPointer(ev);
      onChangeStart(value2);
    },
    onSessionEnd: () => {
      const { isInteractive: isInteractive2, value: value2 } = latestRef.current;
      if (!isInteractive2)
        return;
      setDragging(false);
      onChangeEnd(value2);
    },
    onMove: (ev) => {
      const { isInteractive: isInteractive2 } = latestRef.current;
      if (!isInteractive2)
        return;
      setValueFromPointer(ev);
    }
  });
  const getValueFromPointer = useCallback(
    (ev) => {
      var _a, _b;
      if (!trackRef.current)
        return;
      const { min: min2, max: max2, step: step2 } = latestRef.current;
      eventSourceRef.current = "pointer";
      const { bottom, left, height, width } = trackRef.current.getBoundingClientRect();
      const { clientX, clientY } = (_b = (_a = ev.touches) == null ? void 0 : _a[0]) != null ? _b : ev;
      const diff = isVertical ? bottom - clientY : clientX - left;
      const length = isVertical ? height : width;
      let percent = diff / length;
      if (isReversed)
        percent = 1 - percent;
      let nextValue = percentToValue(percent, min2, max2);
      if (step2)
        nextValue = parseFloat(roundNumberToStep(nextValue, min2, step2));
      nextValue = clampNumber(nextValue, min2, max2);
      return nextValue;
    },
    [isVertical, isReversed, latestRef]
  );
  const setValueFromPointer = (ev) => {
    const { value: value2 } = latestRef.current;
    const nextValue = getValueFromPointer(ev);
    if (nextValue != null && nextValue !== value2)
      setValue(nextValue);
  };
  const focusThumb = useCallback(() => {
    const { focusThumbOnChange: focusThumbOnChange2 } = latestRef.current;
    if (focusThumbOnChange2)
      setTimeout(() => {
        var _a;
        return (_a = thumbRef.current) == null ? void 0 : _a.focus();
      });
  }, [latestRef]);
  const constrain = useCallback(
    (value2) => {
      const { isInteractive: isInteractive2, min: min2, max: max2 } = latestRef.current;
      if (!isInteractive2)
        return;
      value2 = parseFloat(roundNumberToStep(value2, min2, oneStep));
      value2 = clampNumber(value2, min2, max2);
      setValue(value2);
    },
    [oneStep, setValue, latestRef]
  );
  const stepUp = useCallback(
    (step2 = oneStep) => constrain(isReversed ? value - step2 : value + step2),
    [constrain, isReversed, oneStep, value]
  );
  const stepDown = useCallback(
    (step2 = oneStep) => constrain(isReversed ? value + step2 : value - step2),
    [constrain, isReversed, oneStep, value]
  );
  const reset = useCallback(
    () => constrain(defaultValue || 0),
    [constrain, defaultValue]
  );
  const stepTo = useCallback((value2) => constrain(value2), [constrain]);
  const onKeyDown = useCallback(
    (ev) => {
      const { min: min2, max: max2 } = latestRef.current;
      const actions = {
        ArrowRight: () => stepUp(),
        ArrowUp: () => stepUp(),
        ArrowLeft: () => stepDown(),
        ArrowDown: () => stepDown(),
        PageUp: () => stepUp(tenStep),
        PageDown: () => stepDown(tenStep),
        Home: () => constrain(min2),
        End: () => constrain(max2)
      };
      const action = actions[ev.key];
      if (!action)
        return;
      ev.preventDefault();
      ev.stopPropagation();
      action(ev);
      eventSourceRef.current = "keyboard";
    },
    [constrain, latestRef, stepDown, stepUp, tenStep]
  );
  useUpdateEffect(() => {
    const { value: value2 } = latestRef.current;
    focusThumb();
    if (eventSourceRef.current === "keyboard")
      onChangeEnd(value2);
  }, [value, onChangeEnd]);
  const getContainerProps = useCallback(
    (props2 = {}, ref = null) => {
      const { width: w, height: h } = thumbSize != null ? thumbSize : { width: 0, height: 0 };
      const style = {
        ...props2.style,
        position: "relative",
        userSelect: "none",
        touchAction: "none",
        WebkitTapHighlightColor: "rgba(0, 0, 0, 0)",
        outline: 0,
        ...isVertical ? { paddingLeft: w / 2, paddingRight: w / 2 } : { paddingTop: h / 2, paddingBottom: h / 2 }
      };
      return {
        ...rest,
        ...props2,
        ref: mergeRefs(ref, containerRef),
        tabIndex: -1,
        style
      };
    },
    [isVertical, rest, thumbSize]
  );
  const getInputProps = useCallback(
    (props2 = {}, ref = null) => ({
      "aria-readonly": ariaReadonly,
      ...formControlProps,
      ...props2,
      id,
      ref,
      type: "hidden",
      name,
      value,
      required,
      disabled,
      readOnly
    }),
    [
      ariaReadonly,
      disabled,
      formControlProps,
      id,
      name,
      readOnly,
      required,
      value
    ]
  );
  const getTrackProps = useCallback(
    (props2 = {}, ref = null) => {
      const style = {
        ...props2.style,
        position: "absolute",
        ...isVertical ? {
          left: "50%",
          transform: "translateX(-50%)",
          height: "100%"
        } : {
          top: "50%",
          transform: "translateY(-50%)",
          width: "100%"
        }
      };
      return {
        ...formControlProps,
        ...props2,
        ref: mergeRefs(ref, trackRef),
        style
      };
    },
    [isVertical, formControlProps]
  );
  const getFilledTrackProps = useCallback(
    (props2 = {}, ref = null) => {
      const n = Math.abs(isReversed ? 100 - thumbPercent : thumbPercent);
      const style = {
        ...props2.style,
        position: "absolute",
        ...isVertical ? {
          left: "50%",
          transform: "translateX(-50%)",
          height: `${n}%`,
          ...isReversed ? { top: "0%" } : { bottom: "0%" }
        } : {
          top: "50%",
          transform: "translateY(-50%)",
          width: `${n}%`,
          ...isReversed ? { right: "0%" } : { left: "0%" }
        }
      };
      return {
        ...formControlProps,
        ...props2,
        ref,
        style
      };
    },
    [isReversed, isVertical, formControlProps, thumbPercent]
  );
  const getMarkProps = useCallback(
    (props2, ref = null) => {
      let n = valueToPercent(props2.value, min, max);
      n = isReversed ? 100 - n : n;
      const style = {
        ...props2.style,
        position: "absolute",
        pointerEvents: "none",
        ...isVertical ? { bottom: `${n}%` } : { left: `${n}%` }
      };
      return {
        ...formControlProps,
        ...props2,
        ref,
        "aria-hidden": true,
        "data-invalid": dataAttr(props2.value < min || max < props2.value),
        "data-highlighted": dataAttr(props2.value <= value),
        style
      };
    },
    [isReversed, isVertical, max, min, formControlProps, value]
  );
  const getThumbProps = useCallback(
    (props2 = {}, ref = null) => {
      const n = thumbPercent;
      const { width: w, height: h } = thumbSize != null ? thumbSize : { width: 0, height: 0 };
      const style = {
        ...props2.style,
        position: "absolute",
        userSelect: "none",
        touchAction: "none",
        ...isVertical ? { bottom: `calc(${n}% - ${h / 2}px)` } : { left: `calc(${n}% - ${w / 2}px)` }
      };
      return {
        "aria-label": "Slider thumb",
        "aria-readonly": ariaReadonly,
        ...formControlProps,
        ...props2,
        ref: mergeRefs(ref, thumbRef),
        tabIndex: isInteractive && focusThumbOnChange ? 0 : void 0,
        role: "slider",
        "aria-valuemin": min,
        "aria-valuemax": max,
        "aria-valuenow": value,
        "data-active": dataAttr(isDragging && focusThumbOnChange),
        "aria-orientation": orientation,
        onKeyDown: handlerAll(props2.onKeyDown, onKeyDown),
        onFocus: handlerAll(props2.onFocus, onFocus, () => setFocused(true)),
        onBlur: handlerAll(props2.onBlur, onBlur, () => setFocused(false)),
        style
      };
    },
    [
      thumbPercent,
      thumbSize,
      isVertical,
      ariaReadonly,
      formControlProps,
      isInteractive,
      focusThumbOnChange,
      min,
      max,
      value,
      isDragging,
      orientation,
      onKeyDown,
      onFocus,
      onBlur
    ]
  );
  return {
    value,
    isFocused,
    isDragging,
    isVertical,
    stepUp,
    stepDown,
    reset,
    stepTo,
    getContainerProps,
    getInputProps,
    getTrackProps,
    getFilledTrackProps,
    getMarkProps,
    getThumbProps
  };
};
var [SliderProvider, useSliderContext] = createContext({
  name: "SliderContext",
  errorMessage: `useSliderContext returned is 'undefined'. Seems you forgot to wrap the components in "<Slider />" `
});
var Slider = forwardRef((props, ref) => {
  const [styles, mergedProps] = useMultiComponentStyle("Slider", props);
  const {
    className,
    children,
    inputProps,
    trackProps,
    filledTrackProps,
    thumbProps,
    trackColor,
    filledTrackColor,
    trackSize,
    thumbColor,
    thumbSize,
    ...rest
  } = omitThemeProps(mergedProps);
  const {
    isVertical,
    getContainerProps,
    getInputProps,
    getTrackProps,
    getFilledTrackProps,
    getMarkProps,
    getThumbProps
  } = useSlider(rest);
  const css = { ...styles.container };
  const validChildren = getValidChildren(children);
  const [customSliderTrack] = findChildren(validChildren, SliderTrack);
  const [customSliderThumb] = findChildren(validChildren, SliderThumb);
  const hasSliderThumb = includesChildren(validChildren, SliderThumb);
  const cloneChildren = !isEmpty(validChildren) ? omitChildren(validChildren, SliderTrack, SliderThumb) : children;
  return /* @__PURE__ */ jsx(
    SliderProvider,
    {
      value: {
        isVertical,
        getTrackProps,
        getFilledTrackProps,
        getMarkProps,
        getThumbProps,
        trackProps,
        trackColor,
        trackSize,
        filledTrackProps,
        filledTrackColor,
        thumbProps,
        thumbColor,
        thumbSize,
        styles
      },
      children: /* @__PURE__ */ jsxs(
        ui.div,
        {
          className: cx("ui-slider", className),
          __css: css,
          ...getContainerProps(),
          children: [
            /* @__PURE__ */ jsx(ui.input, { ...getInputProps(inputProps, ref) }),
            customSliderTrack != null ? customSliderTrack : /* @__PURE__ */ jsx(SliderTrack, {}),
            cloneChildren,
            customSliderThumb != null ? customSliderThumb : !hasSliderThumb ? /* @__PURE__ */ jsx(SliderThumb, {}) : null
          ]
        }
      )
    }
  );
});
var SliderTrack = forwardRef(
  ({ className, children, filledTrackProps, ...rest }, ref) => {
    const {
      styles,
      trackProps,
      trackColor,
      trackSize,
      isVertical,
      getTrackProps
    } = useSliderContext();
    const css = { ...styles.track };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        className: cx("ui-slider__track", className),
        __css: css,
        ...getTrackProps(
          {
            ...trackColor ? { bg: trackColor } : {},
            ...trackSize ? isVertical ? { w: trackSize } : { h: trackSize } : {},
            ...trackProps,
            ...rest
          },
          ref
        ),
        children: children != null ? children : /* @__PURE__ */ jsx(SliderFilledTrack, { ...filledTrackProps })
      }
    );
  }
);
var SliderFilledTrack = forwardRef(
  ({ className, ...rest }, ref) => {
    const { styles, filledTrackProps, filledTrackColor, getFilledTrackProps } = useSliderContext();
    const css = { ...styles.filledTrack };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        className: cx("ui-slider__track-filled", className),
        __css: css,
        ...getFilledTrackProps(
          {
            ...filledTrackColor ? { bg: filledTrackColor } : {},
            ...filledTrackProps,
            ...rest
          },
          ref
        )
      }
    );
  }
);
var SliderMark = forwardRef(
  ({ className, ...rest }, ref) => {
    const { styles, getMarkProps } = useSliderContext();
    const css = {
      display: "inline-flex",
      justifyContent: "center",
      alignItems: "center",
      ...styles.mark
    };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        className: cx("ui-slider__mark", className),
        __css: css,
        ...getMarkProps(rest, ref)
      }
    );
  }
);
var SliderThumb = forwardRef(
  ({ className, ...rest }, ref) => {
    const { styles, thumbProps, thumbColor, thumbSize, getThumbProps } = useSliderContext();
    const css = { ...styles.thumb };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        className: cx("ui-slider__thumb", className),
        __css: css,
        ...getThumbProps(
          {
            ...thumbColor ? { bg: thumbColor } : {},
            ...thumbSize ? { boxSize: thumbSize } : {},
            ...thumbProps,
            ...rest
          },
          ref
        )
      }
    );
  }
);

export {
  useSlider,
  Slider,
  SliderTrack,
  SliderFilledTrack,
  SliderMark,
  SliderThumb
};
