// @yamada-ui/slider/dist/chunk-W3VE2JEC.mjs
"use client"

// src/range-slider.tsx
import {
  ui,
  forwardRef,
  useMultiComponentStyle,
  omitThemeProps
} from "@yamada-ui/core";
import {
  useFormControlProps,
  formControlProperties
} from "@yamada-ui/form-control";
import { useControllableState } from "@yamada-ui/use-controllable-state";
import { useLatestRef } from "@yamada-ui/use-latest-ref";
import { usePanEvent } from "@yamada-ui/use-pan-event";
import { useSizes } from "@yamada-ui/use-size";
import {
  clampNumber,
  createContext,
  cx,
  dataAttr,
  omitChildren,
  findChildren,
  getValidChildren,
  handlerAll,
  isArray,
  isEmpty,
  mergeRefs,
  percentToValue,
  pickObject,
  roundNumberToStep,
  useCallbackRef,
  useUpdateEffect,
  valueToPercent,
  includesChildren
} from "@yamada-ui/utils";
import { useCallback, useId, useRef, useState } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
var useRangeSlider = ({
  focusThumbOnChange = true,
  ...props
}) => {
  if (!focusThumbOnChange)
    props.isReadOnly = true;
  let {
    id,
    name,
    min = 0,
    max = 100,
    step = 1,
    defaultValue,
    orientation = "horizontal",
    isReversed,
    betweenThumbs = 0,
    required,
    disabled,
    readOnly,
    value: valueProp,
    onChange,
    onChangeStart: onChangeStartProp,
    onChangeEnd: onChangeEndProp,
    onFocus,
    onBlur,
    "aria-readonly": ariaReadonly,
    ...rest
  } = useFormControlProps(props);
  const formControlProps = pickObject(rest, formControlProperties);
  defaultValue = defaultValue != null ? defaultValue : [min + (max - min) / 4, max - (max - min) / 4];
  if (max < min)
    throw new Error("Do not assign a number less than 'min' to 'max'");
  const onChangeStart = useCallbackRef(onChangeStartProp);
  const onChangeEnd = useCallbackRef(onChangeEndProp);
  const [computedValues, setValues] = useControllableState({
    value: valueProp,
    defaultValue,
    onChange
  });
  const [isDragging, setDragging] = useState(false);
  const [isFocused, setFocused] = useState(false);
  const [activeIndex, setActiveIndex] = useState(-1);
  const isInteractive = !(disabled || readOnly);
  const tenStep = (max - min) / 10;
  const oneStep = step || (max - min) / 100;
  const spacing = betweenThumbs * step;
  const values = computedValues.map(
    (value) => clampNumber(value, min, max)
  );
  const [startValue, endValue] = values;
  const reversedValues = values.map((value) => max - value + min);
  const thumbValues = isReversed ? reversedValues : values;
  const thumbPercents = thumbValues.map(
    (value) => valueToPercent(value, min, max)
  );
  const valueBounds = [
    { min, max: endValue - spacing },
    { min: startValue + spacing, max }
  ];
  const isVertical = orientation === "vertical";
  const latestRef = useLatestRef({
    min,
    max,
    step,
    disabled,
    values,
    valueBounds,
    isInteractive,
    isReversed,
    isVertical,
    focusThumbOnChange,
    betweenThumbs,
    orientation
  });
  const eventSourceRef = useRef(null);
  const containerRef = useRef(null);
  const trackRef = useRef(null);
  const thumbSizes = useSizes({
    getNodes: () => {
      var _a;
      const nodes = (_a = containerRef.current) == null ? void 0 : _a.querySelectorAll("[role=slider]");
      return nodes ? Array.from(nodes) : [];
    }
  });
  id != null ? id : id = useId();
  name != null ? name : name = id;
  const getThumbId = useCallback((i) => `slider-thumb-${id}-${i}`, [id]);
  const getInputId = useCallback((i) => `slider-input-${id}-${i}`, [id]);
  const getMarkerId = useCallback(
    (i) => `slider-marker-${id}-${i}`,
    [id]
  );
  usePanEvent(containerRef, {
    onSessionStart: (ev) => {
      const { isInteractive: isInteractive2, values: values2 } = latestRef.current;
      if (!isInteractive2)
        return;
      setDragging(true);
      const pointValue = getValueFromPointer(ev) || 0;
      const distances = values2.map((value) => Math.abs(value - pointValue));
      const closest = Math.min(...distances);
      let i = distances.indexOf(closest);
      const thumbsPosition = distances.filter(
        (distance) => distance === closest
      );
      const isThumbStacked = thumbsPosition.length > 1;
      if (isThumbStacked && pointValue > values2[i])
        i = i + thumbsPosition.length - 1;
      setActiveIndex(i);
      constrain(i, pointValue);
      focusThumb(i);
      onChangeStart(values2);
    },
    onSessionEnd: () => {
      const { isInteractive: isInteractive2, values: values2 } = latestRef.current;
      if (!isInteractive2)
        return;
      setDragging(false);
      onChangeEnd(values2);
    },
    onMove: (ev) => {
      const { isInteractive: isInteractive2 } = latestRef.current;
      if (!isInteractive2 || activeIndex == -1)
        return;
      const pointValue = getValueFromPointer(ev) || 0;
      setActiveIndex(activeIndex);
      constrain(activeIndex, pointValue);
      focusThumb(activeIndex);
    }
  });
  const getValueFromPointer = useCallback(
    (ev) => {
      var _a, _b;
      if (!trackRef.current)
        return;
      const { min: min2, max: max2 } = latestRef.current;
      eventSourceRef.current = "pointer";
      const { bottom, left, height, width } = trackRef.current.getBoundingClientRect();
      const { clientX, clientY } = (_b = (_a = ev.touches) == null ? void 0 : _a[0]) != null ? _b : ev;
      const diff = isVertical ? bottom - clientY : clientX - left;
      const length = isVertical ? height : width;
      let percent = diff / length;
      if (isReversed)
        percent = 1 - percent;
      let nextValue = percentToValue(percent, min2, max2);
      return nextValue;
    },
    [latestRef, isVertical, isReversed]
  );
  const focusThumb = useCallback(
    (i) => {
      var _a;
      i != null ? i : i = activeIndex;
      if (i === -1 || !focusThumbOnChange)
        return;
      const id2 = getThumbId(i);
      const el = (_a = containerRef.current) == null ? void 0 : _a.ownerDocument.getElementById(id2);
      if (el)
        setTimeout(() => el.focus());
    },
    [activeIndex, focusThumbOnChange, getThumbId]
  );
  const constrain = useCallback(
    (i, value) => {
      const { isInteractive: isInteractive2, valueBounds: valueBounds2, values: values2 } = latestRef.current;
      if (!isInteractive2)
        return;
      const { min: min2, max: max2 } = valueBounds2[i];
      value = parseFloat(roundNumberToStep(value, min2, oneStep));
      value = clampNumber(value, min2, max2);
      const nextValues = [...values2];
      nextValues[i] = value;
      setValues(nextValues);
    },
    [latestRef, oneStep, setValues]
  );
  const stepUp = useCallback(
    (i, step2 = oneStep) => {
      const { values: values2 } = latestRef.current;
      const value = values2[i];
      constrain(i, isReversed ? value - step2 : value + step2);
    },
    [constrain, isReversed, latestRef, oneStep]
  );
  const stepDown = useCallback(
    (i, step2 = oneStep) => {
      const { values: values2 } = latestRef.current;
      const value = values2[i];
      constrain(i, isReversed ? value + step2 : value - step2);
    },
    [constrain, isReversed, latestRef, oneStep]
  );
  const reset = useCallback(
    () => setValues(defaultValue),
    [defaultValue, setValues]
  );
  const onKeyDown = useCallback(
    (ev) => {
      const { valueBounds: valueBounds2 } = latestRef.current;
      const { min: min2, max: max2 } = valueBounds2[activeIndex];
      const actions = {
        ArrowRight: () => stepUp(activeIndex),
        ArrowUp: () => stepUp(activeIndex),
        ArrowLeft: () => stepDown(activeIndex),
        ArrowDown: () => stepDown(activeIndex),
        PageUp: () => stepUp(activeIndex, tenStep),
        PageDown: () => stepDown(activeIndex, tenStep),
        Home: () => constrain(activeIndex, min2),
        End: () => constrain(activeIndex, max2)
      };
      const action = actions[ev.key];
      if (!action)
        return;
      ev.preventDefault();
      ev.stopPropagation();
      action(ev);
      eventSourceRef.current = "keyboard";
    },
    [activeIndex, constrain, latestRef, stepDown, stepUp, tenStep]
  );
  useUpdateEffect(() => {
    const { values: values2 } = latestRef.current;
    if (eventSourceRef.current === "keyboard")
      onChangeEnd(values2);
  }, [startValue, endValue, onChangeEnd]);
  const getContainerProps = useCallback(
    (props2 = {}, ref = null) => {
      var _a;
      const z = { width: 0, height: 0 };
      const p = isVertical ? "height" : "width";
      const { width: w, height: h } = (_a = thumbSizes.reduce((a = z, b = z) => a[p] > b[p] ? a : b, z)) != null ? _a : {};
      const style = {
        ...props2.style,
        position: "relative",
        userSelect: "none",
        touchAction: "none",
        WebkitTapHighlightColor: "rgba(0, 0, 0, 0)",
        outline: 0,
        ...isVertical ? w ? { paddingLeft: w / 2, paddingRight: w / 2 } : {} : h ? { paddingTop: h / 2, paddingBottom: h / 2 } : {}
      };
      return {
        ...rest,
        ...props2,
        id: `slider-container-${id}`,
        ref: mergeRefs(ref, containerRef),
        tabIndex: -1,
        style
      };
    },
    [id, isVertical, rest, thumbSizes]
  );
  const getInputProps = useCallback(
    ({ index: i, ...props2 }, ref = null) => ({
      "aria-readonly": ariaReadonly,
      ...formControlProps,
      ...props2,
      ref,
      id: getInputId(i),
      type: "hidden",
      name: isArray(name) ? name[i] : `${name}-${i}`,
      value: values[i],
      required,
      disabled,
      readOnly
    }),
    [
      ariaReadonly,
      disabled,
      getInputId,
      name,
      readOnly,
      required,
      formControlProps,
      values
    ]
  );
  const getTrackProps = useCallback(
    (props2 = {}, ref = null) => {
      const style = {
        ...props2.style,
        position: "absolute",
        ...isVertical ? {
          left: "50%",
          transform: "translateX(-50%)",
          height: "100%"
        } : {
          top: "50%",
          transform: "translateY(-50%)",
          width: "100%"
        }
      };
      return {
        ...formControlProps,
        ...props2,
        id: `slider-track-${id}`,
        ref: mergeRefs(ref, trackRef),
        style
      };
    },
    [id, isVertical, formControlProps]
  );
  const getFilledTrackProps = useCallback(
    (props2 = {}, ref = null) => {
      const n = Math.abs(thumbPercents[1] - thumbPercents[0]);
      const s = isReversed ? 100 - thumbPercents[0] : thumbPercents[0];
      const style = {
        ...props2.style,
        position: "absolute",
        ...isVertical ? {
          left: "50%",
          transform: "translateX(-50%)",
          height: `${n}%`,
          ...isReversed ? { top: `${s}%` } : { bottom: `${s}%` }
        } : {
          top: "50%",
          transform: "translateY(-50%)",
          width: `${n}%`,
          ...isReversed ? { right: `${s}%` } : { left: `${s}%` }
        }
      };
      return {
        ...formControlProps,
        ...props2,
        id: `slider-filled-track-${id}`,
        ref,
        style
      };
    },
    [id, isReversed, isVertical, formControlProps, thumbPercents]
  );
  const getMarkProps = useCallback(
    (props2, ref = null) => {
      let n = valueToPercent(props2.value, min, max);
      n = isReversed ? 100 - n : n;
      const style = {
        ...props2.style,
        position: "absolute",
        pointerEvents: "none",
        ...isVertical ? { bottom: `${n}%` } : { left: `${n}%` }
      };
      return {
        ...formControlProps,
        ...props2,
        ref,
        id: getMarkerId(props2.value),
        "aria-hidden": true,
        "data-invalid": dataAttr(props2.value < min || max < props2.value),
        "data-highlighted": dataAttr(
          values[0] <= props2.value && props2.value <= values[1]
        ),
        style
      };
    },
    [getMarkerId, isReversed, isVertical, max, min, formControlProps, values]
  );
  const getThumbProps = useCallback(
    ({ index: i, ...props2 }, ref = null) => {
      var _a;
      const n = thumbPercents[i];
      const { width: w, height: h } = (_a = thumbSizes[i]) != null ? _a : { width: 0, height: 0 };
      const style = {
        ...props2.style,
        position: "absolute",
        userSelect: "none",
        touchAction: "none",
        ...isVertical ? { bottom: `calc(${n}% - ${h / 2}px)` } : { left: `calc(${n}% - ${w / 2}px)` }
      };
      const value = values[i];
      if (value == null)
        throw new Error(
          `Cannot find value at index '${i}'. The 'value' or 'defaultValue'`
        );
      return {
        "aria-label": "Slider thumb",
        "aria-readonly": ariaReadonly,
        ...formControlProps,
        ...props2,
        ref,
        id: getThumbId(i),
        tabIndex: isInteractive && focusThumbOnChange ? 0 : void 0,
        role: "slider",
        "aria-valuemin": min,
        "aria-valuemax": max,
        "aria-valuenow": value,
        "data-active": dataAttr(
          isDragging && focusThumbOnChange && activeIndex === i
        ),
        "aria-orientation": orientation,
        onKeyDown: handlerAll(props2.onKeyDown, onKeyDown),
        onFocus: handlerAll(props2.onFocus, onFocus, () => {
          setFocused(true);
          setActiveIndex(i);
        }),
        onBlur: handlerAll(props2.onBlur, onBlur, () => {
          setFocused(false);
          setActiveIndex(-1);
        }),
        style
      };
    },
    [
      thumbPercents,
      thumbSizes,
      isVertical,
      values,
      ariaReadonly,
      formControlProps,
      getThumbId,
      isInteractive,
      focusThumbOnChange,
      min,
      max,
      isDragging,
      activeIndex,
      orientation,
      onKeyDown,
      onFocus,
      onBlur
    ]
  );
  return {
    values,
    isFocused,
    isDragging,
    isVertical,
    stepUp,
    stepDown,
    reset,
    setActiveIndex,
    getThumbId,
    getInputId,
    getMarkerId,
    getContainerProps,
    getInputProps,
    getTrackProps,
    getFilledTrackProps,
    getMarkProps,
    getThumbProps
  };
};
var [RangeSliderProvider, useRangeSliderContext] = createContext({
  name: "RangeSliderContext",
  errorMessage: `useRangeSliderContext returned is 'undefined'. Seems you forgot to wrap the components in "<RangeSlider />" `
});
var RangeSlider = forwardRef((props, ref) => {
  const [styles, mergedProps] = useMultiComponentStyle("RangeSlider", props);
  const {
    className,
    children,
    inputProps,
    trackProps,
    filledTrackProps,
    thumbProps,
    trackColor,
    filledTrackColor,
    trackSize,
    thumbColor,
    thumbSize,
    ...rest
  } = omitThemeProps(mergedProps);
  const {
    isVertical,
    getContainerProps,
    getInputProps,
    getTrackProps,
    getFilledTrackProps,
    getMarkProps,
    getThumbProps
  } = useRangeSlider(rest);
  const css = { ...styles.container };
  const validChildren = getValidChildren(children);
  const [customRangeSliderTrack] = findChildren(validChildren, RangeSliderTrack);
  const [customRangeSliderStartThumb] = findChildren(
    validChildren,
    RangeSliderStartThumb
  );
  const [customRangeSliderEndThumb] = findChildren(
    validChildren,
    RangeSliderEndThumb
  );
  const hasRangeSliderStartThumb = includesChildren(
    validChildren,
    RangeSliderStartThumb
  );
  const hasRangeSliderEndThumb = includesChildren(
    validChildren,
    RangeSliderEndThumb
  );
  const cloneChildren = !isEmpty(validChildren) ? omitChildren(
    validChildren,
    RangeSliderTrack,
    RangeSliderStartThumb,
    RangeSliderEndThumb
  ) : children;
  return /* @__PURE__ */ jsx(
    RangeSliderProvider,
    {
      value: {
        isVertical,
        getTrackProps,
        getFilledTrackProps,
        getMarkProps,
        getThumbProps,
        getInputProps,
        trackProps,
        trackColor,
        trackSize,
        filledTrackProps,
        filledTrackColor,
        thumbProps,
        thumbColor,
        thumbSize,
        inputProps,
        styles
      },
      children: /* @__PURE__ */ jsxs(
        ui.div,
        {
          className: cx("ui-slider", className),
          __css: css,
          ...getContainerProps({}, ref),
          children: [
            customRangeSliderTrack != null ? customRangeSliderTrack : /* @__PURE__ */ jsx(RangeSliderTrack, {}),
            cloneChildren,
            customRangeSliderStartThumb != null ? customRangeSliderStartThumb : !hasRangeSliderStartThumb ? /* @__PURE__ */ jsx(RangeSliderStartThumb, {}) : null,
            customRangeSliderEndThumb != null ? customRangeSliderEndThumb : !hasRangeSliderEndThumb ? /* @__PURE__ */ jsx(RangeSliderEndThumb, {}) : null
          ]
        }
      )
    }
  );
});
var RangeSliderTrack = forwardRef(
  ({ className, children, filledTrackProps, ...rest }, ref) => {
    const {
      styles,
      trackProps,
      trackColor,
      trackSize,
      isVertical,
      getTrackProps
    } = useRangeSliderContext();
    const css = { ...styles.track };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        className: cx("ui-slider__track", className),
        __css: css,
        ...getTrackProps(
          {
            ...trackColor ? { bg: trackColor } : {},
            ...trackSize ? isVertical ? { w: trackSize } : { h: trackSize } : {},
            ...trackProps,
            ...rest
          },
          ref
        ),
        children: children != null ? children : /* @__PURE__ */ jsx(RangeSliderFilledTrack, { ...filledTrackProps })
      }
    );
  }
);
var RangeSliderFilledTrack = forwardRef(({ className, ...rest }, ref) => {
  const { styles, filledTrackProps, filledTrackColor, getFilledTrackProps } = useRangeSliderContext();
  const css = { ...styles.filledTrack };
  return /* @__PURE__ */ jsx(
    ui.div,
    {
      className: cx("ui-slider__track-filled", className),
      __css: css,
      ...getFilledTrackProps(
        {
          ...filledTrackColor ? { bg: filledTrackColor } : {},
          ...filledTrackProps,
          ...rest
        },
        ref
      )
    }
  );
});
var RangeSliderMark = forwardRef(
  ({ className, ...rest }, ref) => {
    const { styles, getMarkProps } = useRangeSliderContext();
    const css = {
      display: "inline-flex",
      justifyContent: "center",
      alignItems: "center",
      ...styles.mark
    };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        className: cx("ui-slider__mark", className),
        __css: css,
        ...getMarkProps(rest, ref)
      }
    );
  }
);
var RangeSliderThumb = forwardRef(({ className, index, children, ...rest }, ref) => {
  const {
    styles,
    thumbProps,
    thumbColor,
    thumbSize,
    inputProps,
    getThumbProps,
    getInputProps
  } = useRangeSliderContext();
  const css = { ...styles.thumb };
  const { children: propChildren } = thumbProps != null ? thumbProps : {};
  return /* @__PURE__ */ jsxs(
    ui.div,
    {
      className: cx("ui-slider__thumb", className),
      __css: css,
      ...getThumbProps(
        {
          index,
          ...thumbColor ? { bg: thumbColor } : {},
          ...thumbSize ? { boxSize: thumbSize } : {},
          ...thumbProps,
          ...rest
        },
        ref
      ),
      children: [
        /* @__PURE__ */ jsx(ui.input, { ...getInputProps({ ...inputProps, index }, ref) }),
        children != null ? children : propChildren
      ]
    }
  );
});
var RangeSliderStartThumb = forwardRef(
  (rest, ref) => {
    return /* @__PURE__ */ jsx(RangeSliderThumb, { ref, index: 0, ...rest });
  }
);
var RangeSliderEndThumb = forwardRef(
  (rest, ref) => {
    return /* @__PURE__ */ jsx(RangeSliderThumb, { ref, index: 1, ...rest });
  }
);

export {
  useRangeSlider,
  RangeSlider,
  RangeSliderTrack,
  RangeSliderFilledTrack,
  RangeSliderMark,
  RangeSliderStartThumb,
  RangeSliderEndThumb
};
