// @yamada-ui/card/dist/chunk-F3TZ4KBM.mjs
"use client"

// src/card.tsx
import {
  ui,
  forwardRef,
  omitThemeProps,
  useMultiComponentStyle
} from "@yamada-ui/core";
import { createContext, cx } from "@yamada-ui/utils";
import { jsx } from "react/jsx-runtime";
var [CardProvider, useCard] = createContext({
  name: `CardContext`,
  errorMessage: `useCard returned is 'undefined'. Seems you forgot to wrap the components in "<Card />" `
});
var Card = forwardRef((props, ref) => {
  const [styles, mergedProps] = useMultiComponentStyle("Card", props);
  const {
    className,
    direction: flexDirection = "column",
    justify: justifyContent,
    align: alignItems,
    ...rest
  } = omitThemeProps(mergedProps);
  const css = {
    display: "flex",
    flexDirection,
    justifyContent,
    alignItems,
    wordWrap: "break-word",
    ...styles.container
  };
  return /* @__PURE__ */ jsx(CardProvider, { value: styles, children: /* @__PURE__ */ jsx(
    ui.article,
    {
      ref,
      className: cx("ui-card", className),
      __css: css,
      ...rest
    }
  ) });
});
var CardHeader = forwardRef(
  ({ className, ...rest }, ref) => {
    const styles = useCard();
    const css = {
      display: "flex",
      justifyContent: "flex-start",
      alignItems: "center",
      ...styles.header
    };
    return /* @__PURE__ */ jsx(
      ui.header,
      {
        ref,
        className: cx("ui-card__header", className),
        __css: css,
        ...rest
      }
    );
  }
);
var CardBody = forwardRef(
  ({ className, ...rest }, ref) => {
    const styles = useCard();
    const css = {
      display: "flex",
      flexDirection: "column",
      alignItems: "flex-start",
      ...styles.body
    };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        ref,
        className: cx("ui-card__body", className),
        __css: css,
        ...rest
      }
    );
  }
);
var CardFooter = forwardRef(
  ({ className, ...rest }, ref) => {
    const styles = useCard();
    const css = {
      display: "flex",
      justifyContent: "flex-start",
      alignItems: "center",
      ...styles.footer
    };
    return /* @__PURE__ */ jsx(
      ui.footer,
      {
        ref,
        className: cx("ui-card__footer", className),
        __css: css,
        ...rest
      }
    );
  }
);

export {
  Card,
  CardHeader,
  CardBody,
  CardFooter
};
