// @yamada-ui/color-picker/dist/chunk-7BDU5DSE.mjs
"use client"
import {
  ColorSelectorEyeDropper
} from "./chunk-IU3VEGY6.mjs";
import {
  ColorSelectorSliders
} from "./chunk-JOX5PPY7.mjs";
import {
  ColorSwatch
} from "./chunk-UQWAM6TA.mjs";
import {
  useColorSelectorContext
} from "./chunk-5QT7SRA4.mjs";

// src/color-selector-body.tsx
import { forwardRef, ui } from "@yamada-ui/core";
import { cx } from "@yamada-ui/utils";
import { jsx, jsxs } from "react/jsx-runtime";
var ColorSelectorBody = forwardRef(
  ({
    className,
    withEyeDropper = true,
    withResult,
    eyeDropperRef,
    eyeDropperProps,
    hueSliderRef,
    hueSliderProps,
    alphaSliderRef,
    alphaSliderProps,
    ...rest
  }, ref) => {
    const { value, eyeDropperSupported, styles } = useColorSelectorContext();
    const css = {
      display: "flex",
      w: "100%",
      ...styles.body
    };
    return /* @__PURE__ */ jsxs(
      ui.div,
      {
        ref,
        className: cx("ui-color-selector__body", className),
        __css: css,
        ...rest,
        children: [
          /* @__PURE__ */ jsx(
            ColorSelectorSliders,
            {
              ...{
                hueSliderRef,
                hueSliderProps,
                alphaSliderRef,
                alphaSliderProps
              }
            }
          ),
          withEyeDropper && eyeDropperSupported ? /* @__PURE__ */ jsx(ColorSelectorEyeDropper, { ref: eyeDropperRef, ...eyeDropperProps }) : null,
          withResult ? /* @__PURE__ */ jsx(
            ColorSwatch,
            {
              className: "ui-color-selector__result",
              color: value,
              __css: { ...styles.result }
            }
          ) : null
        ]
      }
    );
  }
);

export {
  ColorSelectorBody
};
