// @yamada-ui/color-picker/dist/chunk-RR6V37S5.mjs
"use client"
import {
  useSaturationSlider
} from "./chunk-KKQ3MOXJ.mjs";

// src/saturation-slider.tsx
import {
  ui,
  forwardRef,
  useMultiComponentStyle,
  omitThemeProps
} from "@yamada-ui/core";
import { cx, replaceObject } from "@yamada-ui/utils";
import { jsx, jsxs } from "react/jsx-runtime";
var SaturationSlider = forwardRef(
  (props, ref) => {
    const [styles, mergedProps] = useMultiComponentStyle(
      "SaturationSlider",
      props
    );
    const {
      className,
      ratio = 16 / 9,
      innerProps,
      inputProps,
      trackProps,
      thumbProps,
      __css,
      ...computedProps
    } = omitThemeProps(mergedProps);
    const {
      overlays,
      getContainerProps,
      getInnerProps,
      getTrackProps,
      getInputProps,
      getThumbProps
    } = useSaturationSlider(computedProps);
    const css = {
      position: "relative",
      _before: {
        content: `""`,
        display: "block",
        h: 0,
        pb: replaceObject(ratio, (r) => `${1 / r * 100}%`)
      },
      "& > *": {
        position: "absolute",
        top: "0",
        right: "0",
        bottom: "0",
        left: "0",
        display: "flex",
        justifyContent: "center",
        alignItems: "center",
        w: "100%",
        h: "100%"
      },
      ...styles.container,
      ...__css
    };
    return /* @__PURE__ */ jsx(
      ui.div,
      {
        className: cx("ui-saturation-slider", className),
        __css: css,
        ...getContainerProps(),
        children: /* @__PURE__ */ jsxs(
          ui.div,
          {
            className: cx("ui-saturation-slider__inner", className),
            __css: { ...styles.inner },
            ...getInnerProps(innerProps),
            children: [
              /* @__PURE__ */ jsx(ui.input, { ...getInputProps(inputProps, ref) }),
              overlays.map((props2, index) => /* @__PURE__ */ jsx(
                ui.div,
                {
                  className: "ui-saturation-slider__overlay",
                  __css: {
                    position: "absolute",
                    top: 0,
                    left: 0,
                    right: 0,
                    bottom: 0,
                    ...styles.overlay
                  },
                  ...props2
                },
                index
              )),
              /* @__PURE__ */ jsx(
                ui.div,
                {
                  className: "ui-saturation-slider__track",
                  __css: {
                    position: "relative",
                    w: "100%",
                    h: "100%",
                    ...styles.track
                  },
                  ...getTrackProps(trackProps),
                  children: /* @__PURE__ */ jsx(
                    ui.div,
                    {
                      className: "ui-saturation-slider__thumb",
                      __css: { ...styles.thumb },
                      ...getThumbProps(thumbProps)
                    }
                  )
                }
              )
            ]
          }
        )
      }
    );
  }
);

export {
  SaturationSlider
};
