// @yamada-ui/color-picker/dist/use-color-picker.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/use-color-picker.ts
var use_color_picker_exports = {};
__export(use_color_picker_exports, {
  ColorPickerProvider: () => ColorPickerProvider,
  useColorPicker: () => useColorPicker,
  useColorPickerContext: () => useColorPickerContext
});
module.exports = __toCommonJS(use_color_picker_exports);
var import_core = require("@yamada-ui/core");
var import_form_control = require("@yamada-ui/form-control");
var import_popover = require("@yamada-ui/popover");
var import_use_controllable_state = require("@yamada-ui/use-controllable-state");
var import_use_disclosure = require("@yamada-ui/use-disclosure");
var import_use_eye_dropper = require("@yamada-ui/use-eye-dropper");
var import_use_outside_click = require("@yamada-ui/use-outside-click");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var [ColorPickerProvider, useColorPickerContext] = (0, import_utils.createContext)({
  name: "ColorPickerContext",
  errorMessage: `useColorPickerContext returned is 'undefined'. Seems you forgot to wrap the components in "<ColorPicker />"`
});
var useColorPicker = ({
  value: valueProp,
  defaultValue,
  fallbackValue,
  defaultColor,
  onChange: onChangeProp,
  onChangeStart,
  onChangeEnd,
  onSwatchClick,
  formatInput = (value) => value,
  closeOnBlur = true,
  closeOnEsc = true,
  placement = "bottom-start",
  duration = 0.2,
  isOpen: isOpenProp,
  defaultIsOpen,
  onOpen: onOpenProp,
  onClose: onCloseProp,
  allowInput = true,
  closeOnSelectSwatch,
  format,
  swatchesLabel,
  swatches,
  swatchesColumns,
  withPicker,
  withChannel,
  withResult = false,
  withColorSelectorEyeDropper = false,
  colorSelectorVariant,
  colorSelectorSize,
  colorSelectorColorScheme,
  ...rest
}) => {
  var _a;
  rest = (0, import_form_control.useFormControlProps)(rest);
  const { "aria-readonly": _ariaReadonly, ...formControlProps } = (0, import_utils.pickObject)(
    rest,
    import_form_control.formControlProperties
  );
  const { disabled, readOnly } = formControlProps;
  const [containerProps, inputProps] = (0, import_utils.splitObject)(
    (0, import_utils.omitObject)(rest, [...import_popover.popoverProperties]),
    import_core.layoutStyleProperties
  );
  const containerRef = (0, import_react.useRef)(null);
  const fieldRef = (0, import_react.useRef)(null);
  const { supported: eyeDropperSupported, onOpen: onEyeDropperOpen } = (0, import_use_eye_dropper.useEyeDropper)();
  const [value, setValue] = (0, import_use_controllable_state.useControllableState)({
    value: valueProp,
    defaultValue,
    onChange: onChangeProp
  });
  const formatRef = (0, import_react.useRef)(
    format != null ? format : (0, import_utils.calcFormat)((_a = value != null ? value : defaultColor) != null ? _a : "")
  );
  const isInputFocused = (0, import_react.useRef)(false);
  const [inputValue, setInputValue] = (0, import_react.useState)(value != null ? value : "");
  const {
    isOpen,
    onOpen: onInternalOpen,
    onClose: onInternalClose
  } = (0, import_use_disclosure.useDisclosure)({
    isOpen: isOpenProp,
    defaultIsOpen,
    onOpen: onOpenProp,
    onClose: onCloseProp
  });
  const isColorSelectorFull = colorSelectorSize === "full";
  const onOpen = (0, import_react.useCallback)(() => {
    if (disabled || readOnly)
      return;
    onInternalOpen();
  }, [onInternalOpen, disabled, readOnly]);
  const onClose = (0, import_react.useCallback)(() => {
    if (!isOpen)
      return;
    const next = (0, import_utils.convertColor)(value, fallbackValue)(formatRef.current);
    setValue((prev) => !next || prev === next ? prev : next);
    setInputValue(formatInput(next != null ? next : ""));
    onInternalClose();
  }, [
    formatRef,
    isOpen,
    setValue,
    onInternalClose,
    value,
    formatInput,
    setInputValue,
    fallbackValue
  ]);
  const onContainerClick = (0, import_react.useCallback)(() => {
    if (isOpen)
      return;
    onOpen();
  }, [isOpen, onOpen]);
  const onInputFocus = (0, import_react.useCallback)(() => {
    isInputFocused.current = true;
    if (isOpen)
      return;
    onOpen();
  }, [isOpen, onOpen]);
  const onInputBlur = (0, import_react.useCallback)(() => {
    isInputFocused.current = false;
  }, []);
  const onContainerBlur = (0, import_react.useCallback)(
    (ev) => {
      const relatedTarget = (0, import_utils.getEventRelatedTarget)(ev);
      if ((0, import_utils.isContains)(containerRef.current, relatedTarget))
        return;
      if (!closeOnBlur)
        return;
      if (isOpen)
        onClose();
    },
    [closeOnBlur, isOpen, onClose]
  );
  const onInputKeyDown = (0, import_react.useCallback)(
    (ev) => {
      if (ev.key === " ")
        ev.key = ev.code;
      if (disabled || readOnly)
        return;
      const actions = {
        Space: !isOpen ? onOpen : void 0,
        Enter: !isOpen ? onOpen : void 0,
        Escape: closeOnEsc ? onClose : void 0
      };
      const action = actions[ev.key];
      if (!action)
        return;
      ev.preventDefault();
      ev.stopPropagation();
      action();
    },
    [disabled, readOnly, isOpen, onOpen, closeOnEsc, onClose]
  );
  const onInputChange = (0, import_react.useCallback)(
    (ev) => {
      const value2 = ev.target.value;
      setInputValue(formatInput(value2));
      setValue(value2);
    },
    [setInputValue, formatInput, setValue]
  );
  const onColorSelectorChange = (0, import_react.useCallback)(
    (value2) => {
      setValue(value2);
      setTimeout(() => {
        if (!isInputFocused.current)
          setInputValue(formatInput(value2));
      });
    },
    [setValue, formatInput]
  );
  const onEyeDropperClick = (0, import_react.useCallback)(
    async (ev) => {
      var _a2;
      ev.preventDefault();
      ev.stopPropagation();
      try {
        const { sRGBHex } = (_a2 = await onEyeDropperOpen()) != null ? _a2 : {};
        if (!sRGBHex)
          return;
        onColorSelectorChange(sRGBHex);
        onChangeEnd == null ? void 0 : onChangeEnd(sRGBHex);
      } catch {
      }
    },
    [onEyeDropperOpen, onColorSelectorChange, onChangeEnd]
  );
  (0, import_use_outside_click.useOutsideClick)({
    ref: containerRef,
    handler: onClose,
    enabled: isOpen && closeOnBlur
  });
  (0, import_utils.useUpdateEffect)(() => {
    if (!format || !value)
      return;
    formatRef.current = format;
    const nextValue = (0, import_utils.convertColor)(value, fallbackValue)(format);
    if (!nextValue)
      return;
    setInputValue(formatInput(nextValue));
    setValue(nextValue);
  }, [format]);
  (0, import_utils.useUpdateEffect)(() => {
    if (isInputFocused.current || !valueProp)
      return;
    setInputValue(formatInput(valueProp));
  }, [valueProp]);
  const getPopoverProps = (0, import_react.useCallback)(
    (props) => ({
      matchWidth: isColorSelectorFull,
      ...rest,
      ...props,
      isOpen,
      onOpen,
      onClose,
      placement,
      duration,
      trigger: "never",
      closeOnButton: false
    }),
    [isColorSelectorFull, duration, onClose, onOpen, placement, rest, isOpen]
  );
  const getContainerProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      ref: (0, import_utils.mergeRefs)(containerRef, ref),
      ...containerProps,
      ...props,
      ...formControlProps,
      onClick: (0, import_utils.handlerAll)(props.onClick, rest.onClick, onContainerClick),
      onBlur: (0, import_utils.handlerAll)(props.onBlur, rest.onBlur, onContainerBlur)
    }),
    [containerProps, formControlProps, onContainerBlur, onContainerClick, rest]
  );
  const getFieldProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      ref: (0, import_utils.mergeRefs)(fieldRef, ref),
      tabIndex: !allowInput ? -1 : 0,
      ...inputProps,
      ...props,
      style: {
        ...props.style,
        ...!allowInput ? { pointerEvents: "none" } : {}
      },
      value: inputValue,
      "data-active": (0, import_utils.dataAttr)(isOpen),
      "aria-expanded": (0, import_utils.dataAttr)(isOpen),
      onFocus: (0, import_utils.handlerAll)(props.onFocus, rest.onFocus, onInputFocus),
      onKeyDown: (0, import_utils.handlerAll)(props.onKeyDown, rest.onKeyDown, onInputKeyDown),
      onChange: (0, import_utils.handlerAll)(props.onChange, onInputChange),
      onBlur: (0, import_utils.handlerAll)(props.onFocus, onInputBlur)
    }),
    [
      allowInput,
      inputProps,
      inputValue,
      isOpen,
      rest,
      onInputFocus,
      onInputKeyDown,
      onInputChange,
      onInputBlur
    ]
  );
  const getEyeDropperProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      disabled,
      "aria-label": "Pick a color",
      ...props,
      ref,
      style: { ...props.style, pointerEvents: readOnly ? "none" : void 0 },
      onClick: (0, import_utils.handlerAll)(props.onClick, onEyeDropperClick)
    }),
    [disabled, onEyeDropperClick, readOnly]
  );
  const getSelectorProps = (0, import_react.useCallback)(
    (props) => ({
      ...formControlProps,
      ...props,
      value,
      defaultValue: defaultColor,
      fallbackValue,
      onChange: onColorSelectorChange,
      onChangeStart,
      onChangeEnd,
      onSwatchClick: (0, import_utils.handlerAll)(
        onSwatchClick,
        closeOnSelectSwatch ? onClose : void 0
      ),
      format: formatRef.current,
      withPicker,
      withChannel,
      withResult,
      withEyeDropper: withColorSelectorEyeDropper,
      swatchesLabel,
      swatches,
      swatchesColumns,
      variant: colorSelectorVariant,
      size: colorSelectorSize,
      colorScheme: colorSelectorColorScheme
    }),
    [
      formControlProps,
      value,
      fallbackValue,
      defaultColor,
      onColorSelectorChange,
      onChangeStart,
      onChangeEnd,
      onSwatchClick,
      onClose,
      closeOnSelectSwatch,
      formatRef,
      withPicker,
      withChannel,
      withResult,
      withColorSelectorEyeDropper,
      swatchesLabel,
      swatches,
      swatchesColumns,
      colorSelectorColorScheme,
      colorSelectorSize,
      colorSelectorVariant
    ]
  );
  return {
    value,
    eyeDropperSupported,
    allowInput,
    getPopoverProps,
    getContainerProps,
    getFieldProps,
    getSelectorProps,
    getEyeDropperProps
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ColorPickerProvider,
  useColorPicker,
  useColorPickerContext
});
