// @yamada-ui/list/dist/index.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  DecimalList: () => DecimalList,
  DiscList: () => DiscList,
  List: () => List,
  ListIcon: () => ListIcon,
  ListItem: () => ListItem
});
module.exports = __toCommonJS(src_exports);

// src/list.tsx
var import_core = require("@yamada-ui/core");
var import_icon = require("@yamada-ui/icon");
var import_utils = require("@yamada-ui/utils");
var import_jsx_runtime = require("react/jsx-runtime");
var [ListProvider, useList] = (0, import_utils.createContext)({
  name: `ListContext`,
  errorMessage: `useList returned is 'undefined'. Seems you forgot to wrap the components in "<List />" `
});
var List = (0, import_core.forwardRef)((props, ref) => {
  const [styles, mergedProps] = (0, import_core.useMultiComponentStyle)("List", props);
  const {
    className,
    children,
    styleType: listStyleType = "none",
    stylePosition: listStylePosition,
    gap = "fallback(2, 0.5rem)",
    ...rest
  } = (0, import_core.omitThemeProps)(mergedProps);
  const validChildren = (0, import_utils.getValidChildren)(children);
  const css = {
    display: "flex",
    flexDirection: "column",
    gap,
    ms: listStyleType !== "none" ? "1.4em" : void 0,
    "& > li": { ps: listStyleType === "decimal" ? "0.3em" : void 0 },
    ...styles.container
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ListProvider, { value: styles, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_core.ui.ul,
    {
      ref,
      listStyleType,
      listStylePosition,
      role: "list",
      className: (0, import_utils.cx)("ui-list", className),
      __css: css,
      ...rest,
      children: validChildren
    }
  ) });
});
var DiscList = (0, import_core.forwardRef)(({ ...rest }, ref) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(List, { ref, as: "ul", styleType: "disc", ...rest });
});
var DecimalList = (0, import_core.forwardRef)(({ ...rest }, ref) => {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(List, { ref, as: "ol", styleType: "decimal", ms: "1.2em", ...rest });
});
var ListItem = (0, import_core.forwardRef)(
  ({ className, ...rest }, ref) => {
    const styles = useList();
    const css = { ...styles.item };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core.ui.li,
      {
        ref,
        className: (0, import_utils.cx)("ui-list-item", className),
        __css: css,
        ...rest
      }
    );
  }
);
var ListIcon = (0, import_core.forwardRef)(
  ({ className, ...rest }, ref) => {
    const styles = useList();
    const css = { ...styles.icon };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_icon.Icon,
      {
        ref,
        className: (0, import_utils.cx)("ui-list-icon", className),
        role: "presentation",
        __css: css,
        ...rest
      }
    );
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DecimalList,
  DiscList,
  List,
  ListIcon,
  ListItem
});
