// @yamada-ui/ripple/dist/ripple.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/ripple.tsx
var ripple_exports = {};
__export(ripple_exports, {
  Ripple: () => Ripple
});
module.exports = __toCommonJS(ripple_exports);
var import_motion = require("@yamada-ui/motion");
var import_utils = require("@yamada-ui/utils");
var import_jsx_runtime = require("react/jsx-runtime");
var clamp = (value, min, max) => Math.min(Math.max(value, min), max);
var Ripple = ({
  className,
  ripples,
  onAnimationComplete,
  onClear,
  color = "currentColor",
  style,
  isDisabled,
  ...rest
}) => {
  if (isDisabled)
    return null;
  const css = {
    rounded: "fallback(full, 9999px)",
    zIndex: "fallback(kurillin, 9)"
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: ripples.map(({ key, x, y, size }) => {
    const duration = clamp(0.01 * size, 0.2, size > 100 ? 0.75 : 0.5);
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_motion.AnimatePresence, { mode: "popLayout", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_motion.Motion,
      {
        as: "span",
        className: (0, import_utils.cx)("ui-ripple", className),
        initial: { transform: "scale(0)", opacity: 0.35 },
        animate: { transform: "scale(2)", opacity: 0 },
        exit: { opacity: 0 },
        transition: { duration },
        bgColor: color,
        style: {
          position: "absolute",
          transformOrigin: "center",
          pointerEvents: "none",
          left: x,
          top: y,
          width: `${size}px`,
          height: `${size}px`,
          ...style
        },
        __css: css,
        ...rest,
        onAnimationComplete: (0, import_utils.handlerAll)(
          onAnimationComplete,
          () => onClear(key)
        )
      }
    ) }, key);
  }) });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Ripple
});
