// @yamada-ui/progress/dist/index.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  CircleProgress: () => CircleProgress,
  CircleProgressLabel: () => CircleProgressLabel,
  Progress: () => Progress
});
module.exports = __toCommonJS(src_exports);

// src/progress.tsx
var import_core = require("@yamada-ui/core");
var import_use_animation = require("@yamada-ui/use-animation");
var import_utils = require("@yamada-ui/utils");
var import_jsx_runtime = require("react/jsx-runtime");
var [ProgressProvider, useProgress] = (0, import_utils.createContext)({
  name: `ProgressStylesContext`,
  errorMessage: `useProgress returned is 'undefined'. Seems you forgot to wrap the components in "<Progress />" `
});
var Progress = (0, import_core.forwardRef)((props, ref) => {
  var _a;
  const [styles, mergedProps] = (0, import_core.useMultiComponentStyle)("Progress", props);
  const {
    className,
    children,
    value,
    min,
    max,
    hasStripe,
    isStripeAnimation,
    isAnimation,
    speed,
    borderRadius: _borderRadius,
    rounded,
    ...rest
  } = (0, import_core.omitThemeProps)(mergedProps, ["filledTrackColor"]);
  const borderRadius = (_a = _borderRadius != null ? _borderRadius : rounded) != null ? _a : styles.track.borderRadius;
  const css = {
    w: "100%",
    overflow: "hidden",
    pos: "relative",
    ...styles.track
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ProgressProvider, { value: styles, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    import_core.ui.div,
    {
      ref,
      className: (0, import_utils.cx)("ui-progress", className),
      __css: css,
      borderRadius,
      ...rest,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          ProgressFilledTrack,
          {
            min,
            max,
            value,
            hasStripe,
            isStripeAnimation,
            isAnimation,
            speed,
            borderRadius
          }
        ),
        children
      ]
    }
  ) });
});
var ProgressFilledTrack = ({
  value = 0,
  min = 0,
  max = 100,
  hasStripe,
  isStripeAnimation,
  isAnimation,
  speed = "1.4s",
  ...rest
}) => {
  const percent = (0, import_utils.valueToPercent)(value, min, max);
  const styles = useProgress();
  const stripeAnimation = (0, import_use_animation.useAnimation)({
    keyframes: {
      "0%": { bgPosition: "1rem 0" },
      "100%": { bgPosition: "0 0" }
    },
    duration: typeof speed === "string" ? speed : `${speed}s`,
    iterationCount: "infinite",
    timingFunction: "linear"
  });
  const interpolationAnimation = (0, import_use_animation.useAnimation)({
    keyframes: {
      "0%": { left: "-40%" },
      "100%": { left: "100%" }
    },
    duration: typeof speed === "string" ? speed : `${speed}s`,
    iterationCount: "infinite",
    timingFunction: "ease"
  });
  isStripeAnimation = !isAnimation && hasStripe && isStripeAnimation;
  const css = {
    ...isStripeAnimation ? {
      animation: stripeAnimation
    } : {},
    ...isAnimation ? {
      position: "absolute",
      willChange: "left",
      minWidth: "50%",
      animation: interpolationAnimation
    } : {}
  };
  const __css = {
    w: `${percent}%`,
    h: "100%",
    ...styles.filledTrack
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_core.ui.div, { css, __css, ...rest });
};

// src/circle-progress.tsx
var import_core2 = require("@yamada-ui/core");
var import_use_animation2 = require("@yamada-ui/use-animation");
var import_use_token = require("@yamada-ui/use-token");
var import_use_value = require("@yamada-ui/use-value");
var import_utils2 = require("@yamada-ui/utils");
var import_jsx_runtime2 = require("react/jsx-runtime");
var CircleProgress = (0, import_core2.forwardRef)(
  ({
    className,
    children,
    size = "6rem",
    thickness = "0.625rem",
    color = "primary",
    trackColor = "border",
    value = 0,
    min = 0,
    max = 100,
    isAnimation = false,
    isRounded,
    speed = ["1.4s", "2s"],
    ...rest
  }, ref) => {
    var _a, _b;
    size = (_a = (0, import_use_token.useToken)("sizes", (0, import_use_value.useValue)(size))) != null ? _a : size;
    thickness = (_b = (0, import_use_token.useToken)("sizes", (0, import_use_value.useValue)(thickness))) != null ? _b : thickness;
    const isTransparent = value === 0 && !isAnimation;
    const percent = (0, import_utils2.valueToPercent)(value, min, max);
    const interval = !isAnimation ? percent * 2.64 : void 0;
    const animation = (0, import_use_animation2.useAnimation)({
      keyframes: {
        "0%": {
          strokeDasharray: "1, 400",
          strokeDashoffset: "0"
        },
        "50%": {
          strokeDasharray: "400, 400",
          strokeDashoffset: "-100"
        },
        "100%": {
          strokeDasharray: "400, 400",
          strokeDashoffset: "-260"
        }
      },
      duration: typeof speed[0] === "string" ? speed[0] : `${speed[0]}s`,
      iterationCount: "infinite",
      timingFunction: "linear"
    });
    const css = {
      display: "inline-block",
      position: "relative",
      verticalAlign: "middle",
      fontSize: size
    };
    const props = isAnimation ? {
      animation
    } : {
      strokeDashoffset: 66,
      strokeDasharray: interval == null ? void 0 : `${interval} ${264 - interval}`,
      transitionProperty: "stroke-dasharray, stroke",
      transitionDuration: "0.6s",
      transitionTimingFunction: "ease"
    };
    return /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(
      import_core2.ui.div,
      {
        ref,
        className: (0, import_utils2.cx)("ui-circle-progress", className),
        __css: css,
        ...rest,
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime2.jsxs)(Shape, { size, isAnimation, speed, children: [
            /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(Circle, { stroke: trackColor, strokeWidth: thickness }),
            /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
              Circle,
              {
                stroke: color,
                strokeWidth: thickness,
                strokeLinecap: isRounded ? "round" : void 0,
                opacity: isTransparent ? 0 : void 0,
                ...props
              }
            )
          ] }),
          children
        ]
      }
    );
  }
);
var Circle = ({ ...rest }) => /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_core2.ui.circle, { cx: 50, cy: 50, r: 42, fill: "transparent", ...rest });
var Shape = ({ size, isAnimation, speed, ...rest }) => {
  const animation = (0, import_use_animation2.useAnimation)({
    keyframes: {
      "0%": {
        transform: "rotate(0deg)"
      },
      "100%": {
        transform: "rotate(360deg)"
      }
    },
    duration: typeof speed[1] === "string" ? speed[1] : `${speed[1]}s`,
    iterationCount: "infinite",
    timingFunction: "linear"
  });
  const css = {
    display: "block",
    boxSize: size,
    ...isAnimation ? { animation } : {}
  };
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(import_core2.ui.svg, { viewBox: "0 0 100 100", __css: css, ...rest });
};
var CircleProgressLabel = (0, import_core2.ui)("span", {
  baseStyle: {
    position: "absolute",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)",
    width: "100%",
    fontSize: "0.25em",
    textAlign: "center"
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CircleProgress,
  CircleProgressLabel,
  Progress
});
