// @yamada-ui/loading/dist/audio.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/audio.tsx
var audio_exports = {};
__export(audio_exports, {
  Audio: () => Audio
});
module.exports = __toCommonJS(audio_exports);
var import_core = require("@yamada-ui/core");
var import_icon = require("@yamada-ui/icon");
var import_jsx_runtime = require("react/jsx-runtime");
var Audio = (0, import_core.forwardRef)(
  ({ color, duration = "1.4s", ...rest }, ref) => {
    const dur = typeof duration === "string" ? parseFloat(duration) : duration;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icon.Icon, { ref, viewBox: "0 0 55 80", fill: color, ...rest, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("g", { transform: "matrix(1 0 0 -1 0 80)", children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("rect", { width: "10", height: "20", rx: "3", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "animate",
        {
          attributeName: "height",
          begin: "0s",
          dur: `${dur * 3}s`,
          values: "20;45;57;80;64;32;66;45;64;23;66;13;64;56;34;34;2;23;76;79;20",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("rect", { x: "15", width: "10", height: "80", rx: "3", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "animate",
        {
          attributeName: "height",
          begin: "0s",
          dur: `${dur * 1.4}s`,
          values: "80;55;33;5;75;23;73;33;12;14;60;80",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("rect", { x: "30", width: "10", height: "50", rx: "3", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "animate",
        {
          attributeName: "height",
          begin: "0s",
          dur: `${dur}s`,
          values: "50;34;78;23;56;23;34;76;80;54;21;50",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("rect", { x: "45", width: "10", height: "30", rx: "3", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "animate",
        {
          attributeName: "height",
          begin: "0s",
          dur: `${dur * 1.4}s`,
          values: "30;45;13;80;56;72;45;76;34;23;67;30",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) })
    ] }) });
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Audio
});
