// @yamada-ui/loading/dist/chunk-MNPPARYC.mjs
"use client"

// src/oval.tsx
import { forwardRef } from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { jsx, jsxs } from "react/jsx-runtime";
var Oval = forwardRef(
  ({ color, secondaryColor, duration = "1s", ...rest }, ref) => {
    return /* @__PURE__ */ jsx(Icon, { ref, viewBox: "0 0 38 38", stroke: color, ...rest, children: /* @__PURE__ */ jsx("g", { fill: "none", fillRule: "evenodd", children: /* @__PURE__ */ jsxs("g", { transform: "translate(1 1)", strokeWidth: "2", children: [
      /* @__PURE__ */ jsx(
        "circle",
        {
          ...secondaryColor ? { stroke: secondaryColor } : { strokeOpacity: ".5" },
          cx: "18",
          cy: "18",
          r: "18"
        }
      ),
      /* @__PURE__ */ jsx("path", { d: "M36 18c0-9.94-8.06-18-18-18", children: /* @__PURE__ */ jsx(
        "animateTransform",
        {
          attributeName: "transform",
          type: "rotate",
          from: "0 18 18",
          to: "360 18 18",
          dur: duration,
          repeatCount: "indefinite"
        }
      ) })
    ] }) }) });
  }
);

export {
  Oval
};
