// @yamada-ui/loading/dist/chunk-TIRPNXA7.mjs
"use client"

// src/dots.tsx
import { forwardRef } from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { jsx, jsxs } from "react/jsx-runtime";
var Dots = forwardRef(
  ({ color, duration = "1s", ...rest }, ref) => {
    return /* @__PURE__ */ jsxs(Icon, { ref, viewBox: "0 0 120 30", fill: color, ...rest, children: [
      /* @__PURE__ */ jsxs("circle", { cx: "15", cy: "15", r: "15", children: [
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "r",
            from: "15",
            to: "15",
            begin: "0s",
            dur: duration,
            values: "15;9;15",
            calcMode: "linear",
            repeatCount: "indefinite"
          }
        ),
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "fill-opacity",
            from: "1",
            to: "1",
            begin: "0s",
            dur: duration,
            values: "1;.5;1",
            calcMode: "linear",
            repeatCount: "indefinite"
          }
        )
      ] }),
      /* @__PURE__ */ jsxs("circle", { cx: "60", cy: "15", r: "9", fillOpacity: "0.3", children: [
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "r",
            from: "9",
            to: "9",
            begin: "0s",
            dur: duration,
            values: "9;15;9",
            calcMode: "linear",
            repeatCount: "indefinite"
          }
        ),
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "fill-opacity",
            from: "0.5",
            to: "0.5",
            begin: "0s",
            dur: duration,
            values: ".5;1;.5",
            calcMode: "linear",
            repeatCount: "indefinite"
          }
        )
      ] }),
      /* @__PURE__ */ jsxs("circle", { cx: "105", cy: "15", r: "15", children: [
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "r",
            from: "15",
            to: "15",
            begin: "0s",
            dur: duration,
            values: "15;9;15",
            calcMode: "linear",
            repeatCount: "indefinite"
          }
        ),
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "fill-opacity",
            from: "1",
            to: "1",
            begin: "0s",
            dur: duration,
            values: "1;.5;1",
            calcMode: "linear",
            repeatCount: "indefinite"
          }
        )
      ] })
    ] });
  }
);

export {
  Dots
};
