// @yamada-ui/loading/dist/chunk-TP5LE3DK.mjs
"use client"

// src/circles.tsx
import { forwardRef } from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { jsx, jsxs } from "react/jsx-runtime";
var Circles = forwardRef(
  ({ color, duration = "3s", ...rest }, ref) => {
    return /* @__PURE__ */ jsx(Icon, { ref, viewBox: "0 0 58 58", stroke: color, ...rest, children: /* @__PURE__ */ jsx("g", { fill: "none", fillRule: "evenodd", children: /* @__PURE__ */ jsxs("g", { transform: "translate(2 1)", stroke: color, strokeWidth: "1.5", children: [
      /* @__PURE__ */ jsx("circle", { cx: "42.601", cy: "11.462", r: "5", fillOpacity: "1", fill: color, children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "1;0;0;0;0;0;0;0",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "49.063", cy: "27.063", r: "5", fillOpacity: "0", fill: color, children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "0;1;0;0;0;0;0;0",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "42.601", cy: "42.663", r: "5", fillOpacity: "0", fill: color, children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "0;0;1;0;0;0;0;0",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "27", cy: "49.125", r: "5", fillOpacity: "0", fill: color, children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "0;0;0;1;0;0;0;0",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "11.399", cy: "42.663", r: "5", fillOpacity: "0", fill: color, children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "0;0;0;0;1;0;0;0",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "4.938", cy: "27.063", r: "5", fillOpacity: "0", fill: color, children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "0;0;0;0;0;1;0;0",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "11.399", cy: "11.462", r: "5", fillOpacity: "0", fill: color, children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "0;0;0;0;0;0;1;0",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "27", cy: "5", r: "5", fillOpacity: "0", fill: color, children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "0;0;0;0;0;0;0;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) })
    ] }) }) });
  }
);

export {
  Circles
};
