// @yamada-ui/loading/dist/chunk-4JKPWQSY.mjs
"use client"

// src/rings.tsx
import { forwardRef } from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { jsx, jsxs } from "react/jsx-runtime";
var Rings = forwardRef(
  ({ color, duration = "3s", ...rest }, ref) => {
    const dur = typeof duration === "string" ? parseFloat(duration) : duration;
    return /* @__PURE__ */ jsx(Icon, { ref, viewBox: "0 0 45 45", stroke: color, ...rest, children: /* @__PURE__ */ jsxs(
      "g",
      {
        fill: "none",
        fillRule: "evenodd",
        transform: "translate(1 1)",
        strokeWidth: "2",
        children: [
          /* @__PURE__ */ jsxs("circle", { cx: "22", cy: "22", r: "6", strokeOpacity: "0", children: [
            /* @__PURE__ */ jsx(
              "animate",
              {
                attributeName: "r",
                begin: "1.5s",
                dur: `${dur}s`,
                values: "6;22",
                calcMode: "linear",
                repeatCount: "indefinite"
              }
            ),
            /* @__PURE__ */ jsx(
              "animate",
              {
                attributeName: "stroke-opacity",
                begin: "1.5s",
                dur: `${dur}s`,
                values: "1;0",
                calcMode: "linear",
                repeatCount: "indefinite"
              }
            ),
            /* @__PURE__ */ jsx(
              "animate",
              {
                attributeName: "stroke-width",
                begin: "1.5s",
                dur: `${dur}s`,
                values: "2;0",
                calcMode: "linear",
                repeatCount: "indefinite"
              }
            )
          ] }),
          /* @__PURE__ */ jsxs("circle", { cx: "22", cy: "22", r: "6", strokeOpacity: "0", children: [
            /* @__PURE__ */ jsx(
              "animate",
              {
                attributeName: "r",
                begin: "3s",
                dur: `${dur}s`,
                values: "6;22",
                calcMode: "linear",
                repeatCount: "indefinite"
              }
            ),
            /* @__PURE__ */ jsx(
              "animate",
              {
                attributeName: "stroke-opacity",
                begin: "3s",
                dur: `${dur}s`,
                values: "1;0",
                calcMode: "linear",
                repeatCount: "indefinite"
              }
            ),
            /* @__PURE__ */ jsx(
              "animate",
              {
                attributeName: "stroke-width",
                begin: "3s",
                dur: `${dur}s`,
                values: "2;0",
                calcMode: "linear",
                repeatCount: "indefinite"
              }
            )
          ] }),
          /* @__PURE__ */ jsx("circle", { cx: "22", cy: "22", r: "8", children: /* @__PURE__ */ jsx(
            "animate",
            {
              attributeName: "r",
              begin: "0s",
              dur: `${dur / 2}s`,
              values: "6;1;2;3;4;5;6",
              calcMode: "linear",
              repeatCount: "indefinite"
            }
          ) })
        ]
      }
    ) });
  }
);

export {
  Rings
};
