// @yamada-ui/loading/dist/chunk-TIONMRJE.mjs
"use client"

// src/grid.tsx
import { forwardRef } from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { jsx, jsxs } from "react/jsx-runtime";
var Grid = forwardRef(
  ({ color, duration = "1s", ...rest }, ref) => {
    return /* @__PURE__ */ jsxs(Icon, { ref, viewBox: "0 0 105 105", fill: color, ...rest, children: [
      /* @__PURE__ */ jsx("circle", { cx: "12.5", cy: "12.5", r: "12.5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "0s",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "12.5", cy: "52.5", r: "12.5", fillOpacity: ".5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "100ms",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "52.5", cy: "12.5", r: "12.5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "300ms",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "52.5", cy: "52.5", r: "12.5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "600ms",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "92.5", cy: "12.5", r: "12.5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "800ms",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "92.5", cy: "52.5", r: "12.5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "400ms",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "12.5", cy: "92.5", r: "12.5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "700ms",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "52.5", cy: "92.5", r: "12.5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "500ms",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("circle", { cx: "92.5", cy: "92.5", r: "12.5", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "fill-opacity",
          begin: "200ms",
          dur: duration,
          values: "1;.2;1",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) })
    ] });
  }
);

export {
  Grid
};
