// @yamada-ui/loading/dist/chunk-ENUKQWPP.mjs
"use client"

// src/puff.tsx
import { forwardRef } from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { jsx, jsxs } from "react/jsx-runtime";
var Puff = forwardRef(
  ({ color, duration = "1.8s", ...rest }, ref) => {
    return /* @__PURE__ */ jsx(Icon, { ref, viewBox: "0 0 44 44", stroke: color, ...rest, children: /* @__PURE__ */ jsxs("g", { fill: "none", fillRule: "evenodd", strokeWidth: "2", children: [
      /* @__PURE__ */ jsxs("circle", { cx: "22", cy: "22", r: "1", children: [
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "r",
            begin: "0s",
            dur: duration,
            values: "1; 20",
            calcMode: "spline",
            keyTimes: "0; 1",
            keySplines: "0.165, 0.84, 0.44, 1",
            repeatCount: "indefinite"
          }
        ),
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "stroke-opacity",
            begin: "0s",
            dur: duration,
            values: "1; 0",
            calcMode: "spline",
            keyTimes: "0; 1",
            keySplines: "0.3, 0.61, 0.355, 1",
            repeatCount: "indefinite"
          }
        )
      ] }),
      /* @__PURE__ */ jsxs("circle", { cx: "22", cy: "22", r: "1", children: [
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "r",
            begin: "-0.9s",
            dur: duration,
            values: "1; 20",
            calcMode: "spline",
            keyTimes: "0; 1",
            keySplines: "0.165, 0.84, 0.44, 1",
            repeatCount: "indefinite"
          }
        ),
        /* @__PURE__ */ jsx(
          "animate",
          {
            attributeName: "stroke-opacity",
            begin: "-0.9s",
            dur: duration,
            values: "1; 0",
            calcMode: "spline",
            keyTimes: "0; 1",
            keySplines: "0.3, 0.61, 0.355, 1",
            repeatCount: "indefinite"
          }
        )
      ] })
    ] }) });
  }
);

export {
  Puff
};
