// @yamada-ui/loading/dist/chunk-EYRZLOPC.mjs
"use client"

// src/audio.tsx
import { forwardRef } from "@yamada-ui/core";
import { Icon } from "@yamada-ui/icon";
import { jsx, jsxs } from "react/jsx-runtime";
var Audio = forwardRef(
  ({ color, duration = "1.4s", ...rest }, ref) => {
    const dur = typeof duration === "string" ? parseFloat(duration) : duration;
    return /* @__PURE__ */ jsx(Icon, { ref, viewBox: "0 0 55 80", fill: color, ...rest, children: /* @__PURE__ */ jsxs("g", { transform: "matrix(1 0 0 -1 0 80)", children: [
      /* @__PURE__ */ jsx("rect", { width: "10", height: "20", rx: "3", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "height",
          begin: "0s",
          dur: `${dur * 3}s`,
          values: "20;45;57;80;64;32;66;45;64;23;66;13;64;56;34;34;2;23;76;79;20",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("rect", { x: "15", width: "10", height: "80", rx: "3", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "height",
          begin: "0s",
          dur: `${dur * 1.4}s`,
          values: "80;55;33;5;75;23;73;33;12;14;60;80",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("rect", { x: "30", width: "10", height: "50", rx: "3", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "height",
          begin: "0s",
          dur: `${dur}s`,
          values: "50;34;78;23;56;23;34;76;80;54;21;50",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) }),
      /* @__PURE__ */ jsx("rect", { x: "45", width: "10", height: "30", rx: "3", children: /* @__PURE__ */ jsx(
        "animate",
        {
          attributeName: "height",
          begin: "0s",
          dur: `${dur * 1.4}s`,
          values: "30;45;13;80;56;72;45;76;34;23;67;30",
          calcMode: "linear",
          repeatCount: "indefinite"
        }
      ) })
    ] }) });
  }
);

export {
  Audio
};
