// @yamada-ui/use-popper/dist/index.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  popperProperties: () => popperProperties,
  usePopper: () => usePopper
});
module.exports = __toCommonJS(src_exports);
var import_core = require("@popperjs/core");
var import_use_value = require("@yamada-ui/use-value");
var import_utils = require("@yamada-ui/utils");
var import_react = require("react");
var popperProperties = [
  "enabled",
  "offset",
  "gutter",
  "preventOverflow",
  "flip",
  "matchWidth",
  "boundary",
  "eventListeners",
  "strategy",
  "placement",
  "modifiers"
];
var defaultEventListeners = {
  scroll: true,
  resize: true
};
var transforms = {
  top: "bottom center",
  "top-start": "bottom left",
  "top-end": "bottom right",
  bottom: "top center",
  "bottom-start": "top left",
  "bottom-end": "top right",
  left: "right center",
  "left-start": "right top",
  "left-end": "right bottom",
  right: "left center",
  "right-start": "left top",
  "right-end": "left bottom"
};
var usePopper = ({
  enabled = true,
  eventListeners = true,
  matchWidth,
  offset,
  gutter: _gutter = 8,
  flip = true,
  preventOverflow = true,
  boundary = "clippingParents",
  strategy = "absolute",
  placement: _placement = "bottom",
  modifiers
} = {}) => {
  const reference = (0, import_react.useRef)(null);
  const popper = (0, import_react.useRef)(null);
  const instance = (0, import_react.useRef)(null);
  const gutter = (0, import_use_value.useValue)(_gutter);
  const placement = (0, import_use_value.useValue)(_placement);
  const cleanup = (0, import_react.useRef)(() => {
  });
  const setupPopper = (0, import_react.useCallback)(() => {
    var _a;
    if (!enabled || !reference.current || !popper.current)
      return;
    (_a = cleanup.current) == null ? void 0 : _a.call(cleanup);
    const modifierTransformOrigin = {
      name: "transformOrigin",
      enabled: true,
      phase: "write",
      fn: ({ state }) => {
        state.elements.popper.style.setProperty(
          "--popper-transform-origin",
          transforms[state.placement]
        );
      },
      effect: ({ state }) => () => {
        state.elements.popper.style.setProperty(
          "--popper-transform-origin",
          transforms[state.placement]
        );
      }
    };
    const modifierEventListeners = {
      name: "eventListeners",
      ...typeof eventListeners === "object" ? {
        enabled: true,
        options: { ...defaultEventListeners, ...eventListeners }
      } : {
        enabled: eventListeners,
        options: defaultEventListeners
      }
    };
    const modifierOffset = {
      name: "offset",
      options: { offset: offset != null ? offset : [0, gutter] }
    };
    const modifierFlip = {
      name: "flip",
      enabled: !!flip,
      options: { padding: 8 }
    };
    const modifierPreventOverflow = {
      name: "preventOverflow",
      enabled: !!preventOverflow,
      options: { boundary }
    };
    const modifierMatchWidth = {
      name: "matchWidth",
      enabled: !!matchWidth,
      phase: "beforeWrite",
      requires: ["computeStyles"],
      fn: ({ state }) => {
        state.styles.popper.width = `${state.rects.reference.width}px`;
      },
      effect: ({ state }) => () => {
        state.elements.popper.style.width = `${state.elements.reference.offsetWidth}px`;
      }
    };
    instance.current = (0, import_core.createPopper)(reference.current, popper.current, {
      placement,
      modifiers: [
        modifierTransformOrigin,
        modifierMatchWidth,
        modifierEventListeners,
        modifierOffset,
        modifierFlip,
        modifierPreventOverflow,
        ...modifiers != null ? modifiers : []
      ],
      strategy
    });
    instance.current.forceUpdate();
    cleanup.current = instance.current.destroy;
  }, [
    placement,
    enabled,
    modifiers,
    matchWidth,
    eventListeners,
    offset,
    gutter,
    flip,
    preventOverflow,
    boundary,
    strategy
  ]);
  (0, import_react.useEffect)(() => {
    return () => {
      var _a;
      if (reference.current || popper.current)
        return;
      (_a = instance.current) == null ? void 0 : _a.destroy();
      instance.current = null;
    };
  }, []);
  const referenceRef = (0, import_react.useCallback)(
    (el) => {
      reference.current = el;
      setupPopper();
    },
    [setupPopper]
  );
  const getReferenceProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      ...props,
      ref: (0, import_utils.mergeRefs)(referenceRef, ref)
    }),
    [referenceRef]
  );
  const popperRef = (0, import_react.useCallback)(
    (el) => {
      popper.current = el;
      setupPopper();
    },
    [setupPopper]
  );
  const getPopperProps = (0, import_react.useCallback)(
    (props = {}, ref = null) => ({
      ...props,
      ref: (0, import_utils.mergeRefs)(popperRef, ref),
      style: {
        ...props.style,
        position: strategy,
        minWidth: matchWidth ? void 0 : "max-content",
        inset: "0 auto auto 0"
      }
    }),
    [strategy, popperRef, matchWidth]
  );
  return {
    update: () => {
      var _a;
      return (_a = instance.current) == null ? void 0 : _a.update();
    },
    forceUpdate: () => {
      var _a;
      return (_a = instance.current) == null ? void 0 : _a.forceUpdate();
    },
    transformOrigin: "var(--popper-transform-origin)",
    referenceRef,
    popperRef,
    getPopperProps,
    getReferenceProps
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  popperProperties,
  usePopper
});
