// @yamada-ui/native-table/dist/native-table.js
"use client"
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/native-table.tsx
var native_table_exports = {};
__export(native_table_exports, {
  NativeTable: () => NativeTable,
  TableStyleProvider: () => TableStyleProvider,
  useTableStyles: () => useTableStyles
});
module.exports = __toCommonJS(native_table_exports);
var import_core = require("@yamada-ui/core");
var import_utils = require("@yamada-ui/utils");
var import_jsx_runtime = require("react/jsx-runtime");
var [TableStyleProvider, useTableStyles] = (0, import_utils.createContext)({
  name: "TableStyleContext",
  errorMessage: `useTableStyles returned is 'undefined'. Seems you forgot to wrap the components in "<Table />" or "<NativeTable />" or "<PagingTable />"`
});
var NativeTable = (0, import_core.forwardRef)(
  (props, ref) => {
    const [styles, mergedProps] = (0, import_core.useMultiComponentStyle)("NativeTable", props);
    const { className, layout, ...rest } = (0, import_core.omitThemeProps)(mergedProps, [
      "withBorder",
      "withColumnBorders",
      "highlightOnHover"
    ]);
    const css = { tableLayout: layout, ...styles.table };
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TableStyleProvider, { value: styles, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_core.ui.table,
      {
        ref,
        className: (0, import_utils.cx)("ui-table", className),
        __css: css,
        ...rest
      }
    ) });
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NativeTable,
  TableStyleProvider,
  useTableStyles
});
